// @wordpress/components/build/higher-order/with-filters/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = withFilters;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _lodash = require("lodash");

var _hooks = require("@wordpress/hooks");

var _compose = require("@wordpress/compose");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ANIMATION_FRAME_PERIOD = 16;
/**
 * Creates a higher-order component which adds filtering capability to the
 * wrapped component. Filters get applied when the original component is about
 * to be mounted. When a filter is added or removed that matches the hook name,
 * the wrapped component re-renders.
 *
 * @param {string} hookName Hook name exposed to be used by filters.
 *
 * @return {Function} Higher-order component factory.
 */

function withFilters(hookName) {
  return (0, _compose.createHigherOrderComponent)(function (OriginalComponent) {
    var namespace = 'core/with-filters/' + hookName;
    /**
     * The component definition with current filters applied. Each instance
     * reuse this shared reference as an optimization to avoid excessive
     * calls to `applyFilters` when many instances exist.
     *
     * @type {?Component}
     */

    var FilteredComponent;
    /**
     * Initializes the FilteredComponent variable once, if not already
     * assigned. Subsequent calls are effectively a noop.
     */

    function ensureFilteredComponent() {
      if (FilteredComponent === undefined) {
        FilteredComponent = (0, _hooks.applyFilters)(hookName, OriginalComponent);
      }
    }

    var FilteredComponentRenderer = /*#__PURE__*/function (_Component) {
      (0, _inherits2.default)(FilteredComponentRenderer, _Component);

      var _super = _createSuper(FilteredComponentRenderer);

      function FilteredComponentRenderer() {
        var _this;

        (0, _classCallCheck2.default)(this, FilteredComponentRenderer);
        _this = _super.apply(this, arguments);
        ensureFilteredComponent();
        return _this;
      }

      (0, _createClass2.default)(FilteredComponentRenderer, [{
        key: "componentDidMount",
        value: function componentDidMount() {
          FilteredComponentRenderer.instances.push(this); // If there were previously no mounted instances for components
          // filtered on this hook, add the hook handler.

          if (FilteredComponentRenderer.instances.length === 1) {
            (0, _hooks.addAction)('hookRemoved', namespace, onHooksUpdated);
            (0, _hooks.addAction)('hookAdded', namespace, onHooksUpdated);
          }
        }
      }, {
        key: "componentWillUnmount",
        value: function componentWillUnmount() {
          FilteredComponentRenderer.instances = (0, _lodash.without)(FilteredComponentRenderer.instances, this); // If this was the last of the mounted components filtered on
          // this hook, remove the hook handler.

          if (FilteredComponentRenderer.instances.length === 0) {
            (0, _hooks.removeAction)('hookRemoved', namespace);
            (0, _hooks.removeAction)('hookAdded', namespace);
          }
        }
      }, {
        key: "render",
        value: function render() {
          return (0, _element.createElement)(FilteredComponent, this.props);
        }
      }]);
      return FilteredComponentRenderer;
    }(_element.Component);

    FilteredComponentRenderer.instances = [];
    /**
     * Updates the FilteredComponent definition, forcing a render for each
     * mounted instance. This occurs a maximum of once per animation frame.
     */

    var throttledForceUpdate = (0, _lodash.debounce)(function () {
      // Recreate the filtered component, only after delay so that it's
      // computed once, even if many filters added.
      FilteredComponent = (0, _hooks.applyFilters)(hookName, OriginalComponent); // Force each instance to render.

      FilteredComponentRenderer.instances.forEach(function (instance) {
        instance.forceUpdate();
      });
    }, ANIMATION_FRAME_PERIOD);
    /**
     * When a filter is added or removed for the matching hook name, each
     * mounted instance should re-render with the new filters having been
     * applied to the original component.
     *
     * @param {string} updatedHookName Name of the hook that was updated.
     */

    function onHooksUpdated(updatedHookName) {
      if (updatedHookName === hookName) {
        throttledForceUpdate();
      }
    }

    return FilteredComponentRenderer;
  }, 'withFilters');
}
