// @wordpress/components/build/input-control/input-base.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InputBase = InputBase;
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _compose = require("@wordpress/compose");

var _backdrop = _interopRequireDefault(require("./backdrop"));

var _label = _interopRequireDefault(require("./label"));

var _inputControlStyles = require("./styles/input-control-styles");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function useUniqueId(idProp) {
  var instanceId = (0, _compose.useInstanceId)(InputBase);
  var id = "input-base-control-".concat(instanceId);
  return idProp || id;
}

function InputBase(_ref, ref) {
  var children = _ref.children,
      className = _ref.className,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      _ref$hideLabelFromVis = _ref.hideLabelFromVision,
      hideLabelFromVision = _ref$hideLabelFromVis === void 0 ? false : _ref$hideLabelFromVis,
      idProp = _ref.id,
      _ref$isFocused = _ref.isFocused,
      isFocused = _ref$isFocused === void 0 ? false : _ref$isFocused,
      label = _ref.label,
      prefix = _ref.prefix,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? 'default' : _ref$size,
      suffix = _ref.suffix,
      props = (0, _objectWithoutProperties2.default)(_ref, ["children", "className", "disabled", "hideLabelFromVision", "id", "isFocused", "label", "prefix", "size", "suffix"]);
  var id = useUniqueId(idProp);
  return (0, _element.createElement)(_inputControlStyles.Root, (0, _extends2.default)({}, props, {
    className: className,
    isFocused: isFocused,
    ref: ref
  }), (0, _element.createElement)(_inputControlStyles.LabelWrapper, null, (0, _element.createElement)(_label.default, {
    className: "components-input-control__label",
    hideLabelFromVision: hideLabelFromVision,
    htmlFor: id,
    size: size
  }, label)), (0, _element.createElement)(_inputControlStyles.Container, {
    className: "components-input-control__container",
    disabled: disabled,
    isFocused: isFocused
  }, prefix && (0, _element.createElement)(_inputControlStyles.Prefix, {
    className: "components-input-control__prefix"
  }, prefix), children, suffix && (0, _element.createElement)(_inputControlStyles.Suffix, {
    className: "components-input-control__suffix"
  }, suffix), (0, _element.createElement)(_backdrop.default, {
    "aria-hidden": "true",
    disabled: disabled,
    isFocused: isFocused,
    label: label,
    size: size
  })));
}

var _default = (0, _element.forwardRef)(InputBase);

exports.default = _default;
