// @wordpress/block-editor/build/components/page-template-picker/templates/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "About", {
  enumerable: true,
  get: function get() {
    return _about.default;
  }
});
Object.defineProperty(exports, "Blog", {
  enumerable: true,
  get: function get() {
    return _blog.default;
  }
});
Object.defineProperty(exports, "Contact", {
  enumerable: true,
  get: function get() {
    return _contact.default;
  }
});
Object.defineProperty(exports, "Portfolio", {
  enumerable: true,
  get: function get() {
    return _portfolio.default;
  }
});
Object.defineProperty(exports, "Services", {
  enumerable: true,
  get: function get() {
    return _services.default;
  }
});
Object.defineProperty(exports, "Team", {
  enumerable: true,
  get: function get() {
    return _team.default;
  }
});

var _about = _interopRequireDefault(require("./about"));

var _blog = _interopRequireDefault(require("./blog"));

var _contact = _interopRequireDefault(require("./contact"));

var _portfolio = _interopRequireDefault(require("./portfolio"));

var _services = _interopRequireDefault(require("./services"));

var _team = _interopRequireDefault(require("./team"));
