// @wordpress/components/build/mobile/html-text-input/container.ios.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _reactNative = require("react-native");

var _keyboardAvoidingView = _interopRequireDefault(require("../keyboard-avoiding-view"));

var _styleIos = _interopRequireDefault(require("./style.ios.scss"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var HTMLInputContainer = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(HTMLInputContainer, _Component);

  var _super = _createSuper(HTMLInputContainer);

  function HTMLInputContainer() {
    var _this;

    (0, _classCallCheck2.default)(this, HTMLInputContainer);
    _this = _super.apply(this, arguments);
    _this.panResponder = _reactNative.PanResponder.create({
      onStartShouldSetPanResponderCapture: function onStartShouldSetPanResponderCapture() {
        return true;
      },
      onPanResponderMove: function onPanResponderMove(e, gestureState) {
        if (gestureState.dy > 100 && gestureState.dy < 110) {
          //Keyboard.dismiss() and this.textInput.blur() are not working here
          //They require to know the currentlyFocusedID under the hood but
          //during this gesture there's no currentlyFocusedID
          _reactNative.UIManager.blur(e.target);
        }
      }
    });
    return _this;
  }

  (0, _createClass2.default)(HTMLInputContainer, [{
    key: "render",
    value: function render() {
      return (0, _element.createElement)(_keyboardAvoidingView.default, (0, _extends2.default)({
        style: _styleIos.default.keyboardAvoidingView
      }, this.panResponder.panHandlers, {
        parentHeight: this.props.parentHeight
      }), this.props.children);
    }
  }]);
  return HTMLInputContainer;
}(_element.Component);

HTMLInputContainer.scrollEnabled = true;
var _default = HTMLInputContainer;
exports.default = _default;
