// @wordpress/components/build-module/mobile/picker/index.ios.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

/**
 * External dependencies
 */
import { ActionSheetIOS } from 'react-native';
/**
 * WordPress dependencies
 */

import { __ } from '@wordpress/i18n';
import { Component } from '@wordpress/element';

var Picker = /*#__PURE__*/function (_Component) {
  _inherits(Picker, _Component);

  var _super = _createSuper(Picker);

  function Picker() {
    _classCallCheck(this, Picker);

    return _super.apply(this, arguments);
  }

  _createClass(Picker, [{
    key: "presentPicker",
    value: function presentPicker() {
      var _this$props = this.props,
          options = _this$props.options,
          onChange = _this$props.onChange,
          title = _this$props.title,
          destructiveButtonIndex = _this$props.destructiveButtonIndex,
          disabledButtonIndices = _this$props.disabledButtonIndices,
          anchor = _this$props.anchor;
      var labels = options.map(function (_ref) {
        var label = _ref.label;
        return label;
      });
      var fullOptions = [__('Cancel')].concat(labels);
      ActionSheetIOS.showActionSheetWithOptions({
        title: title,
        options: fullOptions,
        cancelButtonIndex: 0,
        destructiveButtonIndex: destructiveButtonIndex,
        disabledButtonIndices: disabledButtonIndices,
        anchor: anchor
      }, function (buttonIndex) {
        if (buttonIndex === 0) {
          return;
        }

        var selected = options[buttonIndex - 1];
        onChange(selected.value);
      });
    }
  }, {
    key: "render",
    value: function render() {
      return null;
    }
  }]);

  return Picker;
}(Component);

export default Picker;
