// @wordpress/components/build/range-control/rail.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = RangeRail;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _mark = _interopRequireDefault(require("./mark"));

var _rangeControlStyles = require("./styles/range-control-styles");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function RangeRail(_ref) {
  var _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      _ref$marks = _ref.marks,
      marks = _ref$marks === void 0 ? false : _ref$marks,
      _ref$min = _ref.min,
      min = _ref$min === void 0 ? 0 : _ref$min,
      _ref$max = _ref.max,
      max = _ref$max === void 0 ? 100 : _ref$max,
      _ref$step = _ref.step,
      step = _ref$step === void 0 ? 1 : _ref$step,
      _ref$value = _ref.value,
      value = _ref$value === void 0 ? 0 : _ref$value,
      restProps = (0, _objectWithoutProperties2.default)(_ref, ["disabled", "marks", "min", "max", "step", "value"]);
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_rangeControlStyles.Rail, (0, _extends2.default)({
    disabled: disabled
  }, restProps)), marks && (0, _element.createElement)(Marks, {
    disabled: disabled,
    marks: marks,
    min: min,
    max: max,
    step: step,
    value: value
  }));
}

function Marks(_ref2) {
  var _ref2$disabled = _ref2.disabled,
      disabled = _ref2$disabled === void 0 ? false : _ref2$disabled,
      _ref2$marks = _ref2.marks,
      marks = _ref2$marks === void 0 ? false : _ref2$marks,
      _ref2$min = _ref2.min,
      min = _ref2$min === void 0 ? 0 : _ref2$min,
      _ref2$max = _ref2.max,
      max = _ref2$max === void 0 ? 100 : _ref2$max,
      _ref2$step = _ref2.step,
      step = _ref2$step === void 0 ? 1 : _ref2$step,
      _ref2$value = _ref2.value,
      value = _ref2$value === void 0 ? 0 : _ref2$value;
  var marksData = useMarks({
    marks: marks,
    min: min,
    max: max,
    step: step,
    value: value
  });
  return (0, _element.createElement)(_rangeControlStyles.MarksWrapper, {
    "aria-hidden": "true",
    className: "components-range-control__marks"
  }, marksData.map(function (mark) {
    return (0, _element.createElement)(_mark.default, (0, _extends2.default)({}, mark, {
      key: mark.key,
      "aria-hidden": "true",
      disabled: disabled
    }));
  }));
}

function useMarks(_ref3) {
  var marks = _ref3.marks,
      _ref3$min = _ref3.min,
      min = _ref3$min === void 0 ? 0 : _ref3$min,
      _ref3$max = _ref3.max,
      max = _ref3$max === void 0 ? 100 : _ref3$max,
      _ref3$step = _ref3.step,
      step = _ref3$step === void 0 ? 1 : _ref3$step,
      _ref3$value = _ref3.value,
      value = _ref3$value === void 0 ? 0 : _ref3$value;
  var isRTL = document.documentElement.dir === 'rtl';

  if (!marks) {
    return [];
  }

  var isCustomMarks = Array.isArray(marks);
  var markCount = Math.round((max - min) / step);
  var marksArray = isCustomMarks ? marks : (0, _toConsumableArray2.default)(Array(markCount + 1)).map(function (_, index) {
    return {
      value: index
    };
  });
  var enhancedMarks = marksArray.map(function (mark, index) {
    var markValue = mark.value !== undefined ? mark.value : value;
    var key = "mark-".concat(index);
    var isFilled = markValue * step + min <= value;
    var offset = "".concat(markValue / markCount * 100, "%");
    var offsetStyle = (0, _defineProperty2.default)({}, isRTL ? 'right' : 'left', offset);
    return _objectSpread(_objectSpread({}, mark), {}, {
      isFilled: isFilled,
      key: key,
      style: offsetStyle
    });
  });
  return enhancedMarks;
}
