// @wordpress/components/build-module/mobile/color-settings/gradient-picker-screen.native.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { View } from 'react-native';
import { useNavigation, useRoute } from '@react-navigation/native';
/**
 * WordPress dependencies
 */

import { __ } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import CustomGradientPicker from '../../custom-gradient-picker';
import NavigationHeader from '../bottom-sheet/navigation-header';

var GradientPickerScreen = function GradientPickerScreen() {
  var navigation = useNavigation();
  var route = useRoute();
  var _route$params = route.params,
      setColor = _route$params.setColor,
      currentValue = _route$params.currentValue,
      isGradientColor = _route$params.isGradientColor;
  return createElement(View, null, createElement(NavigationHeader, {
    screen: __('Customize Gradient'),
    leftButtonOnPress: navigation.goBack
  }), createElement(CustomGradientPicker, {
    setColor: setColor,
    currentValue: currentValue,
    isGradientColor: isGradientColor
  }));
};

export default GradientPickerScreen;
