// @wordpress/components/build-module/mobile/color-settings/picker-screen.native.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { useRoute, useNavigation } from '@react-navigation/native';
import React from 'react';
/**
 * WordPress dependencies
 */

import { useContext, useMemo } from '@wordpress/element';
import { BottomSheetContext } from '@wordpress/components';
/**
 * Internal dependencies
 */

import ColorPicker from '../../color-picker';

var PickerScreen = function PickerScreen() {
  var route = useRoute();
  var navigation = useNavigation();

  var _useContext = useContext(BottomSheetContext),
      onShouldEnableInnerHandling = _useContext.onShouldEnableInnerHandling,
      shouldEnableBottomSheetMaxHeight = _useContext.shouldEnableBottomSheetMaxHeight,
      onHandleClosingBottomSheet = _useContext.onHandleClosingBottomSheet,
      isBottomSheetContentScrolling = _useContext.isBottomSheetContentScrolling,
      shouldEnableBottomSheetScroll = _useContext.shouldEnableBottomSheetScroll,
      onHandleHardwareButtonPress = _useContext.onHandleHardwareButtonPress;

  var _route$params = route.params,
      setColor = _route$params.setColor,
      currentValue = _route$params.currentValue,
      isGradientColor = _route$params.isGradientColor;
  return useMemo(function () {
    return createElement(ColorPicker, {
      onShouldEnableInnerHandling: onShouldEnableInnerHandling,
      shouldEnableBottomSheetMaxHeight: shouldEnableBottomSheetMaxHeight,
      setColor: setColor,
      activeColor: currentValue,
      isGradientColor: isGradientColor,
      onNavigationBack: navigation.goBack,
      onHandleClosingBottomSheet: onHandleClosingBottomSheet,
      isBottomSheetContentScrolling: isBottomSheetContentScrolling,
      shouldEnableBottomSheetScroll: shouldEnableBottomSheetScroll,
      onHandleHardwareButtonPress: onHandleHardwareButtonPress
    });
  }, [setColor, currentValue, isGradientColor, onShouldEnableInnerHandling, shouldEnableBottomSheetMaxHeight, onHandleClosingBottomSheet, isBottomSheetContentScrolling, shouldEnableBottomSheetScroll, onHandleHardwareButtonPress]);
};

export default PickerScreen;
