// @wordpress/components/build/sandbox/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _compose = require("@wordpress/compose");

var _focusableIframe = _interopRequireDefault(require("../focusable-iframe"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var Sandbox = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Sandbox, _Component);

  var _super = _createSuper(Sandbox);

  function Sandbox() {
    var _this;

    (0, _classCallCheck2.default)(this, Sandbox);
    _this = _super.apply(this, arguments);
    _this.trySandbox = _this.trySandbox.bind((0, _assertThisInitialized2.default)(_this));
    _this.checkMessageForResize = _this.checkMessageForResize.bind((0, _assertThisInitialized2.default)(_this));
    _this.iframe = (0, _element.createRef)();
    _this.state = {
      width: 0,
      height: 0
    };
    return _this;
  }

  (0, _createClass2.default)(Sandbox, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      this.trySandbox();

      this.trySandboxWithoutRerender = function () {
        _this2.trySandbox(false);
      }; // This used to be registered using <iframe onLoad={} />, but it made the iframe blank
      // after reordering the containing block. See these two issues for more details:
      // https://github.com/WordPress/gutenberg/issues/6146
      // https://github.com/facebook/react/issues/18752


      this.iframe.current.addEventListener('load', this.trySandboxWithoutRerender, false);
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var forceRerender = prevProps.html !== this.props.html;
      this.trySandbox(forceRerender);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.iframe.current.removeEventListener('load', this.trySandboxWithoutRerender);
    }
  }, {
    key: "isFrameAccessible",
    value: function isFrameAccessible() {
      try {
        return !!this.iframe.current.contentDocument.body;
      } catch (e) {
        return false;
      }
    }
  }, {
    key: "checkMessageForResize",
    value: function checkMessageForResize(event) {
      var iframe = this.iframe.current; // Attempt to parse the message data as JSON if passed as string

      var data = event.data || {};

      if ('string' === typeof data) {
        try {
          data = JSON.parse(data);
        } catch (e) {}
      } // Verify that the mounted element is the source of the message


      if (!iframe || iframe.contentWindow !== event.source) {
        return;
      } // Update the state only if the message is formatted as we expect, i.e.
      // as an object with a 'resize' action, width, and height


      var _data = data,
          action = _data.action,
          width = _data.width,
          height = _data.height;
      var _this$state = this.state,
          oldWidth = _this$state.width,
          oldHeight = _this$state.height;

      if ('resize' === action && (oldWidth !== width || oldHeight !== height)) {
        this.setState({
          width: width,
          height: height
        });
      }
    }
  }, {
    key: "trySandbox",
    value: function trySandbox() {
      var forceRerender = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;

      if (!this.isFrameAccessible()) {
        return;
      }

      var body = this.iframe.current.contentDocument.body;

      if (!forceRerender && null !== body.getAttribute('data-resizable-iframe-connected')) {
        return;
      }

      var observeAndResizeJS = "\n\t\t\t( function() {\n\t\t\t\tvar observer;\n\n\t\t\t\tif ( ! window.MutationObserver || ! document.body || ! window.parent ) {\n\t\t\t\t\treturn;\n\t\t\t\t}\n\n\t\t\t\tfunction sendResize() {\n\t\t\t\t\tvar clientBoundingRect = document.body.getBoundingClientRect();\n\n\t\t\t\t\twindow.parent.postMessage( {\n\t\t\t\t\t\taction: 'resize',\n\t\t\t\t\t\twidth: clientBoundingRect.width,\n\t\t\t\t\t\theight: clientBoundingRect.height,\n\t\t\t\t\t}, '*' );\n\t\t\t\t}\n\n\t\t\t\tobserver = new MutationObserver( sendResize );\n\t\t\t\tobserver.observe( document.body, {\n\t\t\t\t\tattributes: true,\n\t\t\t\t\tattributeOldValue: false,\n\t\t\t\t\tcharacterData: true,\n\t\t\t\t\tcharacterDataOldValue: false,\n\t\t\t\t\tchildList: true,\n\t\t\t\t\tsubtree: true\n\t\t\t\t} );\n\n\t\t\t\twindow.addEventListener( 'load', sendResize, true );\n\n\t\t\t\t// Hack: Remove viewport unit styles, as these are relative\n\t\t\t\t// the iframe root and interfere with our mechanism for\n\t\t\t\t// determining the unconstrained page bounds.\n\t\t\t\tfunction removeViewportStyles( ruleOrNode ) {\n\t\t\t\t\tif( ruleOrNode.style ) {\n\t\t\t\t\t\t[ 'width', 'height', 'minHeight', 'maxHeight' ].forEach( function( style ) {\n\t\t\t\t\t\t\tif ( /^\\d+(vmin|vmax|vh|vw)$/.test( ruleOrNode.style[ style ] ) ) {\n\t\t\t\t\t\t\t\truleOrNode.style[ style ] = '';\n\t\t\t\t\t\t\t}\n\t\t\t\t\t\t} );\n\t\t\t\t\t}\n\t\t\t\t}\n\n\t\t\t\tArray.prototype.forEach.call( document.querySelectorAll( '[style]' ), removeViewportStyles );\n\t\t\t\tArray.prototype.forEach.call( document.styleSheets, function( stylesheet ) {\n\t\t\t\t\tArray.prototype.forEach.call( stylesheet.cssRules || stylesheet.rules, removeViewportStyles );\n\t\t\t\t} );\n\n\t\t\t\tdocument.body.style.position = 'absolute';\n\t\t\t\tdocument.body.style.width = '100%';\n\t\t\t\tdocument.body.setAttribute( 'data-resizable-iframe-connected', '' );\n\n\t\t\t\tsendResize();\n\n\t\t\t\t// Resize events can change the width of elements with 100% width, but we don't\n\t\t\t\t// get an DOM mutations for that, so do the resize when the window is resized, too.\n\t\t\t\twindow.addEventListener( 'resize', sendResize, true );\n\t\t} )();";
      var style = "\n\t\t\tbody {\n\t\t\t\tmargin: 0;\n\t\t\t}\n\t\t\thtml,\n\t\t\tbody,\n\t\t\tbody > div,\n\t\t\tbody > div > iframe {\n\t\t\t\twidth: 100%;\n\t\t\t}\n\t\t\thtml.wp-has-aspect-ratio,\n\t\t\tbody.wp-has-aspect-ratio,\n\t\t\tbody.wp-has-aspect-ratio > div,\n\t\t\tbody.wp-has-aspect-ratio > div > iframe {\n\t\t\t\theight: 100%;\n\t\t\t\toverflow: hidden; /* If it has an aspect ratio, it shouldn't scroll. */\n\t\t\t}\n\t\t\tbody > div > * {\n\t\t\t\tmargin-top: 0 !important; /* Has to have !important to override inline styles. */\n\t\t\t\tmargin-bottom: 0 !important;\n\t\t\t}\n\t\t"; // put the html snippet into a html document, and then write it to the iframe's document
      // we can use this in the future to inject custom styles or scripts.
      // Scripts go into the body rather than the head, to support embedded content such as Instagram
      // that expect the scripts to be part of the body.

      var htmlDoc = (0, _element.createElement)("html", {
        lang: document.documentElement.lang,
        className: this.props.type
      }, (0, _element.createElement)("head", null, (0, _element.createElement)("title", null, this.props.title), (0, _element.createElement)("style", {
        dangerouslySetInnerHTML: {
          __html: style
        }
      }), this.props.styles && this.props.styles.map(function (rules, i) {
        return (0, _element.createElement)("style", {
          key: i,
          dangerouslySetInnerHTML: {
            __html: rules
          }
        });
      })), (0, _element.createElement)("body", {
        "data-resizable-iframe-connected": "data-resizable-iframe-connected",
        className: this.props.type
      }, (0, _element.createElement)("div", {
        dangerouslySetInnerHTML: {
          __html: this.props.html
        }
      }), (0, _element.createElement)("script", {
        type: "text/javascript",
        dangerouslySetInnerHTML: {
          __html: observeAndResizeJS
        }
      }), this.props.scripts && this.props.scripts.map(function (src) {
        return (0, _element.createElement)("script", {
          key: src,
          src: src
        });
      }))); // writing the document like this makes it act in the same way as if it was
      // loaded over the network, so DOM creation and mutation, script execution, etc.
      // all work as expected

      var iframeDocument = this.iframe.current.contentWindow.document;
      iframeDocument.open();
      iframeDocument.write('<!DOCTYPE html>' + (0, _element.renderToString)(htmlDoc));
      iframeDocument.close();
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          title = _this$props.title,
          onFocus = _this$props.onFocus;
      return (0, _element.createElement)(_focusableIframe.default, {
        iframeRef: this.iframe,
        title: title,
        className: "components-sandbox",
        sandbox: "allow-scripts allow-same-origin allow-presentation",
        onFocus: onFocus,
        width: Math.ceil(this.state.width),
        height: Math.ceil(this.state.height)
      });
    }
  }], [{
    key: "defaultProps",
    get: function get() {
      return {
        html: '',
        title: ''
      };
    }
  }]);
  return Sandbox;
}(_element.Component);

Sandbox = (0, _compose.withGlobalEvents)({
  message: 'checkMessageForResize'
})(Sandbox);
var _default = Sandbox;
exports.default = _default;
