// @wordpress/components/build/navigation/back-button/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = NavigationBackButton;

var _element = require("@wordpress/element");

var _classnames = _interopRequireDefault(require("classnames"));

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _context = require("../context");

var _navigationStyles = require("../styles/navigation-styles");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function NavigationBackButton(_ref) {
  var _navigationTree$getMe;

  var backButtonLabel = _ref.backButtonLabel,
      className = _ref.className,
      href = _ref.href,
      _onClick = _ref.onClick,
      parentMenu = _ref.parentMenu;

  var _useNavigationContext = (0, _context.useNavigationContext)(),
      setActiveMenu = _useNavigationContext.setActiveMenu,
      navigationTree = _useNavigationContext.navigationTree;

  var classes = (0, _classnames.default)('components-navigation__back-button', className);
  var parentMenuTitle = (_navigationTree$getMe = navigationTree.getMenu(parentMenu)) === null || _navigationTree$getMe === void 0 ? void 0 : _navigationTree$getMe.title;
  return (0, _element.createElement)(_navigationStyles.MenuBackButtonUI, {
    className: classes,
    isTertiary: true,
    href: href,
    onClick: function onClick() {
      return parentMenu ? setActiveMenu(parentMenu, 'right') : _onClick;
    }
  }, (0, _element.createElement)(_icons.Icon, {
    icon: _icons.chevronLeft
  }), backButtonLabel || parentMenuTitle || (0, _i18n.__)('Back'));
}
