// @wordpress/block-editor/build/components/index.native.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  BlockAlignmentToolbar: true,
  BlockContextProvider: true,
  BlockControls: true,
  BlockEdit: true,
  useBlockEditContext: true,
  BlockFormatControls: true,
  BlockIcon: true,
  BlockVerticalAlignmentToolbar: true,
  AlignmentToolbar: true,
  InnerBlocks: true,
  InspectorAdvancedControls: true,
  InspectorControls: true,
  __experimentalLineHeightControl: true,
  PlainText: true,
  RichText: true,
  RichTextShortcut: true,
  RichTextToolbarButton: true,
  __unstableRichTextInputEvent: true,
  MediaPlaceholder: true,
  MediaUpload: true,
  MEDIA_TYPE_IMAGE: true,
  MEDIA_TYPE_VIDEO: true,
  MediaUploadProgress: true,
  URLInput: true,
  BlockInvalidWarning: true,
  BlockCaption: true,
  Caption: true,
  PanelColorSettings: true,
  __experimentalPanelColorGradientSettings: true,
  __experimentalUseEditorFeature: true,
  BottomSheetSettings: true,
  BlockSettingsButton: true,
  blockSettingsScreens: true,
  VideoPlayer: true,
  VIDEO_ASPECT_RATIO: true,
  __experimentalPageTemplatePicker: true,
  __experimentalWithPageTemplatePicker: true,
  Preview: true,
  BlockList: true,
  BlockMover: true,
  BlockToolbar: true,
  BlockVariationPicker: true,
  BlockStyles: true,
  DefaultBlockAppender: true,
  __unstableEditorStyles: true,
  Inserter: true,
  __experimentalBlock: true,
  __experimentalUseBlockWrapperProps: true,
  FloatingToolbar: true,
  BlockEditorProvider: true
};
Object.defineProperty(exports, "BlockAlignmentToolbar", {
  enumerable: true,
  get: function get() {
    return _blockAlignmentToolbar.default;
  }
});
Object.defineProperty(exports, "BlockContextProvider", {
  enumerable: true,
  get: function get() {
    return _blockContext.BlockContextProvider;
  }
});
Object.defineProperty(exports, "BlockControls", {
  enumerable: true,
  get: function get() {
    return _blockControls.default;
  }
});
Object.defineProperty(exports, "BlockEdit", {
  enumerable: true,
  get: function get() {
    return _blockEdit.default;
  }
});
Object.defineProperty(exports, "useBlockEditContext", {
  enumerable: true,
  get: function get() {
    return _blockEdit.useBlockEditContext;
  }
});
Object.defineProperty(exports, "BlockFormatControls", {
  enumerable: true,
  get: function get() {
    return _blockFormatControls.default;
  }
});
Object.defineProperty(exports, "BlockIcon", {
  enumerable: true,
  get: function get() {
    return _blockIcon.default;
  }
});
Object.defineProperty(exports, "BlockVerticalAlignmentToolbar", {
  enumerable: true,
  get: function get() {
    return _blockVerticalAlignmentToolbar.default;
  }
});
Object.defineProperty(exports, "AlignmentToolbar", {
  enumerable: true,
  get: function get() {
    return _alignmentToolbar.default;
  }
});
Object.defineProperty(exports, "InnerBlocks", {
  enumerable: true,
  get: function get() {
    return _innerBlocks.default;
  }
});
Object.defineProperty(exports, "InspectorAdvancedControls", {
  enumerable: true,
  get: function get() {
    return _inspectorAdvancedControls.default;
  }
});
Object.defineProperty(exports, "InspectorControls", {
  enumerable: true,
  get: function get() {
    return _inspectorControls.default;
  }
});
Object.defineProperty(exports, "__experimentalLineHeightControl", {
  enumerable: true,
  get: function get() {
    return _lineHeightControl.default;
  }
});
Object.defineProperty(exports, "PlainText", {
  enumerable: true,
  get: function get() {
    return _plainText.default;
  }
});
Object.defineProperty(exports, "RichText", {
  enumerable: true,
  get: function get() {
    return _richText.default;
  }
});
Object.defineProperty(exports, "RichTextShortcut", {
  enumerable: true,
  get: function get() {
    return _richText.RichTextShortcut;
  }
});
Object.defineProperty(exports, "RichTextToolbarButton", {
  enumerable: true,
  get: function get() {
    return _richText.RichTextToolbarButton;
  }
});
Object.defineProperty(exports, "__unstableRichTextInputEvent", {
  enumerable: true,
  get: function get() {
    return _richText.__unstableRichTextInputEvent;
  }
});
Object.defineProperty(exports, "MediaPlaceholder", {
  enumerable: true,
  get: function get() {
    return _mediaPlaceholder.default;
  }
});
Object.defineProperty(exports, "MediaUpload", {
  enumerable: true,
  get: function get() {
    return _mediaUpload.default;
  }
});
Object.defineProperty(exports, "MEDIA_TYPE_IMAGE", {
  enumerable: true,
  get: function get() {
    return _mediaUpload.MEDIA_TYPE_IMAGE;
  }
});
Object.defineProperty(exports, "MEDIA_TYPE_VIDEO", {
  enumerable: true,
  get: function get() {
    return _mediaUpload.MEDIA_TYPE_VIDEO;
  }
});
Object.defineProperty(exports, "MediaUploadProgress", {
  enumerable: true,
  get: function get() {
    return _mediaUploadProgress.default;
  }
});
Object.defineProperty(exports, "URLInput", {
  enumerable: true,
  get: function get() {
    return _urlInput.default;
  }
});
Object.defineProperty(exports, "BlockInvalidWarning", {
  enumerable: true,
  get: function get() {
    return _blockInvalidWarning.default;
  }
});
Object.defineProperty(exports, "BlockCaption", {
  enumerable: true,
  get: function get() {
    return _blockCaption.default;
  }
});
Object.defineProperty(exports, "Caption", {
  enumerable: true,
  get: function get() {
    return _caption.default;
  }
});
Object.defineProperty(exports, "PanelColorSettings", {
  enumerable: true,
  get: function get() {
    return _panelColorSettings.default;
  }
});
Object.defineProperty(exports, "__experimentalPanelColorGradientSettings", {
  enumerable: true,
  get: function get() {
    return _panelColorGradientSettings.default;
  }
});
Object.defineProperty(exports, "__experimentalUseEditorFeature", {
  enumerable: true,
  get: function get() {
    return _useEditorFeature.default;
  }
});
Object.defineProperty(exports, "BottomSheetSettings", {
  enumerable: true,
  get: function get() {
    return _blockSettings.BottomSheetSettings;
  }
});
Object.defineProperty(exports, "BlockSettingsButton", {
  enumerable: true,
  get: function get() {
    return _blockSettings.BlockSettingsButton;
  }
});
Object.defineProperty(exports, "blockSettingsScreens", {
  enumerable: true,
  get: function get() {
    return _blockSettings.blockSettingsScreens;
  }
});
Object.defineProperty(exports, "VideoPlayer", {
  enumerable: true,
  get: function get() {
    return _videoPlayer.default;
  }
});
Object.defineProperty(exports, "VIDEO_ASPECT_RATIO", {
  enumerable: true,
  get: function get() {
    return _videoPlayer.VIDEO_ASPECT_RATIO;
  }
});
Object.defineProperty(exports, "__experimentalPageTemplatePicker", {
  enumerable: true,
  get: function get() {
    return _pageTemplatePicker.__experimentalPageTemplatePicker;
  }
});
Object.defineProperty(exports, "__experimentalWithPageTemplatePicker", {
  enumerable: true,
  get: function get() {
    return _pageTemplatePicker.__experimentalWithPageTemplatePicker;
  }
});
Object.defineProperty(exports, "Preview", {
  enumerable: true,
  get: function get() {
    return _pageTemplatePicker.Preview;
  }
});
Object.defineProperty(exports, "BlockList", {
  enumerable: true,
  get: function get() {
    return _blockList.default;
  }
});
Object.defineProperty(exports, "BlockMover", {
  enumerable: true,
  get: function get() {
    return _blockMover.default;
  }
});
Object.defineProperty(exports, "BlockToolbar", {
  enumerable: true,
  get: function get() {
    return _blockToolbar.default;
  }
});
Object.defineProperty(exports, "BlockVariationPicker", {
  enumerable: true,
  get: function get() {
    return _blockVariationPicker.default;
  }
});
Object.defineProperty(exports, "BlockStyles", {
  enumerable: true,
  get: function get() {
    return _blockStyles.default;
  }
});
Object.defineProperty(exports, "DefaultBlockAppender", {
  enumerable: true,
  get: function get() {
    return _defaultBlockAppender.default;
  }
});
Object.defineProperty(exports, "__unstableEditorStyles", {
  enumerable: true,
  get: function get() {
    return _editorStyles.default;
  }
});
Object.defineProperty(exports, "Inserter", {
  enumerable: true,
  get: function get() {
    return _inserter.default;
  }
});
Object.defineProperty(exports, "__experimentalBlock", {
  enumerable: true,
  get: function get() {
    return _blockWrapper.Block;
  }
});
Object.defineProperty(exports, "__experimentalUseBlockWrapperProps", {
  enumerable: true,
  get: function get() {
    return _blockWrapper.useBlockWrapperProps;
  }
});
Object.defineProperty(exports, "FloatingToolbar", {
  enumerable: true,
  get: function get() {
    return _floatingToolbar.default;
  }
});
Object.defineProperty(exports, "BlockEditorProvider", {
  enumerable: true,
  get: function get() {
    return _provider.default;
  }
});

var _blockAlignmentToolbar = _interopRequireDefault(require("./block-alignment-toolbar"));

var _blockContext = require("./block-context");

var _blockControls = _interopRequireDefault(require("./block-controls"));

var _blockEdit = _interopRequireWildcard(require("./block-edit"));

var _blockFormatControls = _interopRequireDefault(require("./block-format-controls"));

var _blockIcon = _interopRequireDefault(require("./block-icon"));

var _blockVerticalAlignmentToolbar = _interopRequireDefault(require("./block-vertical-alignment-toolbar"));

var _colors = require("./colors");

Object.keys(_colors).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _colors[key];
    }
  });
});

var _gradients = require("./gradients");

Object.keys(_gradients).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _gradients[key];
    }
  });
});

var _fontSizes = require("./font-sizes");

Object.keys(_fontSizes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _fontSizes[key];
    }
  });
});

var _alignmentToolbar = _interopRequireDefault(require("./alignment-toolbar"));

var _innerBlocks = _interopRequireDefault(require("./inner-blocks"));

var _inspectorAdvancedControls = _interopRequireDefault(require("./inspector-advanced-controls"));

var _inspectorControls = _interopRequireDefault(require("./inspector-controls"));

var _lineHeightControl = _interopRequireDefault(require("./line-height-control"));

var _plainText = _interopRequireDefault(require("./plain-text"));

var _richText = _interopRequireWildcard(require("./rich-text"));

var _mediaPlaceholder = _interopRequireDefault(require("./media-placeholder"));

var _mediaUpload = _interopRequireWildcard(require("./media-upload"));

var _mediaUploadProgress = _interopRequireDefault(require("./media-upload-progress"));

var _urlInput = _interopRequireDefault(require("./url-input"));

var _blockInvalidWarning = _interopRequireDefault(require("./block-list/block-invalid-warning"));

var _blockCaption = _interopRequireDefault(require("./block-caption"));

var _caption = _interopRequireDefault(require("./caption"));

var _panelColorSettings = _interopRequireDefault(require("./panel-color-settings"));

var _panelColorGradientSettings = _interopRequireDefault(require("./colors-gradients/panel-color-gradient-settings"));

var _useEditorFeature = _interopRequireDefault(require("./use-editor-feature"));

var _blockSettings = require("./block-settings");

var _videoPlayer = _interopRequireWildcard(require("./video-player"));

var _pageTemplatePicker = require("./page-template-picker");

var _blockList = _interopRequireDefault(require("./block-list"));

var _blockMover = _interopRequireDefault(require("./block-mover"));

var _blockToolbar = _interopRequireDefault(require("./block-toolbar"));

var _blockVariationPicker = _interopRequireDefault(require("./block-variation-picker"));

var _blockStyles = _interopRequireDefault(require("./block-styles"));

var _defaultBlockAppender = _interopRequireDefault(require("./default-block-appender"));

var _editorStyles = _interopRequireDefault(require("./editor-styles"));

var _inserter = _interopRequireDefault(require("./inserter"));

var _blockWrapper = require("./block-list/block-wrapper");

var _floatingToolbar = _interopRequireDefault(require("./floating-toolbar"));

var _provider = _interopRequireDefault(require("./provider"));
