// @wordpress/components/build/custom-gradient-picker/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GRADIENT_OPTIONS = exports.HORIZONTAL_GRADIENT_ORIENTATION = exports.DEFAULT_LINEAR_GRADIENT_ANGLE = exports.COLOR_POPOVER_PROPS = exports.DEFAULT_GRADIENT = exports.MINIMUM_SIGNIFICANT_MOVE = exports.KEYBOARD_CONTROL_POINT_VARIATION = exports.MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT = exports.MINIMUM_DISTANCE_BETWEEN_POINTS = exports.MINIMUM_ABSOLUTE_LEFT_POSITION = exports.MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_MARKER = exports.GRADIENT_MARKERS_WIDTH = exports.INSERT_POINT_WIDTH = void 0;

var _i18n = require("@wordpress/i18n");

/**
 * WordPress dependencies
 */
var INSERT_POINT_WIDTH = 23;
exports.INSERT_POINT_WIDTH = INSERT_POINT_WIDTH;
var GRADIENT_MARKERS_WIDTH = 18;
exports.GRADIENT_MARKERS_WIDTH = GRADIENT_MARKERS_WIDTH;
var MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_MARKER = (INSERT_POINT_WIDTH + GRADIENT_MARKERS_WIDTH) / 2;
exports.MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_MARKER = MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_MARKER;
var MINIMUM_ABSOLUTE_LEFT_POSITION = 5;
exports.MINIMUM_ABSOLUTE_LEFT_POSITION = MINIMUM_ABSOLUTE_LEFT_POSITION;
var MINIMUM_DISTANCE_BETWEEN_POINTS = 0;
exports.MINIMUM_DISTANCE_BETWEEN_POINTS = MINIMUM_DISTANCE_BETWEEN_POINTS;
var MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT = 10;
exports.MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT = MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT;
var KEYBOARD_CONTROL_POINT_VARIATION = MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT;
exports.KEYBOARD_CONTROL_POINT_VARIATION = KEYBOARD_CONTROL_POINT_VARIATION;
var MINIMUM_SIGNIFICANT_MOVE = 5;
exports.MINIMUM_SIGNIFICANT_MOVE = MINIMUM_SIGNIFICANT_MOVE;
var DEFAULT_GRADIENT = 'linear-gradient(135deg, rgba(6, 147, 227, 1) 0%, rgb(155, 81, 224) 100%)';
exports.DEFAULT_GRADIENT = DEFAULT_GRADIENT;
var COLOR_POPOVER_PROPS = {
  className: 'components-custom-gradient-picker__color-picker-popover',
  position: 'top'
};
exports.COLOR_POPOVER_PROPS = COLOR_POPOVER_PROPS;
var DEFAULT_LINEAR_GRADIENT_ANGLE = 180;
exports.DEFAULT_LINEAR_GRADIENT_ANGLE = DEFAULT_LINEAR_GRADIENT_ANGLE;
var HORIZONTAL_GRADIENT_ORIENTATION = {
  type: 'angular',
  value: 90
};
exports.HORIZONTAL_GRADIENT_ORIENTATION = HORIZONTAL_GRADIENT_ORIENTATION;
var GRADIENT_OPTIONS = [{
  value: 'linear-gradient',
  label: (0, _i18n.__)('Linear')
}, {
  value: 'radial-gradient',
  label: (0, _i18n.__)('Radial')
}];
exports.GRADIENT_OPTIONS = GRADIENT_OPTIONS;
