// @wordpress/components/build/custom-gradient-picker/control-points.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ControlPoints;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _classnames = _interopRequireDefault(require("classnames"));

var _i18n = require("@wordpress/i18n");

var _compose = require("@wordpress/compose");

var _button = _interopRequireDefault(require("../button"));

var _colorPicker = _interopRequireDefault(require("../color-picker"));

var _dropdown = _interopRequireDefault(require("../dropdown"));

var _visuallyHidden = _interopRequireDefault(require("../visually-hidden"));

var _utils = require("./utils");

var _constants = require("./constants");

var _keyboardShortcuts = _interopRequireDefault(require("../keyboard-shortcuts"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var ControlPointKeyboardMove = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(ControlPointKeyboardMove, _Component);

  var _super = _createSuper(ControlPointKeyboardMove);

  function ControlPointKeyboardMove() {
    var _this;

    (0, _classCallCheck2.default)(this, ControlPointKeyboardMove);
    _this = _super.apply(this, arguments);
    _this.increase = _this.increase.bind((0, _assertThisInitialized2.default)(_this));
    _this.decrease = _this.decrease.bind((0, _assertThisInitialized2.default)(_this));
    _this.shortcuts = {
      right: _this.increase,
      left: _this.decrease
    };
    return _this;
  }

  (0, _createClass2.default)(ControlPointKeyboardMove, [{
    key: "increase",
    value: function increase(event) {
      // Stop propagation of the key press event to avoid focus moving
      // to another editor area.
      event.stopPropagation();
      var _this$props = this.props,
          gradientIndex = _this$props.gradientIndex,
          onChange = _this$props.onChange,
          gradientAST = _this$props.gradientAST;
      onChange((0, _utils.getGradientWithPositionAtIndexIncreased)(gradientAST, gradientIndex));
    }
  }, {
    key: "decrease",
    value: function decrease(event) {
      // Stop propagation of the key press event to avoid focus moving
      // to another editor area.
      event.stopPropagation();
      var _this$props2 = this.props,
          gradientIndex = _this$props2.gradientIndex,
          onChange = _this$props2.onChange,
          gradientAST = _this$props2.gradientAST;
      onChange((0, _utils.getGradientWithPositionAtIndexDecreased)(gradientAST, gradientIndex));
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children;
      return (0, _element.createElement)(_keyboardShortcuts.default, {
        shortcuts: this.shortcuts
      }, children);
    }
  }]);
  return ControlPointKeyboardMove;
}(_element.Component);

function ControlPointButton(_ref) {
  var isOpen = _ref.isOpen,
      position = _ref.position,
      color = _ref.color,
      onChange = _ref.onChange,
      gradientIndex = _ref.gradientIndex,
      gradientAST = _ref.gradientAST,
      additionalProps = (0, _objectWithoutProperties2.default)(_ref, ["isOpen", "position", "color", "onChange", "gradientIndex", "gradientAST"]);
  var instanceId = (0, _compose.useInstanceId)(ControlPointButton);
  var descriptionId = "components-custom-gradient-picker__control-point-button-description-".concat(instanceId);
  return (0, _element.createElement)(ControlPointKeyboardMove, {
    onChange: onChange,
    gradientIndex: gradientIndex,
    gradientAST: gradientAST
  }, (0, _element.createElement)(_button.default, (0, _extends2.default)({
    "aria-label": (0, _i18n.sprintf)( // translators: %1$s: gradient position e.g: 70%, %2$s: gradient color code e.g: rgb(52,121,151).
    (0, _i18n.__)('Gradient control point at position %1$s with color code %2$s.'), position, color),
    "aria-describedby": descriptionId,
    "aria-haspopup": "true",
    "aria-expanded": isOpen,
    className: (0, _classnames.default)('components-custom-gradient-picker__control-point-button', {
      'is-active': isOpen
    }),
    style: {
      left: position
    }
  }, additionalProps)), (0, _element.createElement)(_visuallyHidden.default, {
    id: descriptionId
  }, (0, _i18n.__)('Use your left or right arrow keys or drag and drop with the mouse to change the gradient position. Press the button to change the color or remove the control point.')));
}

function ControlPoints(_ref2) {
  var gradientPickerDomRef = _ref2.gradientPickerDomRef,
      ignoreMarkerPosition = _ref2.ignoreMarkerPosition,
      markerPoints = _ref2.markerPoints,
      onChange = _ref2.onChange,
      gradientAST = _ref2.gradientAST,
      onStartControlPointChange = _ref2.onStartControlPointChange,
      onStopControlPointChange = _ref2.onStopControlPointChange;
  var controlPointMoveState = (0, _element.useRef)();

  var onMouseMove = function onMouseMove(event) {
    var relativePosition = (0, _utils.getHorizontalRelativeGradientPosition)(event.clientX, gradientPickerDomRef.current, _constants.GRADIENT_MARKERS_WIDTH);
    var _controlPointMoveStat = controlPointMoveState.current,
        referenceGradientAST = _controlPointMoveStat.gradientAST,
        position = _controlPointMoveStat.position,
        significantMoveHappened = _controlPointMoveStat.significantMoveHappened;

    if (!significantMoveHappened) {
      var initialPosition = referenceGradientAST.colorStops[position].length.value;

      if (Math.abs(initialPosition - relativePosition) >= _constants.MINIMUM_SIGNIFICANT_MOVE) {
        controlPointMoveState.current.significantMoveHappened = true;
      }
    }

    if (!(0, _utils.isControlPointOverlapping)(referenceGradientAST, relativePosition, position)) {
      onChange((0, _utils.getGradientWithPositionAtIndexChanged)(referenceGradientAST, position, relativePosition));
    }
  };

  var cleanEventListeners = function cleanEventListeners() {
    if (window && window.removeEventListener && controlPointMoveState.current && controlPointMoveState.current.listenersActivated) {
      window.removeEventListener('mousemove', onMouseMove);
      window.removeEventListener('mouseup', cleanEventListeners);
      onStopControlPointChange();
      controlPointMoveState.current.listenersActivated = false;
    }
  };

  (0, _element.useEffect)(function () {
    return function () {
      cleanEventListeners();
    };
  }, []);
  return markerPoints.map(function (point, index) {
    return point && ignoreMarkerPosition !== point.positionValue && (0, _element.createElement)(_dropdown.default, {
      key: index,
      onClose: onStopControlPointChange,
      renderToggle: function renderToggle(_ref3) {
        var isOpen = _ref3.isOpen,
            onToggle = _ref3.onToggle;
        return (0, _element.createElement)(ControlPointButton, {
          key: index,
          onClick: function onClick() {
            if (controlPointMoveState.current && controlPointMoveState.current.significantMoveHappened) {
              return;
            }

            if (isOpen) {
              onStopControlPointChange();
            } else {
              onStartControlPointChange();
            }

            onToggle();
          },
          onMouseDown: function onMouseDown() {
            if (window && window.addEventListener) {
              controlPointMoveState.current = {
                gradientAST: gradientAST,
                position: index,
                significantMoveHappened: false,
                listenersActivated: true
              };
              onStartControlPointChange();
              window.addEventListener('mousemove', onMouseMove);
              window.addEventListener('mouseup', cleanEventListeners);
            }
          },
          isOpen: isOpen,
          position: point.position,
          color: point.color,
          onChange: onChange,
          gradientAST: gradientAST,
          gradientIndex: index
        });
      },
      renderContent: function renderContent(_ref4) {
        var onClose = _ref4.onClose;
        return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_colorPicker.default, {
          color: point.color,
          onChangeComplete: function onChangeComplete(_ref5) {
            var rgb = _ref5.rgb;
            onChange((0, _utils.getGradientWithColorAtIndexChanged)(gradientAST, index, rgb));
          }
        }), (0, _element.createElement)(_button.default, {
          className: "components-custom-gradient-picker__remove-control-point",
          onClick: function onClick() {
            onChange((0, _utils.getGradientWithControlPointRemoved)(gradientAST, index));
            onClose();
          },
          isLink: true
        }, (0, _i18n.__)('Remove Control Point')));
      },
      popoverProps: _constants.COLOR_POPOVER_PROPS
    });
  });
}
