// @wordpress/components/build/custom-gradient-picker/custom-gradient-bar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CustomGradientBar;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _lodash = require("lodash");

var _classnames = _interopRequireDefault(require("classnames"));

var _icons = require("@wordpress/icons");

var _button = _interopRequireDefault(require("../button"));

var _colorPicker = _interopRequireDefault(require("../color-picker"));

var _dropdown = _interopRequireDefault(require("../dropdown"));

var _controlPoints = _interopRequireDefault(require("./control-points"));

var _constants = require("./constants");

var _serializer = require("./serializer");

var _utils = require("./utils");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function InsertPoint(_ref) {
  var onChange = _ref.onChange,
      gradientAST = _ref.gradientAST,
      onOpenInserter = _ref.onOpenInserter,
      onCloseInserter = _ref.onCloseInserter,
      insertPosition = _ref.insertPosition;

  var _useState = (0, _element.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      alreadyInsertedPoint = _useState2[0],
      setAlreadyInsertedPoint = _useState2[1];

  return (0, _element.createElement)(_dropdown.default, {
    className: "components-custom-gradient-picker__inserter",
    onClose: function onClose() {
      onCloseInserter();
    },
    renderToggle: function renderToggle(_ref2) {
      var isOpen = _ref2.isOpen,
          onToggle = _ref2.onToggle;
      return (0, _element.createElement)(_button.default, {
        "aria-expanded": isOpen,
        "aria-haspopup": "true",
        onClick: function onClick() {
          if (isOpen) {
            onCloseInserter();
          } else {
            setAlreadyInsertedPoint(false);
            onOpenInserter();
          }

          onToggle();
        },
        className: "components-custom-gradient-picker__insert-point",
        icon: _icons.plus,
        style: {
          left: insertPosition !== null ? "".concat(insertPosition, "%") : undefined
        }
      });
    },
    renderContent: function renderContent() {
      return (0, _element.createElement)(_colorPicker.default, {
        onChangeComplete: function onChangeComplete(_ref3) {
          var rgb = _ref3.rgb;
          var newGradient;

          if (alreadyInsertedPoint) {
            newGradient = (0, _utils.getGradientWithColorAtPositionChanged)(gradientAST, insertPosition, rgb);
          } else {
            newGradient = (0, _utils.getGradientWithColorStopAdded)(gradientAST, insertPosition, rgb);
            setAlreadyInsertedPoint(true);
          }

          onChange(newGradient);
        }
      });
    },
    popoverProps: _constants.COLOR_POPOVER_PROPS
  });
}

function customGradientBarReducer(state, action) {
  switch (action.type) {
    case 'MOVE_INSERTER':
      if (state.id === 'IDLE' || state.id === 'MOVING_INSERTER') {
        return {
          id: 'MOVING_INSERTER',
          insertPosition: action.insertPosition
        };
      }

      break;

    case 'STOP_INSERTER_MOVE':
      if (state.id === 'MOVING_INSERTER') {
        return {
          id: 'IDLE'
        };
      }

      break;

    case 'OPEN_INSERTER':
      if (state.id === 'MOVING_INSERTER') {
        return {
          id: 'INSERTING_CONTROL_POINT',
          insertPosition: state.insertPosition
        };
      }

      break;

    case 'CLOSE_INSERTER':
      if (state.id === 'INSERTING_CONTROL_POINT') {
        return {
          id: 'IDLE'
        };
      }

      break;

    case 'START_CONTROL_CHANGE':
      if (state.id === 'IDLE') {
        return {
          id: 'MOVING_CONTROL_POINT'
        };
      }

      break;

    case 'STOP_CONTROL_CHANGE':
      if (state.id === 'MOVING_CONTROL_POINT') {
        return {
          id: 'IDLE'
        };
      }

      break;
  }

  return state;
}

var customGradientBarReducerInitialState = {
  id: 'IDLE'
};

function CustomGradientBar(_ref4) {
  var value = _ref4.value,
      onChange = _ref4.onChange;

  var _getGradientParsed = (0, _utils.getGradientParsed)(value),
      gradientAST = _getGradientParsed.gradientAST,
      gradientValue = _getGradientParsed.gradientValue,
      hasGradient = _getGradientParsed.hasGradient;

  var onGradientStructureChange = function onGradientStructureChange(newGradientStructure) {
    onChange((0, _serializer.serializeGradient)(newGradientStructure));
  };

  var gradientPickerDomRef = (0, _element.useRef)();
  var markerPoints = (0, _utils.getMarkerPoints)(gradientAST);

  var _useReducer = (0, _element.useReducer)(customGradientBarReducer, customGradientBarReducerInitialState),
      _useReducer2 = (0, _slicedToArray2.default)(_useReducer, 2),
      gradientBarState = _useReducer2[0],
      gradientBarStateDispatch = _useReducer2[1];

  var onMouseEnterAndMove = function onMouseEnterAndMove(event) {
    var insertPosition = (0, _utils.getHorizontalRelativeGradientPosition)(event.clientX, gradientPickerDomRef.current, _constants.INSERT_POINT_WIDTH); // If the insert point is close to an existing control point don't show it.

    if ((0, _lodash.some)(markerPoints, function (_ref5) {
      var positionValue = _ref5.positionValue;
      return Math.abs(insertPosition - positionValue) < _constants.MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT;
    })) {
      if (gradientBarState.id === 'MOVING_INSERTER') {
        gradientBarStateDispatch({
          type: 'STOP_INSERTER_MOVE'
        });
      }

      return;
    }

    gradientBarStateDispatch({
      type: 'MOVE_INSERTER',
      insertPosition: insertPosition
    });
  };

  var onMouseLeave = function onMouseLeave() {
    gradientBarStateDispatch({
      type: 'STOP_INSERTER_MOVE'
    });
  };

  var isMovingInserter = gradientBarState.id === 'MOVING_INSERTER';
  var isInsertingControlPoint = gradientBarState.id === 'INSERTING_CONTROL_POINT';
  return (0, _element.createElement)("div", {
    ref: gradientPickerDomRef,
    className: (0, _classnames.default)('components-custom-gradient-picker__gradient-bar', {
      'has-gradient': hasGradient
    }),
    onMouseEnter: onMouseEnterAndMove,
    onMouseMove: onMouseEnterAndMove // On radial gradients the bar should display a linear gradient.
    // On radial gradients the bar represents a slice of the gradient from the center until the outside.
    ,
    style: {
      background: gradientAST.type === 'radial-gradient' ? (0, _utils.getLinearGradientRepresentationOfARadial)(gradientAST) : gradientValue
    },
    onMouseLeave: onMouseLeave
  }, (0, _element.createElement)("div", {
    className: "components-custom-gradient-picker__markers-container"
  }, (isMovingInserter || isInsertingControlPoint) && (0, _element.createElement)(InsertPoint, {
    insertPosition: gradientBarState.insertPosition,
    onChange: onGradientStructureChange,
    gradientAST: gradientAST,
    onOpenInserter: function onOpenInserter() {
      gradientBarStateDispatch({
        type: 'OPEN_INSERTER'
      });
    },
    onCloseInserter: function onCloseInserter() {
      gradientBarStateDispatch({
        type: 'CLOSE_INSERTER'
      });
    }
  }), (0, _element.createElement)(_controlPoints.default, {
    gradientPickerDomRef: gradientPickerDomRef,
    ignoreMarkerPosition: isInsertingControlPoint ? gradientBarState.insertPosition : undefined,
    markerPoints: markerPoints,
    onChange: onGradientStructureChange,
    gradientAST: gradientAST,
    onStartControlPointChange: function onStartControlPointChange() {
      gradientBarStateDispatch({
        type: 'START_CONTROL_CHANGE'
      });
    },
    onStopControlPointChange: function onStopControlPointChange() {
      gradientBarStateDispatch({
        type: 'STOP_CONTROL_CHANGE'
      });
    }
  })));
}
