// @wordpress/components/build/custom-gradient-picker/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _lodash = require("lodash");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _utils = require("../mobile/color-settings/utils");

var _utils2 = require("./utils");

var _serializer = require("./serializer");

var _constants = require("./constants");

var _style = _interopRequireDefault(require("./style.scss"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function CustomGradientPicker(_ref) {
  var setColor = _ref.setColor,
      currentValue = _ref.currentValue,
      isGradientColor = _ref.isGradientColor;

  var _useState = (0, _element.useState)(_constants.HORIZONTAL_GRADIENT_ORIENTATION),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      gradientOrientation = _useState2[0],
      setGradientOrientation = _useState2[1];

  var _useState3 = (0, _element.useState)(currentValue),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      currentColor = _useState4[0],
      setCurrentColor = _useState4[1];

  var getGradientType = _utils.colorsUtils.getGradientType,
      gradients = _utils.colorsUtils.gradients,
      gradientOptions = _utils.colorsUtils.gradientOptions;

  var _getGradientParsed = (0, _utils2.getGradientParsed)(currentColor),
      gradientAST = _getGradientParsed.gradientAST;

  var gradientType = getGradientType(currentColor);

  function isLinearGradient(type) {
    return type === gradients.linear;
  }

  function getGradientColor(type) {
    var orientation = (0, _lodash.get)(gradientAST, ['orientation']);

    if (orientation) {
      setGradientOrientation(orientation);
    }

    return (0, _serializer.serializeGradient)(isLinearGradient(type) ? _objectSpread(_objectSpread(_objectSpread({}, gradientAST), gradientAST.orientation ? {} : {
      orientation: gradientOrientation
    }), {}, {
      type: type
    }) : _objectSpread(_objectSpread({}, (0, _lodash.omit)(gradientAST, ['orientation'])), {}, {
      type: type
    }));
  }

  function onGradientTypeChange(type) {
    var gradientColor = getGradientColor(type);
    setCurrentColor(gradientColor);
    setColor(gradientColor);
  }

  function setGradientAngle(value) {
    var gradientColor = (0, _serializer.serializeGradient)(_objectSpread(_objectSpread({}, gradientAST), {}, {
      orientation: {
        type: 'angular',
        value: value
      }
    }));

    if (isGradientColor && gradientColor !== currentColor) {
      setCurrentColor(gradientColor);
      setColor(gradientColor);
    }
  }

  function getGradientAngle() {
    return (0, _lodash.get)(gradientAST, ['orientation', 'value'], _constants.DEFAULT_LINEAR_GRADIENT_ANGLE);
  }

  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_components.PanelBody, {
    title: (0, _i18n.__)('Gradient Type')
  }, (0, _element.createElement)(_components.RadioControl, {
    selected: gradientType,
    options: gradientOptions,
    onChange: onGradientTypeChange
  })), isLinearGradient(gradientType) && (0, _element.createElement)(_components.PanelBody, {
    style: _style.default.angleControl
  }, (0, _element.createElement)(_components.RangeControl, {
    label: (0, _i18n.__)('Angle'),
    minimumValue: 0,
    maximumValue: 360,
    value: getGradientAngle(),
    onChange: setGradientAngle
  })));
}

var _default = CustomGradientPicker;
exports.default = _default;
