// @wordpress/components/build/custom-gradient-picker/serializer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.serializeGradientColor = serializeGradientColor;
exports.serializeGradientPosition = serializeGradientPosition;
exports.serializeGradientColorStop = serializeGradientColorStop;
exports.serializeGradientOrientation = serializeGradientOrientation;
exports.serializeGradient = serializeGradient;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _lodash = require("lodash");

/**
 * External dependencies
 */
function serializeGradientColor(_ref) {
  var type = _ref.type,
      value = _ref.value;

  if (type === 'literal') {
    return value;
  }

  if (type === 'hex') {
    return "#".concat(value);
  }

  return "".concat(type, "(").concat(value.join(','), ")");
}

function serializeGradientPosition(_ref2) {
  var type = _ref2.type,
      value = _ref2.value;
  return "".concat(value).concat(type);
}

function serializeGradientColorStop(_ref3) {
  var type = _ref3.type,
      value = _ref3.value,
      length = _ref3.length;
  return "".concat(serializeGradientColor({
    type: type,
    value: value
  }), " ").concat(serializeGradientPosition(length));
}

function serializeGradientOrientation(orientation) {
  if (!orientation || orientation.type !== 'angular') {
    return;
  }

  return "".concat(orientation.value, "deg");
}

function serializeGradient(_ref4) {
  var type = _ref4.type,
      orientation = _ref4.orientation,
      colorStops = _ref4.colorStops;
  var serializedOrientation = serializeGradientOrientation(orientation);
  var serializedColorStops = colorStops.sort(function (colorStop1, colorStop2) {
    return (0, _lodash.get)(colorStop1, ['length', 'value'], 0) - (0, _lodash.get)(colorStop2, ['length', 'value'], 0);
  }).map(serializeGradientColorStop);
  return "".concat(type, "(").concat((0, _lodash.compact)([serializedOrientation].concat((0, _toConsumableArray2.default)(serializedColorStops))).join(','), ")");
}
