// @wordpress/components/build/mobile/bottom-sheet/cell.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _reactNative = require("react-native");

var _lodash = require("lodash");

var _components = require("@wordpress/components");

var _icons = require("@wordpress/icons");

var _i18n = require("@wordpress/i18n");

var _compose = require("@wordpress/compose");

var _styles = _interopRequireDefault(require("./styles.scss"));

var _cellStyles = _interopRequireDefault(require("./cellStyles.scss"));

var _rippleNative = _interopRequireDefault(require("./ripple.native.js"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var BottomSheetCell = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(BottomSheetCell, _Component);

  var _super = _createSuper(BottomSheetCell);

  function BottomSheetCell(props) {
    var _this;

    (0, _classCallCheck2.default)(this, BottomSheetCell);
    _this = _super.apply(this, arguments);
    _this.state = {
      isEditingValue: props.autoFocus || false,
      isScreenReaderEnabled: false
    };
    _this.handleScreenReaderToggled = _this.handleScreenReaderToggled.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(BottomSheetCell, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps, prevState) {
      if (!prevState.isEditingValue && this.state.isEditingValue) {
        this._valueTextInput.focus();
      }
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      _reactNative.AccessibilityInfo.addEventListener('screenReaderChanged', this.handleScreenReaderToggled);

      _reactNative.AccessibilityInfo.isScreenReaderEnabled().then(function (isScreenReaderEnabled) {
        _this2.setState({
          isScreenReaderEnabled: isScreenReaderEnabled
        });
      });
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      _reactNative.AccessibilityInfo.removeEventListener('screenReaderChanged', this.handleScreenReaderToggled);
    }
  }, {
    key: "handleScreenReaderToggled",
    value: function handleScreenReaderToggled(isScreenReaderEnabled) {
      this.setState({
        isScreenReaderEnabled: isScreenReaderEnabled
      });
    }
  }, {
    key: "typeToKeyboardType",
    value: function typeToKeyboardType(type, step) {
      var keyboardType = "default";

      if (type === "number") {
        if (step && Math.abs(step) < 1) {
          keyboardType = "decimal-pad";
        } else {
          keyboardType = "number-pad";
        }
      }

      return keyboardType;
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props = this.props,
          accessible = _this$props.accessible,
          accessibilityLabel = _this$props.accessibilityLabel,
          accessibilityHint = _this$props.accessibilityHint,
          accessibilityRole = _this$props.accessibilityRole,
          _this$props$disabled = _this$props.disabled,
          disabled = _this$props$disabled === void 0 ? false : _this$props$disabled,
          activeOpacity = _this$props.activeOpacity,
          onPress = _this$props.onPress,
          label = _this$props.label,
          value = _this$props.value,
          _this$props$valuePlac = _this$props.valuePlaceholder,
          valuePlaceholder = _this$props$valuePlac === void 0 ? '' : _this$props$valuePlac,
          icon = _this$props.icon,
          leftAlign = _this$props.leftAlign,
          _this$props$labelStyl = _this$props.labelStyle,
          labelStyle = _this$props$labelStyl === void 0 ? {} : _this$props$labelStyl,
          _this$props$valueStyl = _this$props.valueStyle,
          valueStyle = _this$props$valueStyl === void 0 ? {} : _this$props$valueStyl,
          _this$props$cellConta = _this$props.cellContainerStyle,
          cellContainerStyle = _this$props$cellConta === void 0 ? {} : _this$props$cellConta,
          _this$props$cellRowCo = _this$props.cellRowContainerStyle,
          cellRowContainerStyle = _this$props$cellRowCo === void 0 ? {} : _this$props$cellRowCo,
          onChangeValue = _this$props.onChangeValue,
          onSubmit = _this$props.onSubmit,
          children = _this$props.children,
          _this$props$editable = _this$props.editable,
          editable = _this$props$editable === void 0 ? true : _this$props$editable,
          _this$props$isSelecte = _this$props.isSelected,
          isSelected = _this$props$isSelecte === void 0 ? false : _this$props$isSelecte,
          separatorType = _this$props.separatorType,
          _this$props$style = _this$props.style,
          style = _this$props$style === void 0 ? {} : _this$props$style,
          getStylesFromColorScheme = _this$props.getStylesFromColorScheme,
          customActionButton = _this$props.customActionButton,
          type = _this$props.type,
          step = _this$props.step,
          borderless = _this$props.borderless,
          valueProps = (0, _objectWithoutProperties2.default)(_this$props, ["accessible", "accessibilityLabel", "accessibilityHint", "accessibilityRole", "disabled", "activeOpacity", "onPress", "label", "value", "valuePlaceholder", "icon", "leftAlign", "labelStyle", "valueStyle", "cellContainerStyle", "cellRowContainerStyle", "onChangeValue", "onSubmit", "children", "editable", "isSelected", "separatorType", "style", "getStylesFromColorScheme", "customActionButton", "type", "step", "borderless"]);
      var showValue = value !== undefined;
      var isValueEditable = editable && onChangeValue !== undefined;
      var cellLabelStyle = getStylesFromColorScheme(_styles.default.cellLabel, _styles.default.cellTextDark);
      var cellLabelCenteredStyle = getStylesFromColorScheme(_styles.default.cellLabelCentered, _styles.default.cellTextDark);
      var cellLabelLeftAlignNoIconStyle = getStylesFromColorScheme(_styles.default.cellLabelLeftAlignNoIcon, _styles.default.cellTextDark);
      var defaultMissingIconAndValue = leftAlign ? cellLabelLeftAlignNoIconStyle : cellLabelCenteredStyle;
      var defaultLabelStyle = showValue || customActionButton || icon ? cellLabelStyle : defaultMissingIconAndValue;
      var drawSeparator = separatorType && separatorType !== 'none' || separatorStyle === undefined;
      var drawTopSeparator = drawSeparator && separatorType === 'topFullWidth';
      var cellContainerStyles = [_styles.default.cellContainer, cellContainerStyle];
      var rowContainerStyles = [_styles.default.cellRowContainer, cellRowContainerStyle];

      var onCellPress = function onCellPress() {
        if (isValueEditable) {
          startEditing();
        } else if (onPress !== undefined) {
          onPress();
        }
      };

      var finishEditing = function finishEditing() {
        _this3.setState({
          isEditingValue: false
        });
      };

      var startEditing = function startEditing() {
        if (_this3.state.isEditingValue === false) {
          _this3.setState({
            isEditingValue: true
          });
        }
      };

      var separatorStyle = function separatorStyle() {
        //eslint-disable-next-line @wordpress/no-unused-vars-before-return
        var defaultSeparatorStyle = _this3.props.getStylesFromColorScheme(_styles.default.separator, _styles.default.separatorDark);

        var cellSeparatorStyle = _this3.props.getStylesFromColorScheme(_styles.default.cellSeparator, _styles.default.cellSeparatorDark);

        var leftMarginStyle = _objectSpread(_objectSpread({}, cellSeparatorStyle), _cellStyles.default.separatorMarginLeft);

        switch (separatorType) {
          case 'leftMargin':
            return leftMarginStyle;

          case 'fullWidth':
          case 'topFullWidth':
            return defaultSeparatorStyle;

          case 'none':
            return undefined;

          case undefined:
            if (showValue && icon) {
              return leftMarginStyle;
            }

            return defaultSeparatorStyle;
        }
      };

      var getValueComponent = function getValueComponent() {
        var styleRTL = _reactNative.I18nManager.isRTL && _styles.default.cellValueRTL;

        var cellValueStyle = _this3.props.getStylesFromColorScheme(_styles.default.cellValue, _styles.default.cellTextDark);

        var finalStyle = _objectSpread(_objectSpread(_objectSpread({}, cellValueStyle), valueStyle), styleRTL); // To be able to show the `middle` ellipsizeMode on editable cells
        // we show the TextInput just when the user wants to edit the value,
        // and the Text component to display it.
        // We also show the TextInput to display placeholder.


        var shouldShowPlaceholder = isValueEditable && value === '';
        return _this3.state.isEditingValue || shouldShowPlaceholder ? (0, _element.createElement)(_reactNative.TextInput, (0, _extends2.default)({
          ref: function ref(c) {
            return _this3._valueTextInput = c;
          },
          numberOfLines: 1,
          style: finalStyle,
          value: value,
          placeholder: valuePlaceholder,
          placeholderTextColor: '#87a6bc',
          onChangeText: onChangeValue,
          editable: isValueEditable,
          pointerEvents: _this3.state.isEditingValue ? 'auto' : 'none',
          onFocus: startEditing,
          onBlur: finishEditing,
          onSubmitEditing: onSubmit,
          keyboardType: _this3.typeToKeyboardType(type, step)
        }, valueProps)) : (0, _element.createElement)(_reactNative.Text, {
          style: _objectSpread(_objectSpread({}, cellValueStyle), valueStyle),
          numberOfLines: 1,
          ellipsizeMode: 'middle'
        }, value);
      };

      var getAccessibilityLabel = function getAccessibilityLabel() {
        if (accessibilityLabel || !showValue) {
          return accessibilityLabel || label;
        }

        return (0, _lodash.isEmpty)(value) ? (0, _i18n.sprintf)(
        /* translators: accessibility text. Empty state of a inline textinput cell. %s: The cell's title */
        (0, _i18n._x)('%s. Empty', 'inline textinput cell'), label) : // Separating by ',' is necessary to make a pause on urls (non-capitalized text)
        (0, _i18n.sprintf)(
        /* translators: accessibility text. Inline textinput title and value.%1: Cell title, %2: cell value. */
        (0, _i18n._x)('%1$s, %2$s', 'inline textinput cell'), label, value);
      };

      var iconStyle = getStylesFromColorScheme(_styles.default.icon, _styles.default.iconDark);
      var resetButtonStyle = getStylesFromColorScheme(_styles.default.resetButton, _styles.default.resetButtonDark);
      var containerPointerEvents = this.state.isScreenReaderEnabled && accessible ? 'none' : 'auto';

      var _ref = customActionButton || {},
          title = _ref.title,
          handler = _ref.handler;

      var opacity = activeOpacity !== undefined ? activeOpacity : (0, _lodash.get)(_cellStyles.default, 'activeOpacity.opacity');
      return (0, _element.createElement)(_rippleNative.default, {
        accessible: accessible !== undefined ? accessible : !this.state.isEditingValue,
        accessibilityLabel: getAccessibilityLabel(),
        accessibilityRole: accessibilityRole || 'button',
        accessibilityHint: isValueEditable ?
        /* translators: accessibility text */
        (0, _i18n.__)('Double tap to edit this value') : accessibilityHint,
        disabled: disabled,
        activeOpacity: opacity,
        onPress: onCellPress,
        style: [_styles.default.clipToBounds, style],
        borderless: borderless
      }, drawTopSeparator && (0, _element.createElement)(_reactNative.View, {
        style: separatorStyle()
      }), (0, _element.createElement)(_reactNative.View, {
        style: cellContainerStyles,
        pointerEvents: containerPointerEvents
      }, (0, _element.createElement)(_reactNative.View, {
        style: rowContainerStyles
      }, (0, _element.createElement)(_reactNative.View, {
        style: _styles.default.cellRowContainer
      }, icon && (0, _element.createElement)(_reactNative.View, {
        style: _styles.default.cellRowContainer
      }, (0, _element.createElement)(_components.Icon, {
        icon: icon,
        size: 24,
        fill: iconStyle.color,
        isPressed: false
      }), (0, _element.createElement)(_reactNative.View, {
        style: _cellStyles.default.labelIconSeparator
      })), label && (0, _element.createElement)(_reactNative.Text, {
        style: [defaultLabelStyle, labelStyle]
      }, label)), customActionButton && (0, _element.createElement)(_reactNative.TouchableOpacity, {
        onPress: handler,
        accessibilityRole: 'button'
      }, (0, _element.createElement)(_reactNative.Text, {
        style: resetButtonStyle
      }, title))), isSelected && (0, _element.createElement)(_components.Icon, {
        icon: _icons.check,
        fill: _cellStyles.default.isSelected.color
      }), showValue && getValueComponent(), children), !drawTopSeparator && (0, _element.createElement)(_reactNative.View, {
        style: separatorStyle()
      }));
    }
  }]);
  return BottomSheetCell;
}(_element.Component);

var _default = (0, _compose.withPreferredColorScheme)(BottomSheetCell);

exports.default = _default;
