// @wordpress/components/build-module/custom-gradient-picker/control-points.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import { createElement, Fragment } from "@wordpress/element";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

/**
 * External dependencies
 */
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { Component, useEffect, useRef } from '@wordpress/element';
import { __, sprintf } from '@wordpress/i18n';
import { useInstanceId } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import Button from '../button';
import ColorPicker from '../color-picker';
import Dropdown from '../dropdown';
import VisuallyHidden from '../visually-hidden';
import { getGradientWithColorAtIndexChanged, getGradientWithControlPointRemoved, getGradientWithPositionAtIndexChanged, getGradientWithPositionAtIndexDecreased, getGradientWithPositionAtIndexIncreased, getHorizontalRelativeGradientPosition, isControlPointOverlapping } from './utils';
import { COLOR_POPOVER_PROPS, GRADIENT_MARKERS_WIDTH, MINIMUM_SIGNIFICANT_MOVE } from './constants';
import KeyboardShortcuts from '../keyboard-shortcuts';

var ControlPointKeyboardMove = /*#__PURE__*/function (_Component) {
  _inherits(ControlPointKeyboardMove, _Component);

  var _super = _createSuper(ControlPointKeyboardMove);

  function ControlPointKeyboardMove() {
    var _this;

    _classCallCheck(this, ControlPointKeyboardMove);

    _this = _super.apply(this, arguments);
    _this.increase = _this.increase.bind(_assertThisInitialized(_this));
    _this.decrease = _this.decrease.bind(_assertThisInitialized(_this));
    _this.shortcuts = {
      right: _this.increase,
      left: _this.decrease
    };
    return _this;
  }

  _createClass(ControlPointKeyboardMove, [{
    key: "increase",
    value: function increase(event) {
      // Stop propagation of the key press event to avoid focus moving
      // to another editor area.
      event.stopPropagation();
      var _this$props = this.props,
          gradientIndex = _this$props.gradientIndex,
          onChange = _this$props.onChange,
          gradientAST = _this$props.gradientAST;
      onChange(getGradientWithPositionAtIndexIncreased(gradientAST, gradientIndex));
    }
  }, {
    key: "decrease",
    value: function decrease(event) {
      // Stop propagation of the key press event to avoid focus moving
      // to another editor area.
      event.stopPropagation();
      var _this$props2 = this.props,
          gradientIndex = _this$props2.gradientIndex,
          onChange = _this$props2.onChange,
          gradientAST = _this$props2.gradientAST;
      onChange(getGradientWithPositionAtIndexDecreased(gradientAST, gradientIndex));
    }
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children;
      return createElement(KeyboardShortcuts, {
        shortcuts: this.shortcuts
      }, children);
    }
  }]);

  return ControlPointKeyboardMove;
}(Component);

function ControlPointButton(_ref) {
  var isOpen = _ref.isOpen,
      position = _ref.position,
      color = _ref.color,
      onChange = _ref.onChange,
      gradientIndex = _ref.gradientIndex,
      gradientAST = _ref.gradientAST,
      additionalProps = _objectWithoutProperties(_ref, ["isOpen", "position", "color", "onChange", "gradientIndex", "gradientAST"]);

  var instanceId = useInstanceId(ControlPointButton);
  var descriptionId = "components-custom-gradient-picker__control-point-button-description-".concat(instanceId);
  return createElement(ControlPointKeyboardMove, {
    onChange: onChange,
    gradientIndex: gradientIndex,
    gradientAST: gradientAST
  }, createElement(Button, _extends({
    "aria-label": sprintf( // translators: %1$s: gradient position e.g: 70%, %2$s: gradient color code e.g: rgb(52,121,151).
    __('Gradient control point at position %1$s with color code %2$s.'), position, color),
    "aria-describedby": descriptionId,
    "aria-haspopup": "true",
    "aria-expanded": isOpen,
    className: classnames('components-custom-gradient-picker__control-point-button', {
      'is-active': isOpen
    }),
    style: {
      left: position
    }
  }, additionalProps)), createElement(VisuallyHidden, {
    id: descriptionId
  }, __('Use your left or right arrow keys or drag and drop with the mouse to change the gradient position. Press the button to change the color or remove the control point.')));
}

export default function ControlPoints(_ref2) {
  var gradientPickerDomRef = _ref2.gradientPickerDomRef,
      ignoreMarkerPosition = _ref2.ignoreMarkerPosition,
      markerPoints = _ref2.markerPoints,
      onChange = _ref2.onChange,
      gradientAST = _ref2.gradientAST,
      onStartControlPointChange = _ref2.onStartControlPointChange,
      onStopControlPointChange = _ref2.onStopControlPointChange;
  var controlPointMoveState = useRef();

  var onMouseMove = function onMouseMove(event) {
    var relativePosition = getHorizontalRelativeGradientPosition(event.clientX, gradientPickerDomRef.current, GRADIENT_MARKERS_WIDTH);
    var _controlPointMoveStat = controlPointMoveState.current,
        referenceGradientAST = _controlPointMoveStat.gradientAST,
        position = _controlPointMoveStat.position,
        significantMoveHappened = _controlPointMoveStat.significantMoveHappened;

    if (!significantMoveHappened) {
      var initialPosition = referenceGradientAST.colorStops[position].length.value;

      if (Math.abs(initialPosition - relativePosition) >= MINIMUM_SIGNIFICANT_MOVE) {
        controlPointMoveState.current.significantMoveHappened = true;
      }
    }

    if (!isControlPointOverlapping(referenceGradientAST, relativePosition, position)) {
      onChange(getGradientWithPositionAtIndexChanged(referenceGradientAST, position, relativePosition));
    }
  };

  var cleanEventListeners = function cleanEventListeners() {
    if (window && window.removeEventListener && controlPointMoveState.current && controlPointMoveState.current.listenersActivated) {
      window.removeEventListener('mousemove', onMouseMove);
      window.removeEventListener('mouseup', cleanEventListeners);
      onStopControlPointChange();
      controlPointMoveState.current.listenersActivated = false;
    }
  };

  useEffect(function () {
    return function () {
      cleanEventListeners();
    };
  }, []);
  return markerPoints.map(function (point, index) {
    return point && ignoreMarkerPosition !== point.positionValue && createElement(Dropdown, {
      key: index,
      onClose: onStopControlPointChange,
      renderToggle: function renderToggle(_ref3) {
        var isOpen = _ref3.isOpen,
            onToggle = _ref3.onToggle;
        return createElement(ControlPointButton, {
          key: index,
          onClick: function onClick() {
            if (controlPointMoveState.current && controlPointMoveState.current.significantMoveHappened) {
              return;
            }

            if (isOpen) {
              onStopControlPointChange();
            } else {
              onStartControlPointChange();
            }

            onToggle();
          },
          onMouseDown: function onMouseDown() {
            if (window && window.addEventListener) {
              controlPointMoveState.current = {
                gradientAST: gradientAST,
                position: index,
                significantMoveHappened: false,
                listenersActivated: true
              };
              onStartControlPointChange();
              window.addEventListener('mousemove', onMouseMove);
              window.addEventListener('mouseup', cleanEventListeners);
            }
          },
          isOpen: isOpen,
          position: point.position,
          color: point.color,
          onChange: onChange,
          gradientAST: gradientAST,
          gradientIndex: index
        });
      },
      renderContent: function renderContent(_ref4) {
        var onClose = _ref4.onClose;
        return createElement(Fragment, null, createElement(ColorPicker, {
          color: point.color,
          onChangeComplete: function onChangeComplete(_ref5) {
            var rgb = _ref5.rgb;
            onChange(getGradientWithColorAtIndexChanged(gradientAST, index, rgb));
          }
        }), createElement(Button, {
          className: "components-custom-gradient-picker__remove-control-point",
          onClick: function onClick() {
            onChange(getGradientWithControlPointRemoved(gradientAST, index));
            onClose();
          },
          isLink: true
        }, __('Remove Control Point')));
      },
      popoverProps: COLOR_POPOVER_PROPS
    });
  });
}
