// @wordpress/components/build-module/custom-gradient-picker/custom-gradient-bar.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { some } from 'lodash';
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { useRef, useReducer, useState } from '@wordpress/element';
import { plus } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import Button from '../button';
import ColorPicker from '../color-picker';
import Dropdown from '../dropdown';
import ControlPoints from './control-points';
import { INSERT_POINT_WIDTH, COLOR_POPOVER_PROPS, MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT } from './constants';
import { serializeGradient } from './serializer';
import { getGradientWithColorAtPositionChanged, getGradientWithColorStopAdded, getHorizontalRelativeGradientPosition, getMarkerPoints, getGradientParsed, getLinearGradientRepresentationOfARadial } from './utils';

function InsertPoint(_ref) {
  var onChange = _ref.onChange,
      gradientAST = _ref.gradientAST,
      onOpenInserter = _ref.onOpenInserter,
      onCloseInserter = _ref.onCloseInserter,
      insertPosition = _ref.insertPosition;

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      alreadyInsertedPoint = _useState2[0],
      setAlreadyInsertedPoint = _useState2[1];

  return createElement(Dropdown, {
    className: "components-custom-gradient-picker__inserter",
    onClose: function onClose() {
      onCloseInserter();
    },
    renderToggle: function renderToggle(_ref2) {
      var isOpen = _ref2.isOpen,
          onToggle = _ref2.onToggle;
      return createElement(Button, {
        "aria-expanded": isOpen,
        "aria-haspopup": "true",
        onClick: function onClick() {
          if (isOpen) {
            onCloseInserter();
          } else {
            setAlreadyInsertedPoint(false);
            onOpenInserter();
          }

          onToggle();
        },
        className: "components-custom-gradient-picker__insert-point",
        icon: plus,
        style: {
          left: insertPosition !== null ? "".concat(insertPosition, "%") : undefined
        }
      });
    },
    renderContent: function renderContent() {
      return createElement(ColorPicker, {
        onChangeComplete: function onChangeComplete(_ref3) {
          var rgb = _ref3.rgb;
          var newGradient;

          if (alreadyInsertedPoint) {
            newGradient = getGradientWithColorAtPositionChanged(gradientAST, insertPosition, rgb);
          } else {
            newGradient = getGradientWithColorStopAdded(gradientAST, insertPosition, rgb);
            setAlreadyInsertedPoint(true);
          }

          onChange(newGradient);
        }
      });
    },
    popoverProps: COLOR_POPOVER_PROPS
  });
}

function customGradientBarReducer(state, action) {
  switch (action.type) {
    case 'MOVE_INSERTER':
      if (state.id === 'IDLE' || state.id === 'MOVING_INSERTER') {
        return {
          id: 'MOVING_INSERTER',
          insertPosition: action.insertPosition
        };
      }

      break;

    case 'STOP_INSERTER_MOVE':
      if (state.id === 'MOVING_INSERTER') {
        return {
          id: 'IDLE'
        };
      }

      break;

    case 'OPEN_INSERTER':
      if (state.id === 'MOVING_INSERTER') {
        return {
          id: 'INSERTING_CONTROL_POINT',
          insertPosition: state.insertPosition
        };
      }

      break;

    case 'CLOSE_INSERTER':
      if (state.id === 'INSERTING_CONTROL_POINT') {
        return {
          id: 'IDLE'
        };
      }

      break;

    case 'START_CONTROL_CHANGE':
      if (state.id === 'IDLE') {
        return {
          id: 'MOVING_CONTROL_POINT'
        };
      }

      break;

    case 'STOP_CONTROL_CHANGE':
      if (state.id === 'MOVING_CONTROL_POINT') {
        return {
          id: 'IDLE'
        };
      }

      break;
  }

  return state;
}

var customGradientBarReducerInitialState = {
  id: 'IDLE'
};
export default function CustomGradientBar(_ref4) {
  var value = _ref4.value,
      onChange = _ref4.onChange;

  var _getGradientParsed = getGradientParsed(value),
      gradientAST = _getGradientParsed.gradientAST,
      gradientValue = _getGradientParsed.gradientValue,
      hasGradient = _getGradientParsed.hasGradient;

  var onGradientStructureChange = function onGradientStructureChange(newGradientStructure) {
    onChange(serializeGradient(newGradientStructure));
  };

  var gradientPickerDomRef = useRef();
  var markerPoints = getMarkerPoints(gradientAST);

  var _useReducer = useReducer(customGradientBarReducer, customGradientBarReducerInitialState),
      _useReducer2 = _slicedToArray(_useReducer, 2),
      gradientBarState = _useReducer2[0],
      gradientBarStateDispatch = _useReducer2[1];

  var onMouseEnterAndMove = function onMouseEnterAndMove(event) {
    var insertPosition = getHorizontalRelativeGradientPosition(event.clientX, gradientPickerDomRef.current, INSERT_POINT_WIDTH); // If the insert point is close to an existing control point don't show it.

    if (some(markerPoints, function (_ref5) {
      var positionValue = _ref5.positionValue;
      return Math.abs(insertPosition - positionValue) < MINIMUM_DISTANCE_BETWEEN_INSERTER_AND_POINT;
    })) {
      if (gradientBarState.id === 'MOVING_INSERTER') {
        gradientBarStateDispatch({
          type: 'STOP_INSERTER_MOVE'
        });
      }

      return;
    }

    gradientBarStateDispatch({
      type: 'MOVE_INSERTER',
      insertPosition: insertPosition
    });
  };

  var onMouseLeave = function onMouseLeave() {
    gradientBarStateDispatch({
      type: 'STOP_INSERTER_MOVE'
    });
  };

  var isMovingInserter = gradientBarState.id === 'MOVING_INSERTER';
  var isInsertingControlPoint = gradientBarState.id === 'INSERTING_CONTROL_POINT';
  return createElement("div", {
    ref: gradientPickerDomRef,
    className: classnames('components-custom-gradient-picker__gradient-bar', {
      'has-gradient': hasGradient
    }),
    onMouseEnter: onMouseEnterAndMove,
    onMouseMove: onMouseEnterAndMove // On radial gradients the bar should display a linear gradient.
    // On radial gradients the bar represents a slice of the gradient from the center until the outside.
    ,
    style: {
      background: gradientAST.type === 'radial-gradient' ? getLinearGradientRepresentationOfARadial(gradientAST) : gradientValue
    },
    onMouseLeave: onMouseLeave
  }, createElement("div", {
    className: "components-custom-gradient-picker__markers-container"
  }, (isMovingInserter || isInsertingControlPoint) && createElement(InsertPoint, {
    insertPosition: gradientBarState.insertPosition,
    onChange: onGradientStructureChange,
    gradientAST: gradientAST,
    onOpenInserter: function onOpenInserter() {
      gradientBarStateDispatch({
        type: 'OPEN_INSERTER'
      });
    },
    onCloseInserter: function onCloseInserter() {
      gradientBarStateDispatch({
        type: 'CLOSE_INSERTER'
      });
    }
  }), createElement(ControlPoints, {
    gradientPickerDomRef: gradientPickerDomRef,
    ignoreMarkerPosition: isInsertingControlPoint ? gradientBarState.insertPosition : undefined,
    markerPoints: markerPoints,
    onChange: onGradientStructureChange,
    gradientAST: gradientAST,
    onStartControlPointChange: function onStartControlPointChange() {
      gradientBarStateDispatch({
        type: 'START_CONTROL_CHANGE'
      });
    },
    onStopControlPointChange: function onStopControlPointChange() {
      gradientBarStateDispatch({
        type: 'STOP_CONTROL_CHANGE'
      });
    }
  })));
}
