// @wordpress/components/build-module/custom-gradient-picker/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import { get, omit } from 'lodash';
/**
 * WordPress dependencies
 */

import { __ } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import AnglePickerControl from '../angle-picker-control';
import CustomGradientBar from './custom-gradient-bar';
import { Flex, FlexItem } from '../flex';
import SelectControl from '../select-control';
import { getGradientParsed } from './utils';
import { serializeGradient } from './serializer';
import { DEFAULT_LINEAR_GRADIENT_ANGLE, HORIZONTAL_GRADIENT_ORIENTATION, GRADIENT_OPTIONS } from './constants';
import { SelectWrapper } from './styles/custom-gradient-picker-styles';

var GradientAnglePicker = function GradientAnglePicker(_ref) {
  var gradientAST = _ref.gradientAST,
      hasGradient = _ref.hasGradient,
      onChange = _ref.onChange;
  var angle = get(gradientAST, ['orientation', 'value'], DEFAULT_LINEAR_GRADIENT_ANGLE);

  var onAngleChange = function onAngleChange(newAngle) {
    onChange(serializeGradient(_objectSpread(_objectSpread({}, gradientAST), {}, {
      orientation: {
        type: 'angular',
        value: newAngle
      }
    })));
  };

  return createElement(AnglePickerControl, {
    value: hasGradient ? angle : '',
    onChange: onAngleChange
  });
};

var GradientTypePicker = function GradientTypePicker(_ref2) {
  var gradientAST = _ref2.gradientAST,
      hasGradient = _ref2.hasGradient,
      onChange = _ref2.onChange;
  var type = gradientAST.type;

  var onSetLinearGradient = function onSetLinearGradient() {
    onChange(serializeGradient(_objectSpread(_objectSpread(_objectSpread({}, gradientAST), gradientAST.orientation ? {} : {
      orientation: HORIZONTAL_GRADIENT_ORIENTATION
    }), {}, {
      type: 'linear-gradient'
    })));
  };

  var onSetRadialGradient = function onSetRadialGradient() {
    onChange(serializeGradient(_objectSpread(_objectSpread({}, omit(gradientAST, ['orientation'])), {}, {
      type: 'radial-gradient'
    })));
  };

  var handleOnChange = function handleOnChange(next) {
    if (next === 'linear-gradient') {
      onSetLinearGradient();
    }

    if (next === 'radial-gradient') {
      onSetRadialGradient();
    }
  };

  return createElement(SelectControl, {
    className: "components-custom-gradient-picker__type-picker",
    label: __('Type'),
    onChange: handleOnChange,
    options: GRADIENT_OPTIONS,
    value: hasGradient && type
  });
};

export default function CustomGradientPicker(_ref3) {
  var value = _ref3.value,
      onChange = _ref3.onChange;

  var _getGradientParsed = getGradientParsed(value),
      gradientAST = _getGradientParsed.gradientAST,
      hasGradient = _getGradientParsed.hasGradient;

  var type = gradientAST.type;
  return createElement("div", {
    className: "components-custom-gradient-picker"
  }, createElement(CustomGradientBar, {
    value: value,
    onChange: onChange
  }), createElement(Flex, {
    gap: 3,
    className: "components-custom-gradient-picker__ui-line"
  }, createElement(SelectWrapper, null, createElement(GradientTypePicker, {
    gradientAST: gradientAST,
    hasGradient: hasGradient,
    onChange: onChange
  })), type === 'linear-gradient' && createElement(FlexItem, null, createElement(GradientAnglePicker, {
    gradientAST: gradientAST,
    hasGradient: hasGradient,
    onChange: onChange
  }))));
}
