// @wordpress/components/build-module/custom-gradient-picker/utils.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import { findIndex, map, some } from 'lodash';
import gradientParser from 'gradient-parser';
/**
 * Internal dependencies
 */

import { DEFAULT_GRADIENT, INSERT_POINT_WIDTH, MINIMUM_ABSOLUTE_LEFT_POSITION, MINIMUM_DISTANCE_BETWEEN_POINTS, KEYBOARD_CONTROL_POINT_VARIATION, HORIZONTAL_GRADIENT_ORIENTATION } from './constants';
import { serializeGradientColor, serializeGradientPosition, serializeGradient } from './serializer';

function tinyColorRgbToGradientColorStop(_ref) {
  var r = _ref.r,
      g = _ref.g,
      b = _ref.b,
      a = _ref.a;

  if (a === 1) {
    return {
      type: 'rgb',
      value: [r, g, b]
    };
  }

  return {
    type: 'rgba',
    value: [r, g, b, a]
  };
}

export function getGradientWithColorStopAdded(gradientAST, relativePosition, rgbaColor) {
  var colorStop = tinyColorRgbToGradientColorStop(rgbaColor);
  colorStop.length = {
    type: '%',
    value: relativePosition
  };
  return _objectSpread(_objectSpread({}, gradientAST), {}, {
    colorStops: [].concat(_toConsumableArray(gradientAST.colorStops), [colorStop])
  });
}
export function getGradientWithPositionAtIndexChanged(gradientAST, index, relativePosition) {
  return _objectSpread(_objectSpread({}, gradientAST), {}, {
    colorStops: gradientAST.colorStops.map(function (colorStop, colorStopIndex) {
      if (colorStopIndex !== index) {
        return colorStop;
      }

      return _objectSpread(_objectSpread({}, colorStop), {}, {
        length: _objectSpread(_objectSpread({}, colorStop.length), {}, {
          value: relativePosition
        })
      });
    })
  });
}
export function isControlPointOverlapping(gradientAST, position, initialIndex) {
  var initialPosition = parseInt(gradientAST.colorStops[initialIndex].length.value);
  var minPosition = Math.min(initialPosition, position);
  var maxPosition = Math.max(initialPosition, position);
  return some(gradientAST.colorStops, function (_ref2, index) {
    var length = _ref2.length;
    var itemPosition = parseInt(length.value);
    return index !== initialIndex && (Math.abs(itemPosition - position) < MINIMUM_DISTANCE_BETWEEN_POINTS || minPosition < itemPosition && itemPosition < maxPosition);
  });
}

function getGradientWithPositionAtIndexSummed(gradientAST, index, valueToSum) {
  var currentPosition = gradientAST.colorStops[index].length.value;
  var newPosition = Math.max(0, Math.min(100, parseInt(currentPosition) + valueToSum));

  if (isControlPointOverlapping(gradientAST, newPosition, index)) {
    return gradientAST;
  }

  return getGradientWithPositionAtIndexChanged(gradientAST, index, newPosition);
}

export function getGradientWithPositionAtIndexIncreased(gradientAST, index) {
  return getGradientWithPositionAtIndexSummed(gradientAST, index, KEYBOARD_CONTROL_POINT_VARIATION);
}
export function getGradientWithPositionAtIndexDecreased(gradientAST, index) {
  return getGradientWithPositionAtIndexSummed(gradientAST, index, -KEYBOARD_CONTROL_POINT_VARIATION);
}
export function getGradientWithColorAtIndexChanged(gradientAST, index, rgbaColor) {
  return _objectSpread(_objectSpread({}, gradientAST), {}, {
    colorStops: gradientAST.colorStops.map(function (colorStop, colorStopIndex) {
      if (colorStopIndex !== index) {
        return colorStop;
      }

      return _objectSpread(_objectSpread({}, colorStop), tinyColorRgbToGradientColorStop(rgbaColor));
    })
  });
}
export function getGradientWithColorAtPositionChanged(gradientAST, relativePositionValue, rgbaColor) {
  var index = findIndex(gradientAST.colorStops, function (colorStop) {
    return colorStop && colorStop.length && colorStop.length.type === '%' && colorStop.length.value === relativePositionValue.toString();
  });
  return getGradientWithColorAtIndexChanged(gradientAST, index, rgbaColor);
}
export function getGradientWithControlPointRemoved(gradientAST, index) {
  return _objectSpread(_objectSpread({}, gradientAST), {}, {
    colorStops: gradientAST.colorStops.filter(function (elem, elemIndex) {
      return elemIndex !== index;
    })
  });
}
export function getHorizontalRelativeGradientPosition(mouseXCoordinate, containerElement, positionedElementWidth) {
  if (!containerElement) {
    return;
  }

  var _containerElement$get = containerElement.getBoundingClientRect(),
      x = _containerElement$get.x,
      width = _containerElement$get.width;

  var absolutePositionValue = mouseXCoordinate - x - MINIMUM_ABSOLUTE_LEFT_POSITION - positionedElementWidth / 2;
  var availableWidth = width - MINIMUM_ABSOLUTE_LEFT_POSITION - INSERT_POINT_WIDTH;
  return Math.round(Math.min(Math.max(absolutePositionValue * 100 / availableWidth, 0), 100));
}
/**
 * Returns the marker points from a gradient AST.
 *
 * @param {Object} gradientAST An object representing the gradient AST.
 *
 * @return {Array.<{color: string, position: string, positionValue: number}>}
 *         An array of markerPoint objects.
 *         color:         A string with the color code ready to be used in css style e.g: "rgba( 1, 2 , 3, 0.5)".
 *         position:      A string with the position ready to be used in css style e.g: "70%".
 *         positionValue: A number with the relative position value e.g: 70.
 */

export function getMarkerPoints(gradientAST) {
  if (!gradientAST) {
    return [];
  }

  return map(gradientAST.colorStops, function (colorStop) {
    if (!colorStop || !colorStop.length || colorStop.length.type !== '%') {
      return null;
    }

    return {
      color: serializeGradientColor(colorStop),
      position: serializeGradientPosition(colorStop.length),
      positionValue: parseInt(colorStop.length.value)
    };
  });
}
export function getLinearGradientRepresentationOfARadial(gradientAST) {
  return serializeGradient({
    type: 'linear-gradient',
    orientation: HORIZONTAL_GRADIENT_ORIENTATION,
    colorStops: gradientAST.colorStops
  });
}
var DIRECTIONAL_ORIENTATION_ANGLE_MAP = {
  top: 0,
  'top right': 45,
  'right top': 45,
  right: 90,
  'right bottom': 135,
  'bottom right': 135,
  bottom: 180,
  'bottom left': 225,
  'left bottom': 225,
  left: 270,
  'top left': 315,
  'left top': 315
};
export function getGradientParsed(value) {
  var hasGradient = !!value; // gradientAST will contain the gradient AST as parsed by gradient-parser npm module.
  // More information of its structure available at https://www.npmjs.com/package/gradient-parser#ast.

  var gradientAST;
  var gradientValue;

  try {
    gradientAST = gradientParser.parse(value || DEFAULT_GRADIENT)[0];
    gradientValue = value || DEFAULT_GRADIENT;
  } catch (error) {
    hasGradient = false;
    gradientAST = gradientParser.parse(DEFAULT_GRADIENT)[0];
    gradientValue = DEFAULT_GRADIENT;
  }

  if (gradientAST.orientation && gradientAST.orientation.type === 'directional') {
    gradientAST.orientation.type = 'angular';
    gradientAST.orientation.value = DIRECTIONAL_ORIENTATION_ANGLE_MAP[gradientAST.orientation.value].toString();
  }

  return {
    hasGradient: hasGradient,
    gradientAST: gradientAST,
    gradientValue: gradientValue
  };
}
