// @wordpress/components/build/focusable-iframe/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _lodash = require("lodash");

var _compose = require("@wordpress/compose");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

/**
 * Browser dependencies
 */
var _window = window,
    FocusEvent = _window.FocusEvent;

var FocusableIframe = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(FocusableIframe, _Component);

  var _super = _createSuper(FocusableIframe);

  function FocusableIframe(props) {
    var _this;

    (0, _classCallCheck2.default)(this, FocusableIframe);
    _this = _super.apply(this, arguments);
    _this.checkFocus = _this.checkFocus.bind((0, _assertThisInitialized2.default)(_this));
    _this.node = props.iframeRef || (0, _element.createRef)();
    return _this;
  }
  /**
   * Checks whether the iframe is the activeElement, inferring that it has
   * then received focus, and calls the `onFocus` prop callback.
   */


  (0, _createClass2.default)(FocusableIframe, [{
    key: "checkFocus",
    value: function checkFocus() {
      var iframe = this.node.current;

      if (iframe.ownerDocument.activeElement !== iframe) {
        return;
      }

      var focusEvent = new FocusEvent('focus', {
        bubbles: true
      });
      iframe.dispatchEvent(focusEvent);
      var onFocus = this.props.onFocus;

      if (onFocus) {
        onFocus(focusEvent);
      }
    }
  }, {
    key: "render",
    value: function render() {
      // Disable reason: The rendered iframe is a pass-through component,
      // assigning props inherited from the rendering parent. It's the
      // responsibility of the parent to assign a title.

      /* eslint-disable jsx-a11y/iframe-has-title */
      return (0, _element.createElement)("iframe", (0, _extends2.default)({
        ref: this.node
      }, (0, _lodash.omit)(this.props, ['iframeRef', 'onFocus'])));
      /* eslint-enable jsx-a11y/iframe-has-title */
    }
  }]);
  return FocusableIframe;
}(_element.Component);

var _default = (0, _compose.withGlobalEvents)({
  blur: 'checkFocus'
})(FocusableIframe);

exports.default = _default;
