// @wordpress/components/build/notice/list.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _reactNative = require("react-native");

var _data = require("@wordpress/data");

var _compose = require("@wordpress/compose");

var _ = _interopRequireDefault(require("./"));

var _style = _interopRequireDefault(require("./style.scss"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var NoticeList = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(NoticeList, _Component);

  var _super = _createSuper(NoticeList);

  function NoticeList() {
    var _this;

    (0, _classCallCheck2.default)(this, NoticeList);
    _this = _super.apply(this, arguments);
    _this.removeNotice = _this.removeNotice.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(NoticeList, [{
    key: "removeNotice",
    value: function removeNotice(id) {
      var removeNotice = this.props.removeNotice;
      removeNotice(id);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          notices = _this$props.notices,
          shouldStack = _this$props.shouldStack;

      if (!notices.length) {
        return null;
      }

      return (0, _element.createElement)(_reactNative.View, {
        style: _style.default.list,
        key: notices.length
      }, shouldStack ? notices.reverse().map(function (notice) {
        return (0, _element.createElement)(_.default, (0, _extends2.default)({}, notice, {
          key: notice.id,
          onNoticeHidden: _this2.removeNotice
        }));
      }) : (0, _element.createElement)(_.default, (0, _extends2.default)({}, notices[notices.length - 1], {
        onNoticeHidden: this.removeNotice
      })));
    }
  }]);
  return NoticeList;
}(_element.Component);

var _default = (0, _compose.compose)([(0, _data.withSelect)(function (select) {
  var _select = select('core/editor'),
      getNotices = _select.getNotices;

  return {
    notices: getNotices()
  };
}), (0, _data.withDispatch)(function (dispatch) {
  var _dispatch = dispatch('core/editor'),
      removeNotice = _dispatch.removeNotice,
      removeAllNotices = _dispatch.removeAllNotices;

  return {
    removeNotice: removeNotice,
    removeAllNotices: removeAllNotices
  };
})])(NoticeList);

exports.default = _default;
