// @wordpress/components/build/resizable-box/resize-tooltip/label.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _utils = require("./utils");

var _styleMixins = require("../../utils/style-mixins");

var _resizeTooltip = require("./styles/resize-tooltip.styles");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var CORNER_OFFSET = 4;
var CURSOR_OFFSET_TOP = CORNER_OFFSET * 2.5;

function Label(_ref, ref) {
  var label = _ref.label,
      _ref$position = _ref.position,
      position = _ref$position === void 0 ? _utils.POSITIONS.corner : _ref$position,
      _ref$zIndex = _ref.zIndex,
      zIndex = _ref$zIndex === void 0 ? 1000 : _ref$zIndex,
      props = (0, _objectWithoutProperties2.default)(_ref, ["label", "position", "zIndex"]);
  var isRTL = (0, _styleMixins.useRTL)();
  var showLabel = !!label;
  var isBottom = position === _utils.POSITIONS.bottom;
  var isCorner = position === _utils.POSITIONS.corner;
  if (!showLabel) return null;
  var style = {
    opacity: showLabel ? 1 : null,
    zIndex: zIndex
  };
  var labelStyle = {};

  if (isBottom) {
    style = _objectSpread(_objectSpread({}, style), {}, {
      position: 'absolute',
      bottom: CURSOR_OFFSET_TOP * -1,
      left: '50%',
      transform: 'translate(-50%, 0)'
    });
    labelStyle = {
      transform: "translate(0, 100%)"
    };
  }

  if (isCorner) {
    style = _objectSpread(_objectSpread({}, style), {}, {
      position: 'absolute',
      top: CORNER_OFFSET,
      right: isRTL ? null : CORNER_OFFSET,
      left: isRTL ? CORNER_OFFSET : null
    });
  }

  return (0, _element.createElement)(_resizeTooltip.TooltipWrapper, (0, _extends2.default)({
    "aria-hidden": "true",
    className: "components-resizable-tooltip__tooltip-wrapper",
    isActive: showLabel,
    ref: ref,
    style: style
  }, props), (0, _element.createElement)(_resizeTooltip.Tooltip, {
    className: "components-resizable-tooltip__tooltip",
    style: labelStyle
  }, (0, _element.createElement)(_resizeTooltip.LabelText, {
    as: "span"
  }, label)));
}

var ForwardedComponent = (0, _element.forwardRef)(Label);
var _default = ForwardedComponent;
exports.default = _default;
