// @wordpress/components/build-module/button/index.native.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import { StyleSheet, TouchableOpacity, Text, View, Platform } from 'react-native';
import { isArray } from 'lodash';
/**
 * WordPress dependencies
 */

import { Children, cloneElement } from '@wordpress/element';
import { withPreferredColorScheme } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import Tooltip from '../tooltip';
import Icon from '../icon';
var isAndroid = Platform.OS === 'android';
var marginBottom = isAndroid ? -0.5 : 0;
var marginLeft = -3;
var styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 3,
    justifyContent: 'center',
    alignItems: 'center'
  },
  buttonInactive: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  fixedRatio: {
    aspectRatio: 1
  },
  buttonActive: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 6,
    borderColor: '#2e4453',
    backgroundColor: '#2e4453'
  },
  subscriptInactive: {
    color: '#7b9ab1',
    // $toolbar-button
    fontWeight: 'bold',
    fontSize: 13,
    alignSelf: 'flex-end',
    marginLeft: marginLeft,
    marginBottom: marginBottom
  },
  subscriptInactiveDark: {
    color: '#a7aaad' // $gray_20

  },
  subscriptActive: {
    color: 'white',
    fontWeight: 'bold',
    fontSize: 13,
    alignSelf: 'flex-end',
    marginLeft: marginLeft,
    marginBottom: marginBottom
  }
});
export function Button(props) {
  var children = props.children,
      onClick = props.onClick,
      onLongPress = props.onLongPress,
      disabled = props.disabled,
      hint = props.hint,
      _props$fixedRatio = props.fixedRatio,
      fixedRatio = _props$fixedRatio === void 0 ? true : _props$fixedRatio,
      getStylesFromColorScheme = props.getStylesFromColorScheme,
      isPressed = props.isPressed,
      ariaDisabled = props['aria-disabled'],
      subscript = props['data-subscript'],
      testID = props.testID,
      icon = props.icon,
      iconSize = props.iconSize,
      showTooltip = props.showTooltip,
      label = props.label,
      shortcut = props.shortcut,
      tooltipPosition = props.tooltipPosition;
  var isDisabled = ariaDisabled || disabled;

  var buttonViewStyle = _objectSpread(_objectSpread({
    opacity: isDisabled ? 0.3 : 1
  }, fixedRatio && styles.fixedRatio), isPressed ? styles.buttonActive : styles.buttonInactive);

  var states = [];

  if (isPressed) {
    states.push('selected');
  }

  if (isDisabled) {
    states.push('disabled');
  }

  var subscriptInactive = getStylesFromColorScheme(styles.subscriptInactive, styles.subscriptInactiveDark);
  var newChildren = Children.map(children, function (child) {
    return child ? cloneElement(child, {
      colorScheme: props.preferredColorScheme,
      isPressed: isPressed
    }) : child;
  }); // Should show the tooltip if...

  var shouldShowTooltip = !isDisabled && ( // an explicit tooltip is passed or...
  showTooltip && label || // there's a shortcut or...
  shortcut || // there's a label and...
  !!label && ( // the children are empty and...
  !children || isArray(children) && !children.length) && // the tooltip is not explicitly disabled.
  false !== showTooltip);
  var newIcon = icon ? cloneElement(createElement(Icon, {
    icon: icon,
    size: iconSize
  }), {
    colorScheme: props.preferredColorScheme,
    isPressed: isPressed
  }) : null;
  var element = createElement(TouchableOpacity, {
    activeOpacity: 0.7,
    accessible: true,
    accessibilityLabel: label,
    accessibilityStates: states,
    accessibilityRole: 'button',
    accessibilityHint: hint,
    onPress: onClick,
    onLongPress: onLongPress,
    style: styles.container,
    disabled: isDisabled,
    testID: testID
  }, createElement(View, {
    style: buttonViewStyle
  }, createElement(View, {
    style: {
      flexDirection: 'row'
    }
  }, newIcon, newChildren, subscript && createElement(Text, {
    style: isPressed ? styles.subscriptActive : subscriptInactive
  }, subscript))));

  if (!shouldShowTooltip) {
    return element;
  }

  return createElement(Tooltip, {
    text: label,
    shortcut: shortcut,
    position: tooltipPosition
  }, element);
}
export default withPreferredColorScheme(Button);
