// @wordpress/components/build/box-control/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BoxControl;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _lodash = require("lodash");

var _compose = require("@wordpress/compose");

var _i18n = require("@wordpress/i18n");

var _button = _interopRequireDefault(require("../button"));

var _flex = require("../flex");

var _allInputControl = _interopRequireDefault(require("./all-input-control"));

var _inputControls = _interopRequireDefault(require("./input-controls"));

var _icon = _interopRequireDefault(require("./icon"));

var _text = _interopRequireDefault(require("../text"));

var _linkedButton = _interopRequireDefault(require("./linked-button"));

var _visualizer = _interopRequireDefault(require("./visualizer"));

var _boxControlStyles = require("./styles/box-control-styles");

var _utils = require("./utils");

var _hooks = require("../utils/hooks");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var defaultInputProps = {
  min: 0
};

function useUniqueId(idProp) {
  var instanceId = (0, _compose.useInstanceId)(BoxControl, 'inspector-box-control');
  return idProp || instanceId;
}

function BoxControl(_ref) {
  var idProp = _ref.id,
      _ref$inputProps = _ref.inputProps,
      inputProps = _ref$inputProps === void 0 ? defaultInputProps : _ref$inputProps,
      _ref$onChange = _ref.onChange,
      onChange = _ref$onChange === void 0 ? _lodash.noop : _ref$onChange,
      _ref$onChangeShowVisu = _ref.onChangeShowVisualizer,
      onChangeShowVisualizer = _ref$onChangeShowVisu === void 0 ? _lodash.noop : _ref$onChangeShowVisu,
      _ref$label = _ref.label,
      label = _ref$label === void 0 ? (0, _i18n.__)('Box Control') : _ref$label,
      valuesProp = _ref.values,
      units = _ref.units;

  var _useControlledState = (0, _hooks.useControlledState)(valuesProp, {
    fallback: _utils.DEFAULT_VALUES
  }),
      _useControlledState2 = (0, _slicedToArray2.default)(_useControlledState, 2),
      values = _useControlledState2[0],
      setValues = _useControlledState2[1];

  var inputValues = values || _utils.DEFAULT_VALUES;
  var hasInitialValue = (0, _utils.isValuesDefined)(valuesProp);

  var _useState = (0, _element.useState)(hasInitialValue),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      isDirty = _useState2[0],
      setIsDirty = _useState2[1];

  var _useState3 = (0, _element.useState)(!hasInitialValue || !(0, _utils.isValuesMixed)(inputValues)),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      isLinked = _useState4[0],
      setIsLinked = _useState4[1];

  var _useState5 = (0, _element.useState)(isLinked ? 'all' : 'top'),
      _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
      side = _useState6[0],
      setSide = _useState6[1];

  var id = useUniqueId(idProp);
  var headingId = "".concat(id, "-heading");

  var toggleLinked = function toggleLinked() {
    setIsLinked(!isLinked);
    setSide(!isLinked ? 'all' : 'top');
  };

  var handleOnFocus = function handleOnFocus(event, _ref2) {
    var nextSide = _ref2.side;
    setSide(nextSide);
  };

  var handleOnChange = function handleOnChange(nextValues) {
    onChange(nextValues);
    setValues(nextValues);
    setIsDirty(true);
  };

  var handleOnHoverOn = function handleOnHoverOn() {
    var next = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    onChangeShowVisualizer(_objectSpread(_objectSpread({}, _utils.DEFAULT_VISUALIZER_VALUES), next));
  };

  var handleOnHoverOff = function handleOnHoverOff() {
    var next = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    onChangeShowVisualizer(_objectSpread(_objectSpread({}, _utils.DEFAULT_VISUALIZER_VALUES), next));
  };

  var handleOnReset = function handleOnReset() {
    var initialValues = _utils.DEFAULT_VALUES;
    onChange(initialValues);
    setValues(initialValues);
    setIsDirty(false);
  };

  var inputControlProps = _objectSpread(_objectSpread({}, inputProps), {}, {
    onChange: handleOnChange,
    onFocus: handleOnFocus,
    onHoverOn: handleOnHoverOn,
    onHoverOff: handleOnHoverOff,
    isLinked: isLinked,
    units: units,
    values: inputValues
  });

  return (0, _element.createElement)(_boxControlStyles.Root, {
    id: id,
    role: "region",
    "aria-labelledby": headingId
  }, (0, _element.createElement)(_boxControlStyles.Header, {
    className: "component-box-control__header"
  }, (0, _element.createElement)(_flex.FlexItem, null, (0, _element.createElement)(_text.default, {
    id: headingId,
    className: "component-box-control__label"
  }, label)), (0, _element.createElement)(_flex.FlexItem, null, (0, _element.createElement)(_button.default, {
    className: "component-box-control__reset-button",
    isSecondary: true,
    isSmall: true,
    onClick: handleOnReset,
    disabled: !isDirty
  }, (0, _i18n.__)('Reset')))), (0, _element.createElement)(_boxControlStyles.HeaderControlWrapper, {
    className: "component-box-control__header-control-wrapper"
  }, (0, _element.createElement)(_flex.FlexItem, null, (0, _element.createElement)(_icon.default, {
    side: side
  })), isLinked && (0, _element.createElement)(_flex.FlexBlock, null, (0, _element.createElement)(_allInputControl.default, inputControlProps)), (0, _element.createElement)(_flex.FlexItem, null, (0, _element.createElement)(_linkedButton.default, {
    onClick: toggleLinked,
    isLinked: isLinked
  }))), !isLinked && (0, _element.createElement)(_inputControls.default, inputControlProps));
}

BoxControl.__Visualizer = _visualizer.default;
