// @wordpress/components/build/box-control/unit-control.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BoxUnitControl;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _lodash = require("lodash");

var _reactUseGesture = require("react-use-gesture");

var _tooltip = _interopRequireDefault(require("../tooltip"));

var _boxControlStyles = require("./styles/box-control-styles");

/**
 * External dependencies
 */

/**
 * Internal dependencies
 */
function BoxUnitControl(_ref) {
  var isFirst = _ref.isFirst,
      isLast = _ref.isLast,
      isOnly = _ref.isOnly,
      _ref$onHoverOn = _ref.onHoverOn,
      onHoverOn = _ref$onHoverOn === void 0 ? _lodash.noop : _ref$onHoverOn,
      _ref$onHoverOff = _ref.onHoverOff,
      onHoverOff = _ref$onHoverOff === void 0 ? _lodash.noop : _ref$onHoverOff,
      label = _ref.label,
      value = _ref.value,
      props = (0, _objectWithoutProperties2.default)(_ref, ["isFirst", "isLast", "isOnly", "onHoverOn", "onHoverOff", "label", "value"]);
  var bindHoverGesture = (0, _reactUseGesture.useHover)(function (_ref2) {
    var event = _ref2.event,
        state = (0, _objectWithoutProperties2.default)(_ref2, ["event"]);

    if (state.hovering) {
      onHoverOn(event, state);
    } else {
      onHoverOff(event, state);
    }
  });
  return (0, _element.createElement)(_boxControlStyles.UnitControlWrapper, (0, _extends2.default)({
    "aria-label": label
  }, bindHoverGesture()), (0, _element.createElement)(Tooltip, {
    text: label
  }, (0, _element.createElement)(_boxControlStyles.UnitControl, (0, _extends2.default)({
    className: "component-box-control__unit-control",
    hideHTMLArrows: true,
    isFirst: isFirst,
    isLast: isLast,
    isOnly: isOnly,
    isPressEnterToChange: true,
    isResetValueOnUnitChange: false,
    value: value
  }, props))));
}

function Tooltip(_ref3) {
  var children = _ref3.children,
      text = _ref3.text;
  if (!text) return children;
  /**
   * Wrapping the children in a `<div />` as Tooltip as it attempts
   * to render the <UnitControl />. Using a plain `<div />` appears to
   * resolve this issue.
   *
   * Originally discovered and referenced here:
   * https://github.com/WordPress/gutenberg/pull/24966#issuecomment-685875026
   */

  return (0, _element.createElement)(_tooltip.default, {
    text: text,
    position: "top"
  }, (0, _element.createElement)("div", null, children));
}
