// @wordpress/components/build/mobile/segmented-control/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _reactNative = require("react-native");

var _lodash = require("lodash");

var _compose = require("@wordpress/compose");

var _layoutAnimation = require("../layout-animation");

var _style = _interopRequireDefault(require("./style.scss"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var ANIMATION_DURATION = 200;
var isIOS = _reactNative.Platform.OS === 'ios';

var Segment = function Segment(_ref) {
  var isSelected = _ref.isSelected,
      title = _ref.title,
      onPress = _ref.onPress,
      onLayout = _ref.onLayout,
      props = (0, _objectWithoutProperties2.default)(_ref, ["isSelected", "title", "onPress", "onLayout"]);
  var isSelectedIOS = isIOS && isSelected;
  var segmentStyle = [_style.default.segment, isIOS && _style.default.segmentIOS];
  var textStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.buttonTextDefault, _style.default.buttonTextDefaultDark);
  var selectedTextStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.buttonTextSelected, _style.default.buttonTextSelectedDark);
  var shadowStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.shadowIOS, {});
  return (0, _element.createElement)(_reactNative.View, {
    style: isSelectedIOS && shadowStyle
  }, (0, _element.createElement)(_reactNative.TouchableWithoutFeedback, {
    onPress: onPress
  }, (0, _element.createElement)(_reactNative.View, (0, _extends2.default)({
    style: segmentStyle,
    onLayout: onLayout
  }, props), (0, _element.createElement)(_reactNative.Text, {
    style: [textStyle, isSelected && selectedTextStyle],
    maxFontSizeMultiplier: 2
  }, title))));
};

var SegmentedControls = function SegmentedControls(_ref2) {
  var segments = _ref2.segments,
      segmentHandler = _ref2.segmentHandler,
      selectedIndex = _ref2.selectedIndex,
      addonLeft = _ref2.addonLeft,
      addonRight = _ref2.addonRight;
  var selectedSegmentIndex = selectedIndex || 0;

  var _useState = (0, _element.useState)(selectedSegmentIndex),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      activeSegmentIndex = _useState2[0],
      setActiveSegmentIndex = _useState2[1];

  var _useState3 = (0, _element.useState)((0, _defineProperty2.default)({}, activeSegmentIndex, {
    width: 0,
    height: 0
  })),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      segmentsDimensions = _useState4[0],
      setSegmentsDimensions = _useState4[1];

  var _useState6 = (0, _element.useState)(new _reactNative.Animated.Value(0)),
      _useState7 = (0, _slicedToArray2.default)(_useState6, 1),
      positionAnimationValue = _useState7[0];

  (0, _element.useEffect)(function () {
    setActiveSegmentIndex(selectedSegmentIndex);
    segmentHandler(segments[selectedSegmentIndex]);
  }, []);
  (0, _element.useEffect)(function () {
    positionAnimationValue.setValue(calculateEndValue(activeSegmentIndex));
  }, [segmentsDimensions]);
  var containerStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.container, _style.default.containerDark);

  function performSwatchAnimation(index) {
    _reactNative.Animated.timing(positionAnimationValue, {
      toValue: calculateEndValue(index),
      duration: ANIMATION_DURATION,
      easing: _reactNative.Easing.ease
    }).start();
  }

  function calculateEndValue(index) {
    var _ref3 = isIOS ? _style.default.containerIOS : _style.default.container,
        offset = _ref3.paddingLeft;

    var widths = (0, _lodash.map)((0, _lodash.values)(segmentsDimensions), 'width');
    var widthsDistance = (0, _lodash.take)(widths, index);
    var widthsDistanceSum = (0, _lodash.reduce)(widthsDistance, function (sum, n) {
      return sum + n;
    });
    var endValue = index === 0 ? 0 : widthsDistanceSum;
    return endValue + offset;
  }

  function onHandlePress(segment, index) {
    (0, _layoutAnimation.performLayoutAnimation)(ANIMATION_DURATION);
    setActiveSegmentIndex(index);
    segmentHandler(segment);
    performSwatchAnimation(index);
  }

  function segmentOnLayout(event, index) {
    var _event$nativeEvent$la = event.nativeEvent.layout,
        width = _event$nativeEvent$la.width,
        height = _event$nativeEvent$la.height;
    setSegmentsDimensions(_objectSpread(_objectSpread({}, segmentsDimensions), {}, (0, _defineProperty2.default)({}, index, {
      width: width,
      height: height
    })));
  }

  var selectedStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.selected, _style.default.selectedDark);
  var width = segmentsDimensions[activeSegmentIndex].width;
  var height = segmentsDimensions[activeSegmentIndex].height;
  var outlineStyle = [_style.default.outline, isIOS && _style.default.outlineIOS];
  return (0, _element.createElement)(_reactNative.View, {
    style: _style.default.row
  }, (0, _element.createElement)(_reactNative.View, {
    style: _style.default.flex
  }, addonLeft), (0, _element.createElement)(_reactNative.View, {
    style: [containerStyle, isIOS && _style.default.containerIOS]
  }, segments.map(function (segment, index) {
    return (0, _element.createElement)(Segment, {
      title: segment,
      onPress: function onPress() {
        return onHandlePress(segment, index);
      },
      isSelected: activeSegmentIndex === index,
      key: index,
      onLayout: function onLayout(event) {
        return segmentOnLayout(event, index);
      },
      accessibilityState: {
        selected: activeSegmentIndex === index
      },
      accessibilityRole: 'button',
      accessibilityLabel: segment,
      accessibilityHint: "".concat(index + 1, " on ").concat(segments.length)
    });
  }), (0, _element.createElement)(_reactNative.Animated.View, {
    style: [{
      width: width,
      left: positionAnimationValue,
      height: height
    }, selectedStyle, outlineStyle]
  })), (0, _element.createElement)(_reactNative.View, {
    style: _style.default.flex
  }, addonRight));
};

var _default = SegmentedControls;
exports.default = _default;
