// @wordpress/components/build/color-picker/alpha.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Alpha = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _lodash = require("lodash");

var _i18n = require("@wordpress/i18n");

var _keycodes = require("@wordpress/keycodes");

var _compose = require("@wordpress/compose");

var _utils = require("./utils");

var _keyboardShortcuts = _interopRequireDefault(require("../keyboard-shortcuts"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var Alpha = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Alpha, _Component);

  var _super = _createSuper(Alpha);

  function Alpha() {
    var _this;

    (0, _classCallCheck2.default)(this, Alpha);
    _this = _super.apply(this, arguments);
    _this.container = (0, _element.createRef)();
    _this.increase = _this.increase.bind((0, _assertThisInitialized2.default)(_this));
    _this.decrease = _this.decrease.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleChange = _this.handleChange.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleMouseDown = _this.handleMouseDown.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleMouseUp = _this.handleMouseUp.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(Alpha, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.unbindEventListeners();
    }
  }, {
    key: "increase",
    value: function increase() {
      var amount = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0.01;
      var _this$props = this.props,
          hsl = _this$props.hsl,
          _this$props$onChange = _this$props.onChange,
          onChange = _this$props$onChange === void 0 ? _lodash.noop : _this$props$onChange;
      amount = parseInt(amount * 100, 10);
      var change = {
        h: hsl.h,
        s: hsl.s,
        l: hsl.l,
        a: (parseInt(hsl.a * 100, 10) + amount) / 100,
        source: 'rgb'
      };
      onChange(change);
    }
  }, {
    key: "decrease",
    value: function decrease() {
      var amount = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0.01;
      var _this$props2 = this.props,
          hsl = _this$props2.hsl,
          _this$props2$onChange = _this$props2.onChange,
          onChange = _this$props2$onChange === void 0 ? _lodash.noop : _this$props2$onChange;
      var intValue = parseInt(hsl.a * 100, 10) - parseInt(amount * 100, 10);
      var change = {
        h: hsl.h,
        s: hsl.s,
        l: hsl.l,
        a: hsl.a <= amount ? 0 : intValue / 100,
        source: 'rgb'
      };
      onChange(change);
    }
  }, {
    key: "handleChange",
    value: function handleChange(e) {
      var _this$props$onChange2 = this.props.onChange,
          onChange = _this$props$onChange2 === void 0 ? _lodash.noop : _this$props$onChange2;
      var change = (0, _utils.calculateAlphaChange)(e, this.props, this.container.current);

      if (change) {
        onChange(change, e);
      }
    }
  }, {
    key: "handleMouseDown",
    value: function handleMouseDown(e) {
      this.handleChange(e);
      window.addEventListener('mousemove', this.handleChange);
      window.addEventListener('mouseup', this.handleMouseUp);
    }
  }, {
    key: "handleMouseUp",
    value: function handleMouseUp() {
      this.unbindEventListeners();
    }
  }, {
    key: "preventKeyEvents",
    value: function preventKeyEvents(event) {
      if (event.keyCode === _keycodes.TAB) {
        return;
      }

      event.preventDefault();
    }
  }, {
    key: "unbindEventListeners",
    value: function unbindEventListeners() {
      window.removeEventListener('mousemove', this.handleChange);
      window.removeEventListener('mouseup', this.handleMouseUp);
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var rgb = this.props.rgb;
      var rgbString = "".concat(rgb.r, ",").concat(rgb.g, ",").concat(rgb.b);
      var gradient = {
        background: "linear-gradient(to right, rgba(".concat(rgbString, ", 0) 0%, rgba(").concat(rgbString, ", 1) 100%)")
      };
      var pointerLocation = {
        left: "".concat(rgb.a * 100, "%")
      };
      var shortcuts = {
        up: function up() {
          return _this2.increase();
        },
        right: function right() {
          return _this2.increase();
        },
        'shift+up': function shiftUp() {
          return _this2.increase(0.1);
        },
        'shift+right': function shiftRight() {
          return _this2.increase(0.1);
        },
        pageup: function pageup() {
          return _this2.increase(0.1);
        },
        end: function end() {
          return _this2.increase(1);
        },
        down: function down() {
          return _this2.decrease();
        },
        left: function left() {
          return _this2.decrease();
        },
        'shift+down': function shiftDown() {
          return _this2.decrease(0.1);
        },
        'shift+left': function shiftLeft() {
          return _this2.decrease(0.1);
        },
        pagedown: function pagedown() {
          return _this2.decrease(0.1);
        },
        home: function home() {
          return _this2.decrease(1);
        }
      };
      return (0, _element.createElement)(_keyboardShortcuts.default, {
        shortcuts: shortcuts
      }, (0, _element.createElement)("div", {
        className: "components-color-picker__alpha"
      }, (0, _element.createElement)("div", {
        className: "components-color-picker__alpha-gradient",
        style: gradient
      }), (0, _element.createElement)("div", {
        className: "components-color-picker__alpha-bar",
        ref: this.container,
        onMouseDown: this.handleMouseDown,
        onTouchMove: this.handleChange,
        onTouchStart: this.handleChange
      }, (0, _element.createElement)("div", {
        tabIndex: "0",
        role: "slider",
        "aria-valuemax": "1",
        "aria-valuemin": "0",
        "aria-valuenow": rgb.a,
        "aria-orientation": "horizontal",
        "aria-label": (0, _i18n.__)('Alpha value, from 0 (transparent) to 1 (fully opaque).'),
        className: "components-color-picker__alpha-pointer",
        style: pointerLocation,
        onKeyDown: this.preventKeyEvents
      }))));
    }
  }]);
  return Alpha;
}(_element.Component);

exports.Alpha = Alpha;

var _default = (0, _compose.pure)(Alpha);

exports.default = _default;
