// @wordpress/components/build/color-picker/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _reactNative = require("react-native");

var _react = _interopRequireDefault(require("react"));

var _reactNativeHsvColorPicker = _interopRequireDefault(require("react-native-hsv-color-picker"));

var _tinycolor = _interopRequireDefault(require("tinycolor2"));

var _i18n = require("@wordpress/i18n");

var _components = require("@wordpress/components");

var _compose = require("@wordpress/compose");

var _icons = require("@wordpress/icons");

var _style = _interopRequireDefault(require("./style.scss"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function ColorPicker(_ref) {
  var shouldEnableBottomSheetScroll = _ref.shouldEnableBottomSheetScroll,
      shouldEnableBottomSheetMaxHeight = _ref.shouldEnableBottomSheetMaxHeight,
      isBottomSheetContentScrolling = _ref.isBottomSheetContentScrolling,
      setColor = _ref.setColor,
      activeColor = _ref.activeColor,
      isGradientColor = _ref.isGradientColor,
      onNavigationBack = _ref.onNavigationBack,
      onHandleClosingBottomSheet = _ref.onHandleClosingBottomSheet,
      onBottomSheetClosed = _ref.onBottomSheetClosed,
      onHandleHardwareButtonPress = _ref.onHandleHardwareButtonPress,
      bottomLabelText = _ref.bottomLabelText;
  var didMount = (0, _element.useRef)(false);
  var isIOS = _reactNative.Platform.OS === 'ios';
  var hitSlop = {
    top: 22,
    bottom: 22,
    left: 22,
    right: 22
  };

  var _ref2 = !isGradientColor && activeColor ? (0, _tinycolor.default)(activeColor).toHsv() : {
    h: 0,
    s: 0.5,
    v: 0.5
  },
      initH = _ref2.h,
      initS = _ref2.s,
      initV = _ref2.v;

  var _useState = (0, _element.useState)(initH),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      hue = _useState2[0],
      setHue = _useState2[1];

  var _useState3 = (0, _element.useState)(initS),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      sat = _useState4[0],
      setSaturation = _useState4[1];

  var _useState5 = (0, _element.useState)(initV),
      _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
      val = _useState6[0],
      setValue = _useState6[1];

  var _useState7 = (0, _element.useState)(activeColor),
      _useState8 = (0, _slicedToArray2.default)(_useState7, 1),
      savedColor = _useState8[0];

  var _styles$picker = _style.default.picker,
      spacing = _styles$picker.paddingLeft,
      pickerHeight = _styles$picker.height,
      borderRadius = _styles$picker.borderRadius;
  var pickerPointerSize = _style.default.pickerPointer.height;
  var pickerWidth = _components.BottomSheet.getWidth() - 2 * spacing;
  var applyButtonStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.applyButton, _style.default.applyButtonDark);
  var cancelButtonStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.cancelButton, _style.default.cancelButtonDark);
  var colorTextStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.colorText, _style.default.colorTextDark);
  var selectColorTextStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.selectColorText, _style.default.selectColorTextDark);
  var footerStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.footer, _style.default.footerDark);
  var currentColor = (0, _tinycolor.default)("hsv ".concat(hue, " ").concat(sat, " ").concat(val)).toHexString();
  (0, _element.useEffect)(function () {
    if (!didMount.current) {
      didMount.current = true;
      return;
    }

    setColor(currentColor);
  }, [currentColor]);
  (0, _element.useEffect)(function () {
    shouldEnableBottomSheetMaxHeight(false);
    onHandleClosingBottomSheet(function () {
      setColor(savedColor);

      if (onBottomSheetClosed) {
        onBottomSheetClosed();
      }
    });

    if (onHandleHardwareButtonPress) {
      onHandleHardwareButtonPress(onButtonPress);
    }
  }, []);

  function onHuePickerChange(_ref3) {
    var h = _ref3.hue;
    setHue(h);
  }

  function onSatValPickerChange(_ref4) {
    var s = _ref4.saturation,
        v = _ref4.value;
    setSaturation(s);
    setValue(v);
  }

  function onButtonPress(action) {
    onNavigationBack();
    onHandleClosingBottomSheet(null);
    shouldEnableBottomSheetMaxHeight(true);
    setColor(action === 'apply' ? currentColor : savedColor);

    if (onBottomSheetClosed) {
      onBottomSheetClosed();
    }
  }

  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_reactNativeHsvColorPicker.default, {
    huePickerHue: hue,
    onHuePickerDragMove: onHuePickerChange,
    onHuePickerPress: !isBottomSheetContentScrolling && onHuePickerChange,
    satValPickerHue: hue,
    satValPickerSaturation: sat,
    satValPickerValue: val,
    onSatValPickerDragMove: onSatValPickerChange,
    onSatValPickerPress: !isBottomSheetContentScrolling && onSatValPickerChange,
    onSatValPickerDragStart: function onSatValPickerDragStart() {
      shouldEnableBottomSheetScroll(false);
    },
    onSatValPickerDragEnd: function onSatValPickerDragEnd() {
      return shouldEnableBottomSheetScroll(true);
    },
    onHuePickerDragStart: function onHuePickerDragStart() {
      return shouldEnableBottomSheetScroll(false);
    },
    onHuePickerDragEnd: function onHuePickerDragEnd() {
      return shouldEnableBottomSheetScroll(true);
    },
    huePickerBarWidth: pickerWidth,
    huePickerBarHeight: pickerPointerSize / 2,
    satValPickerSize: {
      width: pickerWidth,
      height: pickerHeight
    },
    satValPickerSliderSize: pickerPointerSize * 2,
    satValPickerBorderRadius: borderRadius,
    huePickerBorderRadius: borderRadius
  }), (0, _element.createElement)(_reactNative.View, {
    style: footerStyle
  }, (0, _element.createElement)(_reactNative.TouchableWithoutFeedback, {
    onPress: function onPress() {
      return onButtonPress('cancel');
    },
    hitSlop: hitSlop
  }, (0, _element.createElement)(_reactNative.View, null, isIOS ? (0, _element.createElement)(_reactNative.Text, {
    style: cancelButtonStyle
  }, (0, _i18n.__)('Cancel')) : (0, _element.createElement)(_icons.Icon, {
    icon: _icons.close,
    size: 24,
    style: cancelButtonStyle
  }))), bottomLabelText ? (0, _element.createElement)(_reactNative.Text, {
    style: selectColorTextStyle
  }, bottomLabelText) : (0, _element.createElement)(_reactNative.Text, {
    style: colorTextStyle,
    selectable: true
  }, currentColor.toUpperCase()), (0, _element.createElement)(_reactNative.TouchableWithoutFeedback, {
    onPress: function onPress() {
      return onButtonPress('apply');
    },
    hitSlop: hitSlop
  }, (0, _element.createElement)(_reactNative.View, null, isIOS ? (0, _element.createElement)(_reactNative.Text, {
    style: applyButtonStyle
  }, (0, _i18n.__)('Apply')) : (0, _element.createElement)(_icons.Icon, {
    icon: _icons.check,
    size: 24,
    style: applyButtonStyle
  })))));
}

var _default = ColorPicker;
exports.default = _default;
