// @wordpress/components/build/mobile/color-settings/gradient-picker-screen.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _reactNative = require("react-native");

var _native = require("@react-navigation/native");

var _i18n = require("@wordpress/i18n");

var _customGradientPicker = _interopRequireDefault(require("../../custom-gradient-picker"));

var _navigationHeader = _interopRequireDefault(require("../bottom-sheet/navigation-header"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var GradientPickerScreen = function GradientPickerScreen() {
  var navigation = (0, _native.useNavigation)();
  var route = (0, _native.useRoute)();
  var _route$params = route.params,
      setColor = _route$params.setColor,
      currentValue = _route$params.currentValue,
      isGradientColor = _route$params.isGradientColor;
  return (0, _element.createElement)(_reactNative.View, null, (0, _element.createElement)(_navigationHeader.default, {
    screen: (0, _i18n.__)('Customize Gradient'),
    leftButtonOnPress: navigation.goBack
  }), (0, _element.createElement)(_customGradientPicker.default, {
    setColor: setColor,
    currentValue: currentValue,
    isGradientColor: isGradientColor
  }));
};

var _default = GradientPickerScreen;
exports.default = _default;
