// @wordpress/components/build/mobile/color-settings/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _components = require("@wordpress/components");

var _native = require("@react-navigation/native");

var _pickerScreen = _interopRequireDefault(require("./picker-screen"));

var _gradientPickerScreen = _interopRequireDefault(require("./gradient-picker-screen"));

var _palette = _interopRequireDefault(require("./palette.screen"));

var _utils = require("./utils");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var ColorSettingsMemo = _react.default.memo(function (_ref) {
  var defaultSettings = _ref.defaultSettings,
      onHandleClosingBottomSheet = _ref.onHandleClosingBottomSheet,
      shouldEnableBottomSheetMaxHeight = _ref.shouldEnableBottomSheetMaxHeight,
      onColorChange = _ref.onColorChange,
      colorValue = _ref.colorValue,
      gradientValue = _ref.gradientValue,
      onGradientChange = _ref.onGradientChange,
      label = _ref.label;
  (0, _element.useEffect)(function () {
    shouldEnableBottomSheetMaxHeight(true);
    onHandleClosingBottomSheet(null);
  }, []);
  return (0, _element.createElement)(_components.BottomSheet.NavigationContainer, null, (0, _element.createElement)(_components.BottomSheet.NavigationScreen, {
    name: _utils.colorsUtils.screens.palette,
    initialParams: {
      defaultSettings: defaultSettings,
      onColorChange: onColorChange,
      colorValue: colorValue,
      gradientValue: gradientValue,
      onGradientChange: onGradientChange,
      label: label
    }
  }, (0, _element.createElement)(_palette.default, null)), (0, _element.createElement)(_components.BottomSheet.NavigationScreen, {
    name: _utils.colorsUtils.screens.picker
  }, (0, _element.createElement)(_pickerScreen.default, null)), (0, _element.createElement)(_components.BottomSheet.NavigationScreen, {
    name: _utils.colorsUtils.screens.gradientPicker
  }, (0, _element.createElement)(_gradientPickerScreen.default, null)));
});

function ColorSettings(props) {
  var route = (0, _native.useRoute)();

  var _useContext = (0, _element.useContext)(_components.BottomSheetContext),
      onHandleClosingBottomSheet = _useContext.onHandleClosingBottomSheet,
      shouldEnableBottomSheetMaxHeight = _useContext.shouldEnableBottomSheetMaxHeight;

  return (0, _element.createElement)(ColorSettingsMemo, (0, _extends2.default)({
    onHandleClosingBottomSheet: onHandleClosingBottomSheet,
    shouldEnableBottomSheetMaxHeight: shouldEnableBottomSheetMaxHeight
  }, props, route.params));
}

var _default = ColorSettings;
exports.default = _default;
