// @wordpress/components/build/mobile/color-settings/picker-screen.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _native = require("@react-navigation/native");

var _react = _interopRequireDefault(require("react"));

var _components = require("@wordpress/components");

var _colorPicker = _interopRequireDefault(require("../../color-picker"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var PickerScreen = function PickerScreen() {
  var route = (0, _native.useRoute)();
  var navigation = (0, _native.useNavigation)();

  var _useContext = (0, _element.useContext)(_components.BottomSheetContext),
      onShouldEnableInnerHandling = _useContext.onShouldEnableInnerHandling,
      shouldEnableBottomSheetMaxHeight = _useContext.shouldEnableBottomSheetMaxHeight,
      onHandleClosingBottomSheet = _useContext.onHandleClosingBottomSheet,
      isBottomSheetContentScrolling = _useContext.isBottomSheetContentScrolling,
      shouldEnableBottomSheetScroll = _useContext.shouldEnableBottomSheetScroll,
      onHandleHardwareButtonPress = _useContext.onHandleHardwareButtonPress;

  var _route$params = route.params,
      setColor = _route$params.setColor,
      currentValue = _route$params.currentValue,
      isGradientColor = _route$params.isGradientColor;
  return (0, _element.useMemo)(function () {
    return (0, _element.createElement)(_colorPicker.default, {
      onShouldEnableInnerHandling: onShouldEnableInnerHandling,
      shouldEnableBottomSheetMaxHeight: shouldEnableBottomSheetMaxHeight,
      setColor: setColor,
      activeColor: currentValue,
      isGradientColor: isGradientColor,
      onNavigationBack: navigation.goBack,
      onHandleClosingBottomSheet: onHandleClosingBottomSheet,
      isBottomSheetContentScrolling: isBottomSheetContentScrolling,
      shouldEnableBottomSheetScroll: shouldEnableBottomSheetScroll,
      onHandleHardwareButtonPress: onHandleHardwareButtonPress
    });
  }, [setColor, currentValue, isGradientColor, onShouldEnableInnerHandling, shouldEnableBottomSheetMaxHeight, onHandleClosingBottomSheet, isBottomSheetContentScrolling, shouldEnableBottomSheetScroll, onHandleHardwareButtonPress]);
};

var _default = PickerScreen;
exports.default = _default;
