// @wordpress/components/build/mobile/image/image-editing-button.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _reactNative = require("react-native");

var _components = require("@wordpress/components");

var _mediaEdit = require("../media-edit");

var _iconCustomize = _interopRequireDefault(require("./icon-customize"));

var _style = _interopRequireDefault(require("./style.scss"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var ImageEditingButton = function ImageEditingButton(_ref) {
  var onSelectMediaUploadOption = _ref.onSelectMediaUploadOption,
      openMediaOptions = _ref.openMediaOptions,
      pickerOptions = _ref.pickerOptions,
      url = _ref.url;
  return (0, _element.createElement)(_mediaEdit.MediaEdit, {
    onSelect: onSelectMediaUploadOption,
    source: {
      uri: url
    },
    openReplaceMediaOptions: openMediaOptions,
    render: function render(_ref2) {
      var open = _ref2.open,
          mediaOptions = _ref2.mediaOptions;
      return (0, _element.createElement)(_reactNative.TouchableWithoutFeedback, {
        onPress: open
      }, (0, _element.createElement)(_reactNative.View, {
        style: _style.default.editContainer
      }, (0, _element.createElement)(_reactNative.View, {
        style: _style.default.edit
      }, mediaOptions(), (0, _element.createElement)(_components.Icon, (0, _extends2.default)({
        size: 16,
        icon: _iconCustomize.default
      }, _style.default.iconCustomise)))));
    },
    pickerOptions: pickerOptions
  });
};

var _default = ImageEditingButton;
exports.default = _default;
