// @wordpress/components/build/select-control/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _lodash = require("lodash");

var _classnames = _interopRequireDefault(require("classnames"));

var _compose = require("@wordpress/compose");

var _icons = require("@wordpress/icons");

var _baseControl = _interopRequireDefault(require("../base-control"));

var _inputBase = _interopRequireDefault(require("../input-control/input-base"));

var _selectControlStyles = require("./styles/select-control-styles");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function useUniqueId(idProp) {
  var instanceId = (0, _compose.useInstanceId)(SelectControl);
  var id = "inspector-select-control-".concat(instanceId);
  return idProp || id;
}

function SelectControl(_ref, ref) {
  var className = _ref.className,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      help = _ref.help,
      hideLabelFromVision = _ref.hideLabelFromVision,
      idProp = _ref.id,
      label = _ref.label,
      _ref$multiple = _ref.multiple,
      multiple = _ref$multiple === void 0 ? false : _ref$multiple,
      _ref$onBlur = _ref.onBlur,
      onBlur = _ref$onBlur === void 0 ? _lodash.noop : _ref$onBlur,
      _ref$onChange = _ref.onChange,
      onChange = _ref$onChange === void 0 ? _lodash.noop : _ref$onChange,
      _ref$onFocus = _ref.onFocus,
      onFocus = _ref$onFocus === void 0 ? _lodash.noop : _ref$onFocus,
      _ref$options = _ref.options,
      options = _ref$options === void 0 ? [] : _ref$options,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? 'default' : _ref$size,
      valueProp = _ref.value,
      _ref$labelPosition = _ref.labelPosition,
      labelPosition = _ref$labelPosition === void 0 ? 'top' : _ref$labelPosition,
      props = (0, _objectWithoutProperties2.default)(_ref, ["className", "disabled", "help", "hideLabelFromVision", "id", "label", "multiple", "onBlur", "onChange", "onFocus", "options", "size", "value", "labelPosition"]);

  var _useState = (0, _element.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      isFocused = _useState2[0],
      setIsFocused = _useState2[1];

  var id = useUniqueId(idProp);
  var helpId = help ? "".concat(id, "__help") : undefined; // Disable reason: A select with an onchange throws a warning

  if ((0, _lodash.isEmpty)(options)) return null;

  var handleOnBlur = function handleOnBlur(event) {
    onBlur(event);
    setIsFocused(false);
  };

  var handleOnFocus = function handleOnFocus(event) {
    onFocus(event);
    setIsFocused(true);
  };

  var handleOnChange = function handleOnChange(event) {
    if (multiple) {
      var selectedOptions = (0, _toConsumableArray2.default)(event.target.options).filter(function (_ref2) {
        var selected = _ref2.selected;
        return selected;
      });
      var newValues = selectedOptions.map(function (_ref3) {
        var value = _ref3.value;
        return value;
      });
      onChange(newValues);
      return;
    }

    onChange(event.target.value, {
      event: event
    });
  };

  var classes = (0, _classnames.default)('components-select-control', className);
  /* eslint-disable jsx-a11y/no-onchange */

  return (0, _element.createElement)(_baseControl.default, {
    help: help
  }, (0, _element.createElement)(_inputBase.default, {
    className: classes,
    disabled: disabled,
    hideLabelFromVision: hideLabelFromVision,
    id: id,
    isFocused: isFocused,
    label: label,
    size: size,
    suffix: (0, _element.createElement)(_selectControlStyles.DownArrowWrapper, null, (0, _element.createElement)(_icons.Icon, {
      icon: _icons.chevronDown,
      size: 18
    })),
    labelPosition: labelPosition
  }, (0, _element.createElement)(_selectControlStyles.Select, (0, _extends2.default)({}, props, {
    "aria-describedby": helpId,
    className: "components-select-control__input",
    disabled: disabled,
    id: id,
    multiple: multiple,
    onBlur: handleOnBlur,
    onChange: handleOnChange,
    onFocus: handleOnFocus,
    ref: ref,
    size: size,
    value: valueProp
  }), options.map(function (option, index) {
    var key = option.id || "".concat(option.label, "-").concat(option.value, "-").concat(index);
    return (0, _element.createElement)("option", {
      key: key,
      value: option.value,
      disabled: option.disabled
    }, option.label);
  }))));
  /* eslint-enable jsx-a11y/no-onchange */
}

var ForwardedComponent = (0, _element.forwardRef)(SelectControl);
var _default = ForwardedComponent;
exports.default = _default;
