// @wordpress/components/build-module/color-indicator/index.native.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { View, Animated } from 'react-native';
/**
 * WordPress dependencies
 */

import { Icon, check } from '@wordpress/icons';
import { Gradient } from '@wordpress/components';
import { usePreferredColorSchemeStyle } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import styles from './style.scss';
import { colorsUtils } from '../mobile/color-settings/utils';

function SelectedIcon(_ref) {
  var opacity = _ref.opacity;
  return createElement(Animated.View, {
    style: [styles.selected, {
      opacity: opacity
    }]
  }, createElement(View, {
    style: [styles.outline, styles.selectedOutline]
  }), createElement(Icon, {
    icon: check,
    style: styles.icon,
    size: 24
  }));
}

function ColorIndicator(_ref2) {
  var color = _ref2.color,
      isSelected = _ref2.isSelected,
      withCustomPicker = _ref2.withCustomPicker,
      style = _ref2.style,
      opacity = _ref2.opacity;
  var isGradient = colorsUtils.isGradient;
  var outlineStyle = usePreferredColorSchemeStyle(styles.outline, styles.outlineDark);

  if (isGradient(color)) {
    return createElement(Gradient, {
      style: [styles.circleOption, style],
      gradientValue: color
    }, createElement(View, {
      style: outlineStyle
    }), isSelected && createElement(SelectedIcon, {
      opacity: opacity
    }));
  } else if (withCustomPicker) {
    return createElement(View, {
      style: [styles.circleOption, style]
    }, createElement(View, {
      style: outlineStyle
    }), color.map(function (gradientValue) {
      return createElement(Gradient, {
        gradientValue: gradientValue,
        style: [styles.circleOption, styles.absolute, style],
        key: gradientValue
      });
    }), isSelected && createElement(SelectedIcon, null));
  }

  return createElement(View, {
    style: [styles.circleOption, style, {
      backgroundColor: color
    }]
  }, createElement(View, {
    style: outlineStyle
  }), isSelected && createElement(SelectedIcon, {
    opacity: opacity
  }));
}

export default ColorIndicator;
