// @wordpress/components/build/color-palette/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _reactNative = require("react-native");

var _lodash = require("lodash");

var _i18n = require("@wordpress/i18n");

var _compose = require("@wordpress/compose");

var _style = _interopRequireDefault(require("./style.scss"));

var _colorIndicator = _interopRequireDefault(require("../color-indicator"));

var _utils = require("../mobile/color-settings/utils");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var ANIMATION_DURATION = 200;
var contentWidth = 0;
var scrollPosition = 0;
var customIndicatorWidth = 0;

function ColorPalette(_ref) {
  var setColor = _ref.setColor,
      activeColor = _ref.activeColor,
      isGradientColor = _ref.isGradientColor,
      defaultSettings = _ref.defaultSettings,
      currentSegment = _ref.currentSegment,
      onCustomPress = _ref.onCustomPress,
      shouldEnableBottomSheetScroll = _ref.shouldEnableBottomSheetScroll,
      _ref$shouldShowCustom = _ref.shouldShowCustomIndicatorOption,
      shouldShowCustomIndicatorOption = _ref$shouldShowCustom === void 0 ? true : _ref$shouldShowCustom,
      _ref$shouldShowCustom2 = _ref.shouldShowCustomLabel,
      shouldShowCustomLabel = _ref$shouldShowCustom2 === void 0 ? true : _ref$shouldShowCustom2,
      _ref$shouldShowCustom3 = _ref.shouldShowCustomVerticalSeparator,
      shouldShowCustomVerticalSeparator = _ref$shouldShowCustom3 === void 0 ? true : _ref$shouldShowCustom3,
      customColorIndicatorStyles = _ref.customColorIndicatorStyles,
      customIndicatorWrapperStyles = _ref.customIndicatorWrapperStyles;
  var customSwatchGradients = ['linear-gradient(120deg, rgba(255,0,0,.8), 0%, rgba(255,255,255,1) 70.71%)', 'linear-gradient(240deg, rgba(0,255,0,.8), 0%, rgba(0,255,0,0) 70.71%)', 'linear-gradient(360deg, rgba(0,0,255,.8), 0%, rgba(0,0,255,0) 70.71%)'];
  var scrollViewRef = (0, _element.useRef)();
  var isIOS = _reactNative.Platform.OS === 'ios';
  var isGradientSegment = currentSegment === _utils.colorsUtils.segments[1];
  var scale = (0, _element.useRef)(new _reactNative.Animated.Value(1)).current;
  var opacity = (0, _element.useRef)(new _reactNative.Animated.Value(1)).current;
  var defaultColors = (0, _lodash.uniq)((0, _lodash.map)(defaultSettings.colors, 'color'));
  var defaultGradientColors = (0, _lodash.uniq)((0, _lodash.map)(defaultSettings.gradients, 'gradient'));
  var colors = isGradientSegment ? defaultGradientColors : defaultColors;
  var customIndicatorColor = isGradientSegment ? activeColor : customSwatchGradients;
  var isCustomGradientColor = isGradientColor && isSelectedCustom();
  var shouldShowCustomIndicator = shouldShowCustomIndicatorOption && (!isGradientSegment || isCustomGradientColor);
  var accessibilityHint = isGradientSegment ? (0, _i18n.__)('Navigates to customize the gradient') : (0, _i18n.__)('Navigates to custom color picker');
  var customText = (0, _i18n.__)('Custom');
  (0, _element.useEffect)(function () {
    if (scrollViewRef.current) {
      if (isSelectedCustom()) {
        scrollViewRef.current.scrollToEnd();
      } else {
        scrollViewRef.current.scrollTo({
          x: 0,
          y: 0
        });
      }
    }
  }, [currentSegment]);

  function isSelectedCustom() {
    var isWithinColors = activeColor && colors.includes(activeColor);

    if (activeColor) {
      if (isGradientSegment) {
        return isGradientColor && !isWithinColors;
      }

      return !isGradientColor && !isWithinColors;
    }

    return false;
  }

  function isSelected(color) {
    return !isSelectedCustom() && activeColor === color;
  }

  function timingAnimation(property, toValue) {
    return _reactNative.Animated.timing(property, {
      toValue: toValue,
      duration: ANIMATION_DURATION,
      easing: _reactNative.Easing.ease,
      useNativeDriver: true
    });
  }

  function performAnimation(color) {
    if (!isSelected(color)) {
      opacity.setValue(0);
    }

    _reactNative.Animated.parallel([timingAnimation(scale, 2), timingAnimation(opacity, 1)]).start(function () {
      opacity.setValue(1);
      scale.setValue(1);
    });
  }

  var scaleInterpolation = scale.interpolate({
    inputRange: [1, 1.5, 2],
    outputRange: [1, 0.7, 1]
  });

  function deselectCustomGradient() {
    var _Dimensions$get = _reactNative.Dimensions.get('window'),
        width = _Dimensions$get.width;

    var isVisible = contentWidth - scrollPosition - customIndicatorWidth < width;

    if (isCustomGradientColor) {
      if (!isIOS) {
        // Scroll position on Android doesn't adjust automatically when removing the last item from the horizontal list.
        // https://github.com/facebook/react-native/issues/27504
        // Workaround: Force the scroll when deselecting custom gradient color and when custom indicator is visible on layout.
        if (isCustomGradientColor && isVisible && scrollViewRef.current) {
          scrollViewRef.current.scrollTo({
            x: scrollPosition - customIndicatorWidth
          });
        }
      }
    }
  }

  function onColorPress(color) {
    deselectCustomGradient();
    performAnimation(color);
    setColor(color);
  }

  function onContentSizeChange(width) {
    contentWidth = width;

    if (isSelectedCustom() && scrollViewRef.current) {
      scrollViewRef.current.scrollToEnd({
        animated: !isIOS
      });
    }
  }

  function onCustomIndicatorLayout(_ref2) {
    var nativeEvent = _ref2.nativeEvent;
    var width = nativeEvent.layout.width;

    if (width !== customIndicatorWidth) {
      customIndicatorWidth = width;
    }
  }

  function onScroll(_ref3) {
    var nativeEvent = _ref3.nativeEvent;
    scrollPosition = nativeEvent.contentOffset.x;
  }

  var verticalSeparatorStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.verticalSeparator, _style.default.verticalSeparatorDark);
  var customTextStyle = (0, _compose.usePreferredColorSchemeStyle)([_style.default.customText, !isIOS && _style.default.customTextAndroid], _style.default.customTextDark);
  var customIndicatorWrapperStyle = [_style.default.customIndicatorWrapper, customIndicatorWrapperStyles];
  return (0, _element.createElement)(_reactNative.ScrollView, {
    contentContainerStyle: _style.default.contentContainer,
    style: _style.default.container,
    horizontal: true,
    showsHorizontalScrollIndicator: false,
    keyboardShouldPersistTaps: "always",
    disableScrollViewPanResponder: true,
    scrollEventThrottle: 16,
    onScroll: onScroll,
    onContentSizeChange: onContentSizeChange,
    onScrollBeginDrag: function onScrollBeginDrag() {
      return shouldEnableBottomSheetScroll(false);
    },
    onScrollEndDrag: function onScrollEndDrag() {
      return shouldEnableBottomSheetScroll(true);
    },
    ref: scrollViewRef
  }, colors.map(function (color) {
    var scaleValue = isSelected(color) ? scaleInterpolation : 1;
    return (0, _element.createElement)(_reactNative.TouchableWithoutFeedback, {
      onPress: function onPress() {
        return onColorPress(color);
      },
      key: "".concat(color, "-").concat(isSelected(color)),
      accessibilityRole: 'button',
      accessibilityState: {
        selected: isSelected(color)
      },
      accessibilityHint: color
    }, (0, _element.createElement)(_reactNative.Animated.View, {
      style: {
        transform: [{
          scale: scaleValue
        }]
      }
    }, (0, _element.createElement)(_colorIndicator.default, {
      color: color,
      isSelected: isSelected(color),
      opacity: opacity,
      style: [_style.default.colorIndicator, customColorIndicatorStyles]
    })));
  }), shouldShowCustomIndicator && (0, _element.createElement)(_reactNative.View, {
    style: customIndicatorWrapperStyle,
    onLayout: onCustomIndicatorLayout
  }, shouldShowCustomVerticalSeparator && (0, _element.createElement)(_reactNative.View, {
    style: verticalSeparatorStyle
  }), (0, _element.createElement)(_reactNative.TouchableWithoutFeedback, {
    onPress: onCustomPress,
    accessibilityRole: 'button',
    accessibilityState: {
      selected: isSelectedCustom()
    },
    accessibilityHint: accessibilityHint
  }, (0, _element.createElement)(_reactNative.View, {
    style: customIndicatorWrapperStyle
  }, (0, _element.createElement)(_colorIndicator.default, {
    withCustomPicker: !isGradientSegment,
    color: customIndicatorColor,
    isSelected: isSelectedCustom(),
    style: [_style.default.colorIndicator, customColorIndicatorStyles]
  }), shouldShowCustomLabel && (0, _element.createElement)(_reactNative.Text, {
    style: customTextStyle
  }, isIOS ? customText : customText.toUpperCase())))));
}

var _default = ColorPalette;
exports.default = _default;
