// @wordpress/components/build-module/mobile/bottom-sheet/button.native.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import { TouchableOpacity, View, Text } from 'react-native';
/**
 * Internal dependencies
 */

import styles from './styles.scss';

var BottomSheetButton = function BottomSheetButton(_ref) {
  var onPress = _ref.onPress,
      disabled = _ref.disabled,
      text = _ref.text,
      color = _ref.color;
  return createElement(TouchableOpacity, {
    accessible: true,
    onPress: onPress,
    disabled: disabled
  }, createElement(View, {
    style: {
      flexDirection: 'row',
      justifyContent: 'center'
    }
  }, createElement(Text, {
    style: _objectSpread(_objectSpread({}, styles.buttonText), {}, {
      color: color
    })
  }, text)));
};

export default BottomSheetButton;
