// @wordpress/components/build-module/mobile/bottom-sheet/cell.native.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

/**
 * External dependencies
 */
import { TouchableOpacity, Text, View, TextInput, I18nManager, AccessibilityInfo } from 'react-native';
import { isEmpty, get } from 'lodash';
/**
 * WordPress dependencies
 */

import { Icon } from '@wordpress/components';
import { check } from '@wordpress/icons';
import { Component } from '@wordpress/element';
import { __, _x, sprintf } from '@wordpress/i18n';
import { withPreferredColorScheme } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import styles from './styles.scss';
import platformStyles from './cellStyles.scss';
import TouchableRipple from './ripple.native.js';

var BottomSheetCell = /*#__PURE__*/function (_Component) {
  _inherits(BottomSheetCell, _Component);

  var _super = _createSuper(BottomSheetCell);

  function BottomSheetCell(props) {
    var _this;

    _classCallCheck(this, BottomSheetCell);

    _this = _super.apply(this, arguments);
    _this.state = {
      isEditingValue: props.autoFocus || false,
      isScreenReaderEnabled: false
    };
    _this.handleScreenReaderToggled = _this.handleScreenReaderToggled.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(BottomSheetCell, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps, prevState) {
      if (!prevState.isEditingValue && this.state.isEditingValue) {
        this._valueTextInput.focus();
      }
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      AccessibilityInfo.addEventListener('screenReaderChanged', this.handleScreenReaderToggled);
      AccessibilityInfo.isScreenReaderEnabled().then(function (isScreenReaderEnabled) {
        _this2.setState({
          isScreenReaderEnabled: isScreenReaderEnabled
        });
      });
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      AccessibilityInfo.removeEventListener('screenReaderChanged', this.handleScreenReaderToggled);
    }
  }, {
    key: "handleScreenReaderToggled",
    value: function handleScreenReaderToggled(isScreenReaderEnabled) {
      this.setState({
        isScreenReaderEnabled: isScreenReaderEnabled
      });
    }
  }, {
    key: "typeToKeyboardType",
    value: function typeToKeyboardType(type, step) {
      var keyboardType = "default";

      if (type === "number") {
        if (step && Math.abs(step) < 1) {
          keyboardType = "decimal-pad";
        } else {
          keyboardType = "number-pad";
        }
      }

      return keyboardType;
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props = this.props,
          accessible = _this$props.accessible,
          accessibilityLabel = _this$props.accessibilityLabel,
          accessibilityHint = _this$props.accessibilityHint,
          accessibilityRole = _this$props.accessibilityRole,
          _this$props$disabled = _this$props.disabled,
          disabled = _this$props$disabled === void 0 ? false : _this$props$disabled,
          activeOpacity = _this$props.activeOpacity,
          onPress = _this$props.onPress,
          label = _this$props.label,
          value = _this$props.value,
          _this$props$valuePlac = _this$props.valuePlaceholder,
          valuePlaceholder = _this$props$valuePlac === void 0 ? '' : _this$props$valuePlac,
          icon = _this$props.icon,
          leftAlign = _this$props.leftAlign,
          _this$props$labelStyl = _this$props.labelStyle,
          labelStyle = _this$props$labelStyl === void 0 ? {} : _this$props$labelStyl,
          _this$props$valueStyl = _this$props.valueStyle,
          valueStyle = _this$props$valueStyl === void 0 ? {} : _this$props$valueStyl,
          _this$props$cellConta = _this$props.cellContainerStyle,
          cellContainerStyle = _this$props$cellConta === void 0 ? {} : _this$props$cellConta,
          _this$props$cellRowCo = _this$props.cellRowContainerStyle,
          cellRowContainerStyle = _this$props$cellRowCo === void 0 ? {} : _this$props$cellRowCo,
          onChangeValue = _this$props.onChangeValue,
          onSubmit = _this$props.onSubmit,
          children = _this$props.children,
          _this$props$editable = _this$props.editable,
          editable = _this$props$editable === void 0 ? true : _this$props$editable,
          _this$props$isSelecte = _this$props.isSelected,
          isSelected = _this$props$isSelecte === void 0 ? false : _this$props$isSelecte,
          separatorType = _this$props.separatorType,
          _this$props$style = _this$props.style,
          style = _this$props$style === void 0 ? {} : _this$props$style,
          getStylesFromColorScheme = _this$props.getStylesFromColorScheme,
          customActionButton = _this$props.customActionButton,
          type = _this$props.type,
          step = _this$props.step,
          borderless = _this$props.borderless,
          valueProps = _objectWithoutProperties(_this$props, ["accessible", "accessibilityLabel", "accessibilityHint", "accessibilityRole", "disabled", "activeOpacity", "onPress", "label", "value", "valuePlaceholder", "icon", "leftAlign", "labelStyle", "valueStyle", "cellContainerStyle", "cellRowContainerStyle", "onChangeValue", "onSubmit", "children", "editable", "isSelected", "separatorType", "style", "getStylesFromColorScheme", "customActionButton", "type", "step", "borderless"]);

      var showValue = value !== undefined;
      var isValueEditable = editable && onChangeValue !== undefined;
      var cellLabelStyle = getStylesFromColorScheme(styles.cellLabel, styles.cellTextDark);
      var cellLabelCenteredStyle = getStylesFromColorScheme(styles.cellLabelCentered, styles.cellTextDark);
      var cellLabelLeftAlignNoIconStyle = getStylesFromColorScheme(styles.cellLabelLeftAlignNoIcon, styles.cellTextDark);
      var defaultMissingIconAndValue = leftAlign ? cellLabelLeftAlignNoIconStyle : cellLabelCenteredStyle;
      var defaultLabelStyle = showValue || customActionButton || icon ? cellLabelStyle : defaultMissingIconAndValue;
      var drawSeparator = separatorType && separatorType !== 'none' || separatorStyle === undefined;
      var drawTopSeparator = drawSeparator && separatorType === 'topFullWidth';
      var cellContainerStyles = [styles.cellContainer, cellContainerStyle];
      var rowContainerStyles = [styles.cellRowContainer, cellRowContainerStyle];

      var onCellPress = function onCellPress() {
        if (isValueEditable) {
          startEditing();
        } else if (onPress !== undefined) {
          onPress();
        }
      };

      var finishEditing = function finishEditing() {
        _this3.setState({
          isEditingValue: false
        });
      };

      var startEditing = function startEditing() {
        if (_this3.state.isEditingValue === false) {
          _this3.setState({
            isEditingValue: true
          });
        }
      };

      var separatorStyle = function separatorStyle() {
        //eslint-disable-next-line @wordpress/no-unused-vars-before-return
        var defaultSeparatorStyle = _this3.props.getStylesFromColorScheme(styles.separator, styles.separatorDark);

        var cellSeparatorStyle = _this3.props.getStylesFromColorScheme(styles.cellSeparator, styles.cellSeparatorDark);

        var leftMarginStyle = _objectSpread(_objectSpread({}, cellSeparatorStyle), platformStyles.separatorMarginLeft);

        switch (separatorType) {
          case 'leftMargin':
            return leftMarginStyle;

          case 'fullWidth':
          case 'topFullWidth':
            return defaultSeparatorStyle;

          case 'none':
            return undefined;

          case undefined:
            if (showValue && icon) {
              return leftMarginStyle;
            }

            return defaultSeparatorStyle;
        }
      };

      var getValueComponent = function getValueComponent() {
        var styleRTL = I18nManager.isRTL && styles.cellValueRTL;

        var cellValueStyle = _this3.props.getStylesFromColorScheme(styles.cellValue, styles.cellTextDark);

        var finalStyle = _objectSpread(_objectSpread(_objectSpread({}, cellValueStyle), valueStyle), styleRTL); // To be able to show the `middle` ellipsizeMode on editable cells
        // we show the TextInput just when the user wants to edit the value,
        // and the Text component to display it.
        // We also show the TextInput to display placeholder.


        var shouldShowPlaceholder = isValueEditable && value === '';
        return _this3.state.isEditingValue || shouldShowPlaceholder ? createElement(TextInput, _extends({
          ref: function ref(c) {
            return _this3._valueTextInput = c;
          },
          numberOfLines: 1,
          style: finalStyle,
          value: value,
          placeholder: valuePlaceholder,
          placeholderTextColor: '#87a6bc',
          onChangeText: onChangeValue,
          editable: isValueEditable,
          pointerEvents: _this3.state.isEditingValue ? 'auto' : 'none',
          onFocus: startEditing,
          onBlur: finishEditing,
          onSubmitEditing: onSubmit,
          keyboardType: _this3.typeToKeyboardType(type, step)
        }, valueProps)) : createElement(Text, {
          style: _objectSpread(_objectSpread({}, cellValueStyle), valueStyle),
          numberOfLines: 1,
          ellipsizeMode: 'middle'
        }, value);
      };

      var getAccessibilityLabel = function getAccessibilityLabel() {
        if (accessibilityLabel || !showValue) {
          return accessibilityLabel || label;
        }

        return isEmpty(value) ? sprintf(
        /* translators: accessibility text. Empty state of a inline textinput cell. %s: The cell's title */
        _x('%s. Empty', 'inline textinput cell'), label) : // Separating by ',' is necessary to make a pause on urls (non-capitalized text)
        sprintf(
        /* translators: accessibility text. Inline textinput title and value.%1: Cell title, %2: cell value. */
        _x('%1$s, %2$s', 'inline textinput cell'), label, value);
      };

      var iconStyle = getStylesFromColorScheme(styles.icon, styles.iconDark);
      var resetButtonStyle = getStylesFromColorScheme(styles.resetButton, styles.resetButtonDark);
      var containerPointerEvents = this.state.isScreenReaderEnabled && accessible ? 'none' : 'auto';

      var _ref = customActionButton || {},
          title = _ref.title,
          handler = _ref.handler;

      var opacity = activeOpacity !== undefined ? activeOpacity : get(platformStyles, 'activeOpacity.opacity');
      return createElement(TouchableRipple, {
        accessible: accessible !== undefined ? accessible : !this.state.isEditingValue,
        accessibilityLabel: getAccessibilityLabel(),
        accessibilityRole: accessibilityRole || 'button',
        accessibilityHint: isValueEditable ?
        /* translators: accessibility text */
        __('Double tap to edit this value') : accessibilityHint,
        disabled: disabled,
        activeOpacity: opacity,
        onPress: onCellPress,
        style: [styles.clipToBounds, style],
        borderless: borderless
      }, drawTopSeparator && createElement(View, {
        style: separatorStyle()
      }), createElement(View, {
        style: cellContainerStyles,
        pointerEvents: containerPointerEvents
      }, createElement(View, {
        style: rowContainerStyles
      }, createElement(View, {
        style: styles.cellRowContainer
      }, icon && createElement(View, {
        style: styles.cellRowContainer
      }, createElement(Icon, {
        icon: icon,
        size: 24,
        fill: iconStyle.color,
        isPressed: false
      }), createElement(View, {
        style: platformStyles.labelIconSeparator
      })), label && createElement(Text, {
        style: [defaultLabelStyle, labelStyle]
      }, label)), customActionButton && createElement(TouchableOpacity, {
        onPress: handler,
        accessibilityRole: 'button'
      }, createElement(Text, {
        style: resetButtonStyle
      }, title))), isSelected && createElement(Icon, {
        icon: check,
        fill: platformStyles.isSelected.color
      }), showValue && getValueComponent(), children), !drawTopSeparator && createElement(View, {
        style: separatorStyle()
      }));
    }
  }]);

  return BottomSheetCell;
}(Component);

export default withPreferredColorScheme(BottomSheetCell);
