// @wordpress/components/build-module/mobile/bottom-sheet/radio-cell.native.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { Icon, check } from '@wordpress/icons';
import { usePreferredColorSchemeStyle } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import Cell from './cell';
import styles from './styles.scss';
export default function BottomSheetColorCell(props) {
  var selected = props.selected,
      cellProps = _objectWithoutProperties(props, ["selected"]);

  var selectedIconStyle = usePreferredColorSchemeStyle(styles.selectedIcon, styles.selectedIconDark);
  return createElement(Cell, _extends({}, cellProps, {
    accessibilityRole: 'radio',
    accessibilityState: {
      selected: selected
    },
    accessibilityHint:
    /* translators: accessibility text (hint for selecting option) */
    __('Double tap to select the option'),
    editable: false,
    value: ''
  }), selected && createElement(Icon, {
    icon: check,
    style: selectedIconStyle
  }));
}
