// @wordpress/components/build/circular-option-picker/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CircularOptionPicker;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classnames = _interopRequireDefault(require("classnames"));

var _icons = require("@wordpress/icons");

var _button = _interopRequireDefault(require("../button"));

var _dropdown = _interopRequireDefault(require("../dropdown"));

var _tooltip = _interopRequireDefault(require("../tooltip"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function Option(_ref) {
  var className = _ref.className,
      isSelected = _ref.isSelected,
      selectedIconProps = _ref.selectedIconProps,
      tooltipText = _ref.tooltipText,
      additionalProps = (0, _objectWithoutProperties2.default)(_ref, ["className", "isSelected", "selectedIconProps", "tooltipText"]);
  var optionButton = (0, _element.createElement)(_button.default, (0, _extends2.default)({
    isPressed: isSelected,
    className: (0, _classnames.default)(className, 'components-circular-option-picker__option')
  }, additionalProps));
  return (0, _element.createElement)("div", {
    className: "components-circular-option-picker__option-wrapper"
  }, tooltipText ? (0, _element.createElement)(_tooltip.default, {
    text: tooltipText
  }, optionButton) : optionButton, isSelected && (0, _element.createElement)(_icons.Icon, (0, _extends2.default)({
    icon: _icons.check
  }, selectedIconProps ? selectedIconProps : {})));
}

function DropdownLinkAction(_ref2) {
  var buttonProps = _ref2.buttonProps,
      className = _ref2.className,
      dropdownProps = _ref2.dropdownProps,
      linkText = _ref2.linkText;
  return (0, _element.createElement)(_dropdown.default, (0, _extends2.default)({
    className: (0, _classnames.default)('components-circular-option-picker__dropdown-link-action', className),
    renderToggle: function renderToggle(_ref3) {
      var isOpen = _ref3.isOpen,
          onToggle = _ref3.onToggle;
      return (0, _element.createElement)(_button.default, (0, _extends2.default)({
        "aria-expanded": isOpen,
        "aria-haspopup": "true",
        onClick: onToggle,
        isLink: true
      }, buttonProps), linkText);
    }
  }, dropdownProps));
}

function ButtonAction(_ref4) {
  var className = _ref4.className,
      children = _ref4.children,
      additionalProps = (0, _objectWithoutProperties2.default)(_ref4, ["className", "children"]);
  return (0, _element.createElement)(_button.default, (0, _extends2.default)({
    className: (0, _classnames.default)('components-circular-option-picker__clear', className),
    isSmall: true,
    isSecondary: true
  }, additionalProps), children);
}

function CircularOptionPicker(_ref5) {
  var actions = _ref5.actions,
      className = _ref5.className,
      options = _ref5.options,
      children = _ref5.children;
  return (0, _element.createElement)("div", {
    className: (0, _classnames.default)('components-circular-option-picker', className)
  }, (0, _element.createElement)("div", {
    className: "components-circular-option-picker__swatches"
  }, options), children, actions && (0, _element.createElement)("div", {
    className: "components-circular-option-picker__custom-clear-wrapper"
  }, actions));
}

CircularOptionPicker.Option = Option;
CircularOptionPicker.ButtonAction = ButtonAction;
CircularOptionPicker.DropdownLinkAction = DropdownLinkAction;
