// @wordpress/components/build/navigation/use-create-navigation-tree.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCreateNavigationTree = void 0;

var _useNavigationTreeNodes = require("./use-navigation-tree-nodes");

/**
 * Internal dependencies
 */
var useCreateNavigationTree = function useCreateNavigationTree() {
  var _useNavigationTreeNod = (0, _useNavigationTreeNodes.useNavigationTreeNodes)(),
      items = _useNavigationTreeNod.nodes,
      getItem = _useNavigationTreeNod.getNode,
      addItem = _useNavigationTreeNod.addNode,
      removeItem = _useNavigationTreeNod.removeNode;

  var _useNavigationTreeNod2 = (0, _useNavigationTreeNodes.useNavigationTreeNodes)(),
      menus = _useNavigationTreeNod2.nodes,
      getMenu = _useNavigationTreeNod2.getNode,
      addMenu = _useNavigationTreeNod2.addNode,
      removeMenu = _useNavigationTreeNod2.removeNode;

  return {
    items: items,
    getItem: getItem,
    addItem: addItem,
    removeItem: removeItem,
    menus: menus,
    getMenu: getMenu,
    addMenu: addMenu,
    removeMenu: removeMenu
  };
};

exports.useCreateNavigationTree = useCreateNavigationTree;
