// @wordpress/components/build/mobile/picker/index.android.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _i18n = require("@wordpress/i18n");

var _compose = require("@wordpress/compose");

var _components = require("@wordpress/components");

var _bottomSheet = _interopRequireDefault(require("../bottom-sheet"));

var _styles = _interopRequireDefault(require("./styles.scss"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function Separator() {
  var separatorStyle = (0, _compose.usePreferredColorSchemeStyle)(_styles.default.separator, _styles.default.separatorDark);
  return (0, _element.createElement)(_reactNative.View, {
    style: separatorStyle
  });
}

var Picker = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Picker, _Component);

  var _super = _createSuper(Picker);

  function Picker() {
    var _this;

    (0, _classCallCheck2.default)(this, Picker);
    _this = _super.apply(this, arguments);
    _this.onClose = _this.onClose.bind((0, _assertThisInitialized2.default)(_this));
    _this.onCellPress = _this.onCellPress.bind((0, _assertThisInitialized2.default)(_this));
    _this.state = {
      isVisible: false
    };
    return _this;
  }

  (0, _createClass2.default)(Picker, [{
    key: "presentPicker",
    value: function presentPicker() {
      this.setState({
        isVisible: true
      });
    }
  }, {
    key: "onClose",
    value: function onClose() {
      this.setState({
        isVisible: false
      });
    }
  }, {
    key: "onCellPress",
    value: function onCellPress(value) {
      var onChange = this.props.onChange;
      onChange(value);
      this.onClose();
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          options = _this$props.options,
          leftAlign = _this$props.leftAlign,
          hideCancelButton = _this$props.hideCancelButton,
          title = _this$props.title;
      var isVisible = this.state.isVisible;
      return (0, _element.createElement)(_bottomSheet.default, {
        isVisible: isVisible,
        onClose: this.onClose,
        style: {
          paddingBottom: 20
        },
        hideHeader: true
      }, (0, _element.createElement)(_components.PanelBody, {
        title: title,
        style: _styles.default.panelBody
      }, options.map(function (option, index) {
        return (0, _element.createElement)(_element.Fragment, null, options.length > 1 && option.separated && (0, _element.createElement)(Separator, null), (0, _element.createElement)(_bottomSheet.default.Cell, {
          icon: option.icon,
          key: index,
          leftAlign: leftAlign,
          label: option.label,
          separatorType: 'none',
          onPress: function onPress() {
            return _this2.onCellPress(option.value);
          },
          disabled: option.disabled,
          style: option.disabled && _styles.default.disabled
        }));
      }), !hideCancelButton && (0, _element.createElement)(_bottomSheet.default.Cell, {
        label: (0, _i18n.__)('Cancel'),
        onPress: this.onClose,
        separatorType: 'none'
      })));
    }
  }]);
  return Picker;
}(_element.Component);

exports.default = Picker;
