// @wordpress/components/build/mobile/picker/index.ios.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _reactNative = require("react-native");

var _i18n = require("@wordpress/i18n");

var _element = require("@wordpress/element");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var Picker = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(Picker, _Component);

  var _super = _createSuper(Picker);

  function Picker() {
    (0, _classCallCheck2.default)(this, Picker);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(Picker, [{
    key: "presentPicker",
    value: function presentPicker() {
      var _this$props = this.props,
          options = _this$props.options,
          onChange = _this$props.onChange,
          title = _this$props.title,
          destructiveButtonIndex = _this$props.destructiveButtonIndex,
          disabledButtonIndices = _this$props.disabledButtonIndices,
          anchor = _this$props.anchor;
      var labels = options.map(function (_ref) {
        var label = _ref.label;
        return label;
      });
      var fullOptions = [(0, _i18n.__)('Cancel')].concat(labels);

      _reactNative.ActionSheetIOS.showActionSheetWithOptions({
        title: title,
        options: fullOptions,
        cancelButtonIndex: 0,
        destructiveButtonIndex: destructiveButtonIndex,
        disabledButtonIndices: disabledButtonIndices,
        anchor: anchor
      }, function (buttonIndex) {
        if (buttonIndex === 0) {
          return;
        }

        var selected = options[buttonIndex - 1];
        onChange(selected.value);
      });
    }
  }, {
    key: "render",
    value: function render() {
      return null;
    }
  }]);
  return Picker;
}(_element.Component);

var _default = Picker;
exports.default = _default;
