// @wordpress/blocks/build/api/templates.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.doBlocksMatchTemplate = doBlocksMatchTemplate;
exports.synchronizeBlocksWithTemplate = synchronizeBlocksWithTemplate;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _lodash = require("lodash");

var _element = require("@wordpress/element");

var _parser = require("./parser");

var _factory = require("./factory");

var _registration = require("./registration");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * Checks whether a list of blocks matches a template by comparing the block names.
 *
 * @param {Array} blocks    Block list.
 * @param {Array} template  Block template.
 *
 * @return {boolean}        Whether the list of blocks matches a templates
 */
function doBlocksMatchTemplate() {
  var blocks = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  var template = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  return blocks.length === template.length && (0, _lodash.every)(template, function (_ref, index) {
    var _ref2 = (0, _slicedToArray2.default)(_ref, 3),
        name = _ref2[0],
        innerBlocksTemplate = _ref2[2];

    var block = blocks[index];
    return name === block.name && doBlocksMatchTemplate(block.innerBlocks, innerBlocksTemplate);
  });
}
/**
 * Synchronize a block list with a block template.
 *
 * Synchronizing a block list with a block template means that we loop over the blocks
 * keep the block as is if it matches the block at the same position in the template
 * (If it has the same name) and if doesn't match, we create a new block based on the template.
 * Extra blocks not present in the template are removed.
 *
 * @param {Array} blocks    Block list.
 * @param {Array} template  Block template.
 *
 * @return {Array}          Updated Block list.
 */


function synchronizeBlocksWithTemplate() {
  var blocks = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
  var template = arguments.length > 1 ? arguments[1] : undefined;

  // If no template is provided, return blocks unmodified.
  if (!template) {
    return blocks;
  }

  return (0, _lodash.map)(template, function (_ref3, index) {
    var _ref4 = (0, _slicedToArray2.default)(_ref3, 3),
        name = _ref4[0],
        attributes = _ref4[1],
        innerBlocksTemplate = _ref4[2];

    var block = blocks[index];

    if (block && block.name === name) {
      var innerBlocks = synchronizeBlocksWithTemplate(block.innerBlocks, innerBlocksTemplate);
      return _objectSpread(_objectSpread({}, block), {}, {
        innerBlocks: innerBlocks
      });
    } // To support old templates that were using the "children" format
    // for the attributes using "html" strings now, we normalize the template attributes
    // before creating the blocks.


    var blockType = (0, _registration.getBlockType)(name);

    var isHTMLAttribute = function isHTMLAttribute(attributeDefinition) {
      return (0, _lodash.get)(attributeDefinition, ['source']) === 'html';
    };

    var isQueryAttribute = function isQueryAttribute(attributeDefinition) {
      return (0, _lodash.get)(attributeDefinition, ['source']) === 'query';
    };

    var normalizeAttributes = function normalizeAttributes(schema, values) {
      return (0, _lodash.mapValues)(values, function (value, key) {
        return normalizeAttribute(schema[key], value);
      });
    };

    var normalizeAttribute = function normalizeAttribute(definition, value) {
      if (isHTMLAttribute(definition) && (0, _lodash.isArray)(value)) {
        // Introduce a deprecated call at this point
        // When we're confident that "children" format should be removed from the templates.
        return (0, _element.renderToString)(value);
      }

      if (isQueryAttribute(definition) && value) {
        return value.map(function (subValues) {
          return normalizeAttributes(definition.query, subValues);
        });
      }

      return value;
    };

    var normalizedAttributes = normalizeAttributes((0, _lodash.get)(blockType, ['attributes'], {}), attributes);

    var _convertLegacyBlocks = (0, _parser.convertLegacyBlocks)(name, normalizedAttributes),
        blockName = _convertLegacyBlocks.name,
        blockAttributes = _convertLegacyBlocks.attributes;

    return (0, _factory.createBlock)(blockName, blockAttributes, synchronizeBlocksWithTemplate([], innerBlocksTemplate));
  });
}
