// @wordpress/components/build-module/mobile/html-text-input/container.ios.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import { createElement } from "@wordpress/element";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

/**
 * External dependencies
 */
import { UIManager, PanResponder } from 'react-native';
/**
 * WordPress dependencies
 */

import { Component } from '@wordpress/element';
/**
 * Internal dependencies
 */

import KeyboardAvoidingView from '../keyboard-avoiding-view';
import styles from './style.ios.scss';

var HTMLInputContainer = /*#__PURE__*/function (_Component) {
  _inherits(HTMLInputContainer, _Component);

  var _super = _createSuper(HTMLInputContainer);

  function HTMLInputContainer() {
    var _this;

    _classCallCheck(this, HTMLInputContainer);

    _this = _super.apply(this, arguments);
    _this.panResponder = PanResponder.create({
      onStartShouldSetPanResponderCapture: function onStartShouldSetPanResponderCapture() {
        return true;
      },
      onPanResponderMove: function onPanResponderMove(e, gestureState) {
        if (gestureState.dy > 100 && gestureState.dy < 110) {
          //Keyboard.dismiss() and this.textInput.blur() are not working here
          //They require to know the currentlyFocusedID under the hood but
          //during this gesture there's no currentlyFocusedID
          UIManager.blur(e.target);
        }
      }
    });
    return _this;
  }

  _createClass(HTMLInputContainer, [{
    key: "render",
    value: function render() {
      return createElement(KeyboardAvoidingView, _extends({
        style: styles.keyboardAvoidingView
      }, this.panResponder.panHandlers, {
        parentHeight: this.props.parentHeight
      }), this.props.children);
    }
  }]);

  return HTMLInputContainer;
}(Component);

HTMLInputContainer.scrollEnabled = true;
export default HTMLInputContainer;
