// @wordpress/components/build-module/mobile/html-text-input/index.native.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import { createElement } from "@wordpress/element";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

/**
 * External dependencies
 */
import { TextInput } from 'react-native';
/**
 * WordPress dependencies
 */

import { Component } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import { parse } from '@wordpress/blocks';
import { withDispatch, withSelect } from '@wordpress/data';
import { addFilter, removeFilter } from '@wordpress/hooks';
import { withInstanceId, compose, withPreferredColorScheme } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import HTMLInputContainer from './container';
import styles from './style.scss';
export var HTMLTextInput = /*#__PURE__*/function (_Component) {
  _inherits(HTMLTextInput, _Component);

  var _super = _createSuper(HTMLTextInput);

  function HTMLTextInput() {
    var _this;

    _classCallCheck(this, HTMLTextInput);

    _this = _super.apply(this, arguments);
    _this.edit = _this.edit.bind(_assertThisInitialized(_this));
    _this.stopEditing = _this.stopEditing.bind(_assertThisInitialized(_this));
    _this.getHTMLForParent = _this.getHTMLForParent.bind(_assertThisInitialized(_this));
    addFilter('native.persist-html', 'html-text-input', _this.getHTMLForParent);
    _this.state = {};
    return _this;
  }

  _createClass(HTMLTextInput, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      removeFilter('native.persist-html', 'html-text-input'); //TODO: Blocking main thread

      this.stopEditing();
    }
  }, {
    key: "edit",
    value: function edit(html) {
      this.props.onChange(html);
      this.setState({
        value: html,
        isDirty: true
      });
    }
  }, {
    key: "getHTMLForParent",
    value: function getHTMLForParent() {
      return this.state.value;
    }
  }, {
    key: "stopEditing",
    value: function stopEditing() {
      if (this.state.isDirty) {
        this.props.onPersist(this.state.value);
        this.setState({
          isDirty: false
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var getStylesFromColorScheme = this.props.getStylesFromColorScheme;
      var htmlStyle = getStylesFromColorScheme(styles.htmlView, styles.htmlViewDark);
      var placeholderStyle = getStylesFromColorScheme(styles.placeholder, styles.placeholderDark);
      return createElement(HTMLInputContainer, {
        parentHeight: this.props.parentHeight
      }, createElement(TextInput, {
        autoCorrect: false,
        accessibilityLabel: "html-view-title",
        textAlignVertical: "center",
        numberOfLines: 1,
        style: styles.htmlViewTitle,
        value: this.props.title,
        placeholder: __('Add title'),
        placeholderTextColor: placeholderStyle.color,
        onChangeText: this.props.editTitle
      }), createElement(TextInput, {
        autoCorrect: false,
        accessibilityLabel: "html-view-content",
        textAlignVertical: "top",
        multiline: true,
        style: htmlStyle,
        value: this.state.value,
        onChangeText: this.edit,
        onBlur: this.stopEditing,
        placeholder: __('Start writing…'),
        placeholderTextColor: placeholderStyle.color,
        scrollEnabled: HTMLInputContainer.scrollEnabled
      }));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      if (state.isDirty) {
        return null;
      }

      return {
        value: props.value,
        isDirty: false
      };
    }
  }]);

  return HTMLTextInput;
}(Component);
export default compose([withSelect(function (select) {
  var _select = select('core/editor'),
      getEditedPostAttribute = _select.getEditedPostAttribute,
      getEditedPostContent = _select.getEditedPostContent;

  return {
    title: getEditedPostAttribute('title'),
    value: getEditedPostContent()
  };
}), withDispatch(function (dispatch) {
  var _dispatch = dispatch('core/editor'),
      editPost = _dispatch.editPost,
      resetEditorBlocks = _dispatch.resetEditorBlocks;

  return {
    editTitle: function editTitle(title) {
      editPost({
        title: title
      });
    },
    onChange: function onChange(content) {
      editPost({
        content: content
      });
    },
    onPersist: function onPersist(content) {
      var blocks = parse(content);
      resetEditorBlocks(blocks);
    }
  };
}), withInstanceId, withPreferredColorScheme])(HTMLTextInput);
