// @wordpress/components/build/mobile/link-settings/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _reactNative = require("react-native");

var _compose = require("@wordpress/compose");

var _data = require("@wordpress/data");

var _url = require("@wordpress/url");

var _icons = require("@wordpress/icons");

var _bottomSheet = _interopRequireDefault(require("../bottom-sheet"));

var _body = _interopRequireDefault(require("../../panel/body"));

var _textControl = _interopRequireDefault(require("../../text-control"));

var _toggleControl = _interopRequireDefault(require("../../toggle-control"));

var _footerMessageControl = _interopRequireDefault(require("../../footer-message-control"));

var _actions = _interopRequireDefault(require("../../panel/actions"));

var _linkRel = _interopRequireDefault(require("./link-rel"));

var _style = _interopRequireDefault(require("./style.scss"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * Internal dependencies
 */
var NEW_TAB_REL = 'noreferrer noopener';

function LinkSettings(_ref) {
  var isVisible = _ref.isVisible,
      onClose = _ref.onClose,
      attributes = _ref.attributes,
      setAttributes = _ref.setAttributes,
      onEmptyURL = _ref.onEmptyURL,
      options = _ref.options,
      withBottomSheet = _ref.withBottomSheet,
      actions = _ref.actions,
      editorSidebarOpened = _ref.editorSidebarOpened,
      showIcon = _ref.showIcon;
  var url = attributes.url,
      label = attributes.label,
      linkTarget = attributes.linkTarget,
      rel = attributes.rel;

  var _useState = (0, _element.useState)(''),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      urlInputValue = _useState2[0],
      setUrlInputValue = _useState2[1];

  var _useState3 = (0, _element.useState)(''),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      labelInputValue = _useState4[0],
      setLabelInputValue = _useState4[1];

  var _useState5 = (0, _element.useState)(''),
      _useState6 = (0, _slicedToArray2.default)(_useState5, 2),
      linkRelInputValue = _useState6[0],
      setLinkRelInputValue = _useState6[1];

  var prevEditorSidebarOpenedRef = (0, _element.useRef)();
  (0, _element.useEffect)(function () {
    prevEditorSidebarOpenedRef.current = editorSidebarOpened;
  });
  var prevEditorSidebarOpened = prevEditorSidebarOpenedRef.current;
  (0, _element.useEffect)(function () {
    setUrlInputValue(url || '');
  }, [url]);
  (0, _element.useEffect)(function () {
    setLabelInputValue(label || '');
  }, [label]);
  (0, _element.useEffect)(function () {
    setLinkRelInputValue(rel || '');
  }, [rel]);
  (0, _element.useEffect)(function () {
    var isSettingSheetOpen = isVisible || editorSidebarOpened;

    if (options.url.autoFill && isSettingSheetOpen && !url) {
      getURLFromClipboard();
    }

    if (prevEditorSidebarOpened && !editorSidebarOpened) {
      onSetAttributes();
    }
  }, [editorSidebarOpened, isVisible]);

  function onChangeURL(value) {
    if (!value && onEmptyURL) {
      onEmptyURL();
    }

    setUrlInputValue(value);
  }

  function onChangeLabel(value) {
    setLabelInputValue(value);
  }

  function onSetAttributes() {
    setAttributes({
      url: (0, _url.prependHTTP)(urlInputValue),
      label: labelInputValue,
      rel: linkRelInputValue
    });
  }

  function onCloseSettingsSheet() {
    onSetAttributes();
    onClose();
  }

  function onChangeOpenInNewTab(value) {
    var newLinkTarget = value ? '_blank' : undefined;
    var updatedRel = rel;

    if (newLinkTarget && !rel) {
      updatedRel = NEW_TAB_REL;
    } else if (!newLinkTarget && rel === NEW_TAB_REL) {
      updatedRel = undefined;
    }

    setAttributes({
      linkTarget: newLinkTarget,
      rel: updatedRel
    });
  }

  function onChangeLinkRel(value) {
    setLinkRelInputValue(value);
  }

  function getURLFromClipboard() {
    return _getURLFromClipboard.apply(this, arguments);
  }

  function _getURLFromClipboard() {
    _getURLFromClipboard = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
      var clipboardText;
      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return _reactNative.Clipboard.getString();

            case 2:
              clipboardText = _context.sent;

              if (clipboardText) {
                _context.next = 5;
                break;
              }

              return _context.abrupt("return");

            case 5:
              if ((0, _url.isURL)(clipboardText)) {
                _context.next = 7;
                break;
              }

              return _context.abrupt("return");

            case 7:
              setAttributes({
                url: clipboardText
              });

            case 8:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));
    return _getURLFromClipboard.apply(this, arguments);
  }

  function getSettings() {
    return (0, _element.createElement)(_element.Fragment, null, options.url && (0, _element.createElement)(_textControl.default, {
      icon: showIcon && _icons.link,
      label: options.url.label,
      value: urlInputValue,
      valuePlaceholder: options.url.placeholder,
      onChange: onChangeURL,
      onSubmit: onCloseSettingsSheet,
      autoCapitalize: "none",
      autoCorrect: false // eslint-disable-next-line jsx-a11y/no-autofocus
      ,
      autoFocus: _reactNative.Platform.OS === 'ios' && options.url.autoFocus,
      keyboardType: "url"
    }), options.linkLabel && (0, _element.createElement)(_textControl.default, {
      label: options.linkLabel.label,
      value: labelInputValue,
      valuePlaceholder: options.linkLabel.placeholder,
      onChange: onChangeLabel
    }), options.openInNewTab && (0, _element.createElement)(_toggleControl.default, {
      icon: showIcon && _icons.external,
      label: options.openInNewTab.label,
      checked: linkTarget === '_blank',
      onChange: onChangeOpenInNewTab
    }), options.linkRel && (0, _element.createElement)(_textControl.default, {
      icon: showIcon && _linkRel.default,
      label: options.linkRel.label,
      value: linkRelInputValue,
      valuePlaceholder: options.linkRel.placeholder,
      onChange: onChangeLinkRel,
      onSubmit: onCloseSettingsSheet,
      autoCapitalize: "none",
      autoCorrect: false,
      keyboardType: "url"
    }));
  }

  if (!withBottomSheet) {
    return getSettings();
  }

  return (0, _element.createElement)(_bottomSheet.default, {
    isVisible: isVisible,
    onClose: onCloseSettingsSheet,
    hideHeader: true
  }, (0, _element.createElement)(_body.default, {
    style: _style.default.linkSettingsPanel
  }, getSettings()), options.footer && (0, _element.createElement)(_body.default, {
    style: _style.default.linkSettingsPanel
  }, (0, _element.createElement)(_footerMessageControl.default, {
    label: options.footer.label,
    textAlign: "left"
  })), actions && (0, _element.createElement)(_actions.default, {
    actions: actions
  }));
}

var _default = (0, _compose.compose)([(0, _data.withSelect)(function (select) {
  var _select = select('core/edit-post'),
      isEditorSidebarOpened = _select.isEditorSidebarOpened;

  return {
    editorSidebarOpened: isEditorSidebarOpened()
  };
})])(LinkSettings);

exports.default = _default;
