// @wordpress/components/build-module/mobile/image/image-editing-button.native.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { TouchableWithoutFeedback, View } from 'react-native';
/**
 * WordPress dependencies
 */

import { Icon } from '@wordpress/components';
/**
 * Internal dependencies
 */

import { MediaEdit } from '../media-edit';
import SvgIconCustomize from './icon-customize';
import styles from './style.scss';

var ImageEditingButton = function ImageEditingButton(_ref) {
  var onSelectMediaUploadOption = _ref.onSelectMediaUploadOption,
      openMediaOptions = _ref.openMediaOptions,
      pickerOptions = _ref.pickerOptions,
      url = _ref.url;
  return createElement(MediaEdit, {
    onSelect: onSelectMediaUploadOption,
    source: {
      uri: url
    },
    openReplaceMediaOptions: openMediaOptions,
    render: function render(_ref2) {
      var open = _ref2.open,
          mediaOptions = _ref2.mediaOptions;
      return createElement(TouchableWithoutFeedback, {
        onPress: open
      }, createElement(View, {
        style: styles.editContainer
      }, createElement(View, {
        style: styles.edit
      }, mediaOptions(), createElement(Icon, _extends({
        size: 16,
        icon: SvgIconCustomize
      }, styles.iconCustomise)))));
    },
    pickerOptions: pickerOptions
  });
};

export default ImageEditingButton;
