// @wordpress/components/build/date-time/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DatePicker", {
  enumerable: true,
  get: function get() {
    return _date.default;
  }
});
Object.defineProperty(exports, "TimePicker", {
  enumerable: true,
  get: function get() {
    return _time.default;
  }
});
exports.DateTimePicker = void 0;

var _element = require("@wordpress/element");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

require("react-dates/initialize");

var _i18n = require("@wordpress/i18n");

var _button = _interopRequireDefault(require("../button"));

var _date = _interopRequireDefault(require("./date"));

var _time = _interopRequireDefault(require("./time"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var DateTimePicker = /*#__PURE__*/function (_Component) {
  (0, _inherits2.default)(DateTimePicker, _Component);

  var _super = _createSuper(DateTimePicker);

  function DateTimePicker() {
    var _this;

    (0, _classCallCheck2.default)(this, DateTimePicker);
    _this = _super.apply(this, arguments);
    _this.state = {
      calendarHelpIsVisible: false
    };
    _this.onClickDescriptionToggle = _this.onClickDescriptionToggle.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(DateTimePicker, [{
    key: "onClickDescriptionToggle",
    value: function onClickDescriptionToggle() {
      this.setState({
        calendarHelpIsVisible: !this.state.calendarHelpIsVisible
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          currentDate = _this$props.currentDate,
          is12Hour = _this$props.is12Hour,
          isInvalidDate = _this$props.isInvalidDate,
          onMonthPreviewed = _this$props.onMonthPreviewed,
          onChange = _this$props.onChange,
          events = _this$props.events;
      return (0, _element.createElement)("div", {
        className: "components-datetime"
      }, !this.state.calendarHelpIsVisible && (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_time.default, {
        currentTime: currentDate,
        onChange: onChange,
        is12Hour: is12Hour
      }), (0, _element.createElement)(_date.default, {
        currentDate: currentDate,
        onChange: onChange,
        isInvalidDate: isInvalidDate,
        onMonthPreviewed: onMonthPreviewed,
        events: events
      })), this.state.calendarHelpIsVisible && (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)("div", {
        className: "components-datetime__calendar-help"
      }, (0, _element.createElement)("h4", null, (0, _i18n.__)('Click to Select')), (0, _element.createElement)("ul", null, (0, _element.createElement)("li", null, (0, _i18n.__)('Click the right or left arrows to select other months in the past or the future.')), (0, _element.createElement)("li", null, (0, _i18n.__)('Click the desired day to select it.'))), (0, _element.createElement)("h4", null, (0, _i18n.__)('Navigating with a keyboard')), (0, _element.createElement)("ul", null, (0, _element.createElement)("li", null, (0, _element.createElement)("abbr", {
        "aria-label": (0, _i18n._x)('Enter', 'keyboard button')
      }, "\u21B5"), ' '
      /* JSX removes whitespace, but a space is required for screen readers. */
      , (0, _element.createElement)("span", null, (0, _i18n.__)('Select the date in focus.'))), (0, _element.createElement)("li", null, (0, _element.createElement)("abbr", {
        "aria-label": (0, _i18n.__)('Left and Right Arrows')
      }, "\u2190/\u2192"), ' '
      /* JSX removes whitespace, but a space is required for screen readers. */
      , (0, _i18n.__)('Move backward (left) or forward (right) by one day.')), (0, _element.createElement)("li", null, (0, _element.createElement)("abbr", {
        "aria-label": (0, _i18n.__)('Up and Down Arrows')
      }, "\u2191/\u2193"), ' '
      /* JSX removes whitespace, but a space is required for screen readers. */
      , (0, _i18n.__)('Move backward (up) or forward (down) by one week.')), (0, _element.createElement)("li", null, (0, _element.createElement)("abbr", {
        "aria-label": (0, _i18n.__)('Page Up and Page Down')
      }, (0, _i18n.__)('PgUp/PgDn')), ' '
      /* JSX removes whitespace, but a space is required for screen readers. */
      , (0, _i18n.__)('Move backward (PgUp) or forward (PgDn) by one month.')), (0, _element.createElement)("li", null, (0, _element.createElement)("abbr", {
        "aria-label": (0, _i18n.__)('Home and End')
      }, (0, _i18n.__)('Home/End')), ' '
      /* JSX removes whitespace, but a space is required for screen readers. */
      , (0, _i18n.__)('Go to the first (home) or last (end) day of a week.'))))), (0, _element.createElement)("div", {
        className: "components-datetime__buttons"
      }, !this.state.calendarHelpIsVisible && currentDate && (0, _element.createElement)(_button.default, {
        className: "components-datetime__date-reset-button",
        isLink: true,
        onClick: function onClick() {
          return onChange(null);
        }
      }, (0, _i18n.__)('Reset')), (0, _element.createElement)(_button.default, {
        className: "components-datetime__date-help-toggle",
        isLink: true,
        onClick: this.onClickDescriptionToggle
      }, this.state.calendarHelpIsVisible ? (0, _i18n.__)('Close') : (0, _i18n.__)('Calendar Help'))));
    }
  }]);
  return DateTimePicker;
}(_element.Component);

exports.DateTimePicker = DateTimePicker;
